{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract male bias genes on the X chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:10:22.971196Z",
     "iopub.status.busy": "2023-03-14T18:10:22.971073Z",
     "iopub.status.idle": "2023-03-14T18:10:23.144307Z",
     "shell.execute_reply": "2023-03-14T18:10:23.143915Z"
    }
   },
   "outputs": [],
   "source": [
    "import session_info\n",
    "import pandas as pd\n",
    "from pyhere import here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:10:23.146381Z",
     "iopub.status.busy": "2023-03-14T18:10:23.146094Z",
     "iopub.status.idle": "2023-03-14T18:10:23.148666Z",
     "shell.execute_reply": "2023-03-14T18:10:23.148324Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_deg():\n",
    "    fn = here(\"differential_expression/tissue_comparison/summary_table\",\n",
    "              \"_m/differential_expression_analysis_4features_sex.txt.gz\")\n",
    "    df = pd.read_csv(fn, sep='\\t').loc[:, [\"Tissue\", \"Feature\", \"ensemblID\", \"Symbol\", \n",
    "                                           \"seqnames\", \"Type\", \"t\", \"Chrom_Type\"]]\n",
    "    return df[(df[\"Type\"] == \"Gene\")].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:10:23.151004Z",
     "iopub.status.busy": "2023-03-14T18:10:23.150875Z",
     "iopub.status.idle": "2023-03-14T18:11:41.126271Z",
     "shell.execute_reply": "2023-03-14T18:11:41.125740Z"
    }
   },
   "outputs": [],
   "source": [
    "df = get_deg()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.128394Z",
     "iopub.status.busy": "2023-03-14T18:11:41.128277Z",
     "iopub.status.idle": "2023-03-14T18:11:41.246882Z",
     "shell.execute_reply": "2023-03-14T18:11:41.246523Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start position</th>\n",
       "      <th>End position</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Combined XCI status</th>\n",
       "      <th>ensemblID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.8</td>\n",
       "      <td>X</td>\n",
       "      <td>192989</td>\n",
       "      <td>220023</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "      <td>ENSG00000182378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>ENSG00000178605.8</td>\n",
       "      <td>X</td>\n",
       "      <td>220025</td>\n",
       "      <td>230886</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "      <td>ENSG00000178605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene name            Gene ID Chr  Start position  End position  \\\n",
       "0    PLCXD1  ENSG00000182378.8   X          192989        220023   \n",
       "1    GTPBP6  ENSG00000178605.8   X          220025        230886   \n",
       "\n",
       "  Transcript type Combined XCI status        ensemblID  \n",
       "0  protein_coding              escape  ENSG00000182378  \n",
       "1  protein_coding              escape  ENSG00000178605  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xci = pd.read_csv(\"../../_h/xci_status_hg19.txt\", sep='\\t')\n",
    "xci[\"ensemblID\"] = xci[\"Gene ID\"].str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "xci.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.291930Z",
     "iopub.status.busy": "2023-03-14T18:11:41.291818Z",
     "iopub.status.idle": "2023-03-14T18:11:41.297361Z",
     "shell.execute_reply": "2023-03-14T18:11:41.296889Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Combined XCI status\n",
       "escape       99\n",
       "inactive    431\n",
       "variable    101\n",
       "dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xci.groupby(\"Combined XCI status\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.300086Z",
     "iopub.status.busy": "2023-03-14T18:11:41.299874Z",
     "iopub.status.idle": "2023-03-14T18:11:41.334483Z",
     "shell.execute_reply": "2023-03-14T18:11:41.333996Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>Type</th>\n",
       "      <th>t</th>\n",
       "      <th>Chrom_Type</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start position</th>\n",
       "      <th>End position</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Combined XCI status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>CD99|ENSG00000002586.20</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>20.718194</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>CD99</td>\n",
       "      <td>ENSG00000002586.13</td>\n",
       "      <td>X</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2659350</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>DLPFC</td>\n",
       "      <td>CD99|ENSG00000002586.20</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>10.690853</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>CD99</td>\n",
       "      <td>ENSG00000002586.13</td>\n",
       "      <td>X</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2659350</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>CD99|ENSG00000002586.20</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>10.917016</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>CD99</td>\n",
       "      <td>ENSG00000002586.13</td>\n",
       "      <td>X</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2659350</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ZBED1|ENSG00000214717.12</td>\n",
       "      <td>ENSG00000214717</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>14.600959</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>ENSG00000214717.5</td>\n",
       "      <td>X</td>\n",
       "      <td>2404455</td>\n",
       "      <td>2419008</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>DLPFC</td>\n",
       "      <td>ZBED1|ENSG00000214717.12</td>\n",
       "      <td>ENSG00000214717</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>6.980985</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>ENSG00000214717.5</td>\n",
       "      <td>X</td>\n",
       "      <td>2404455</td>\n",
       "      <td>2419008</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>DNAAF6|ENSG00000080572.14</td>\n",
       "      <td>ENSG00000080572</td>\n",
       "      <td>DNAAF6</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.315468</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>PIH1D3</td>\n",
       "      <td>ENSG00000080572.8</td>\n",
       "      <td>X</td>\n",
       "      <td>106449862</td>\n",
       "      <td>106487473</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>DLPFC</td>\n",
       "      <td>IQSEC2|ENSG00000124313.18</td>\n",
       "      <td>ENSG00000124313</td>\n",
       "      <td>IQSEC2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.353653</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>IQSEC2</td>\n",
       "      <td>ENSG00000124313.8</td>\n",
       "      <td>X</td>\n",
       "      <td>53262058</td>\n",
       "      <td>53350522</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>NXF5|ENSG00000126952.18</td>\n",
       "      <td>ENSG00000126952</td>\n",
       "      <td>NXF5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.137540</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>NXF5</td>\n",
       "      <td>ENSG00000126952.12</td>\n",
       "      <td>X</td>\n",
       "      <td>101087085</td>\n",
       "      <td>101112549</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>DLPFC</td>\n",
       "      <td>NXF5|ENSG00000126952.18</td>\n",
       "      <td>ENSG00000126952</td>\n",
       "      <td>NXF5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.122630</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>NXF5</td>\n",
       "      <td>ENSG00000126952.12</td>\n",
       "      <td>X</td>\n",
       "      <td>101087085</td>\n",
       "      <td>101112549</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>NXF5|ENSG00000126952.18</td>\n",
       "      <td>ENSG00000126952</td>\n",
       "      <td>NXF5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.196594</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>NXF5</td>\n",
       "      <td>ENSG00000126952.12</td>\n",
       "      <td>X</td>\n",
       "      <td>101087085</td>\n",
       "      <td>101112549</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Tissue                    Feature        ensemblID  Symbol seqnames  \\\n",
       "27       Caudate    CD99|ENSG00000002586.20  ENSG00000002586    CD99     chrX   \n",
       "28         DLPFC    CD99|ENSG00000002586.20  ENSG00000002586    CD99     chrX   \n",
       "29   Hippocampus    CD99|ENSG00000002586.20  ENSG00000002586    CD99     chrX   \n",
       "36       Caudate   ZBED1|ENSG00000214717.12  ENSG00000214717   ZBED1     chrX   \n",
       "37         DLPFC   ZBED1|ENSG00000214717.12  ENSG00000214717   ZBED1     chrX   \n",
       "..           ...                        ...              ...     ...      ...   \n",
       "213      Caudate  DNAAF6|ENSG00000080572.14  ENSG00000080572  DNAAF6     chrX   \n",
       "216        DLPFC  IQSEC2|ENSG00000124313.18  ENSG00000124313  IQSEC2     chrX   \n",
       "218      Caudate    NXF5|ENSG00000126952.18  ENSG00000126952    NXF5     chrX   \n",
       "219        DLPFC    NXF5|ENSG00000126952.18  ENSG00000126952    NXF5     chrX   \n",
       "220  Hippocampus    NXF5|ENSG00000126952.18  ENSG00000126952    NXF5     chrX   \n",
       "\n",
       "     Type          t Chrom_Type Gene name             Gene ID Chr  \\\n",
       "27   Gene  20.718194   Allosome      CD99  ENSG00000002586.13   X   \n",
       "28   Gene  10.690853   Allosome      CD99  ENSG00000002586.13   X   \n",
       "29   Gene  10.917016   Allosome      CD99  ENSG00000002586.13   X   \n",
       "36   Gene  14.600959   Allosome     ZBED1   ENSG00000214717.5   X   \n",
       "37   Gene   6.980985   Allosome     ZBED1   ENSG00000214717.5   X   \n",
       "..    ...        ...        ...       ...                 ...  ..   \n",
       "213  Gene   0.315468   Allosome    PIH1D3   ENSG00000080572.8   X   \n",
       "216  Gene   0.353653   Allosome    IQSEC2   ENSG00000124313.8   X   \n",
       "218  Gene   0.137540   Allosome      NXF5  ENSG00000126952.12   X   \n",
       "219  Gene   0.122630   Allosome      NXF5  ENSG00000126952.12   X   \n",
       "220  Gene   0.196594   Allosome      NXF5  ENSG00000126952.12   X   \n",
       "\n",
       "     Start position  End position Transcript type Combined XCI status  \n",
       "27          2609220       2659350  protein_coding              escape  \n",
       "28          2609220       2659350  protein_coding              escape  \n",
       "29          2609220       2659350  protein_coding              escape  \n",
       "36          2404455       2419008  protein_coding              escape  \n",
       "37          2404455       2419008  protein_coding              escape  \n",
       "..              ...           ...             ...                 ...  \n",
       "213       106449862     106487473  protein_coding              escape  \n",
       "216        53262058      53350522  protein_coding              escape  \n",
       "218       101087085     101112549  protein_coding              escape  \n",
       "219       101087085     101112549  protein_coding              escape  \n",
       "220       101087085     101112549  protein_coding              escape  \n",
       "\n",
       "[78 rows x 15 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt = df.merge(xci[(xci[\"Combined XCI status\"] == \"escape\")], on=\"ensemblID\")\n",
    "tt[(tt['t'] > 0)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Escaped genes are also located on the PAR regions of the Y chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.337410Z",
     "iopub.status.busy": "2023-03-14T18:11:41.337199Z",
     "iopub.status.idle": "2023-03-14T18:11:41.354865Z",
     "shell.execute_reply": "2023-03-14T18:11:41.354437Z"
    }
   },
   "outputs": [],
   "source": [
    "xlinked = df[(df['seqnames'] == 'chrX')].copy()\n",
    "xx_male = df[(df['seqnames'].isin([\"chrX\", \"chrY\"])) & (df[\"t\"] > 0)].copy()\n",
    "xlinked_male = xlinked[(xlinked[\"t\"] > 0)].copy()\n",
    "xlinked_female = xlinked[(xlinked[\"t\"] < 0)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.357578Z",
     "iopub.status.busy": "2023-03-14T18:11:41.357412Z",
     "iopub.status.idle": "2023-03-14T18:11:41.361755Z",
     "shell.execute_reply": "2023-03-14T18:11:41.361365Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "Caudate        871\n",
       "DLPFC          858\n",
       "Hippocampus    876\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked.groupby(\"Tissue\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.364290Z",
     "iopub.status.busy": "2023-03-14T18:11:41.364128Z",
     "iopub.status.idle": "2023-03-14T18:11:41.368263Z",
     "shell.execute_reply": "2023-03-14T18:11:41.367882Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "Caudate        492\n",
       "DLPFC          462\n",
       "Hippocampus    392\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_male.groupby(\"Tissue\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.370797Z",
     "iopub.status.busy": "2023-03-14T18:11:41.370636Z",
     "iopub.status.idle": "2023-03-14T18:11:41.374704Z",
     "shell.execute_reply": "2023-03-14T18:11:41.374323Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "Caudate        379\n",
       "DLPFC          396\n",
       "Hippocampus    484\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_female.groupby(\"Tissue\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.377381Z",
     "iopub.status.busy": "2023-03-14T18:11:41.376996Z",
     "iopub.status.idle": "2023-03-14T18:11:41.386257Z",
     "shell.execute_reply": "2023-03-14T18:11:41.385871Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>Type</th>\n",
       "      <th>t</th>\n",
       "      <th>Chrom_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>CD99|ENSG00000002586.20</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>20.718194</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>PRKCIP1|ENSG00000237682.2</td>\n",
       "      <td>ENSG00000237682</td>\n",
       "      <td>PRKCIP1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>17.897618</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ZBED1|ENSG00000214717.12</td>\n",
       "      <td>ENSG00000214717</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>14.600959</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>DHRSX|ENSG00000169084.15</td>\n",
       "      <td>ENSG00000169084</td>\n",
       "      <td>DHRSX</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13.689327</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000289007|ENSG00000289007.2</td>\n",
       "      <td>ENSG00000289007</td>\n",
       "      <td>ENSG00000289007</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>12.947296</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1359849</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>ZC4H2|ENSG00000126970.17</td>\n",
       "      <td>ENSG00000126970</td>\n",
       "      <td>ZC4H2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.011004</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1359856</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>ZDHHC15|ENSG00000102383.14</td>\n",
       "      <td>ENSG00000102383</td>\n",
       "      <td>ZDHHC15</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.010772</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1359895</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>AMOT|ENSG00000126016.17</td>\n",
       "      <td>ENSG00000126016</td>\n",
       "      <td>AMOT</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.009385</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360040</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>DANT2|ENSG00000235244.6</td>\n",
       "      <td>ENSG00000235244</td>\n",
       "      <td>DANT2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.002949</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360062</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>NUS1P1|ENSG00000235636.1</td>\n",
       "      <td>ENSG00000235636</td>\n",
       "      <td>NUS1P1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.001518</td>\n",
       "      <td>Allosome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1346 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Tissue                            Feature        ensemblID  \\\n",
       "53           Caudate            CD99|ENSG00000002586.20  ENSG00000002586   \n",
       "56           Caudate          PRKCIP1|ENSG00000237682.2  ENSG00000237682   \n",
       "62           Caudate           ZBED1|ENSG00000214717.12  ENSG00000214717   \n",
       "65           Caudate           DHRSX|ENSG00000169084.15  ENSG00000169084   \n",
       "66           Caudate  ENSG00000289007|ENSG00000289007.2  ENSG00000289007   \n",
       "...              ...                                ...              ...   \n",
       "1359849  Hippocampus           ZC4H2|ENSG00000126970.17  ENSG00000126970   \n",
       "1359856  Hippocampus         ZDHHC15|ENSG00000102383.14  ENSG00000102383   \n",
       "1359895  Hippocampus            AMOT|ENSG00000126016.17  ENSG00000126016   \n",
       "1360040  Hippocampus            DANT2|ENSG00000235244.6  ENSG00000235244   \n",
       "1360062  Hippocampus           NUS1P1|ENSG00000235636.1  ENSG00000235636   \n",
       "\n",
       "                  Symbol seqnames  Type          t Chrom_Type  \n",
       "53                  CD99     chrX  Gene  20.718194   Allosome  \n",
       "56               PRKCIP1     chrX  Gene  17.897618   Allosome  \n",
       "62                 ZBED1     chrX  Gene  14.600959   Allosome  \n",
       "65                 DHRSX     chrX  Gene  13.689327   Allosome  \n",
       "66       ENSG00000289007     chrX  Gene  12.947296   Allosome  \n",
       "...                  ...      ...   ...        ...        ...  \n",
       "1359849            ZC4H2     chrX  Gene   0.011004   Allosome  \n",
       "1359856          ZDHHC15     chrX  Gene   0.010772   Allosome  \n",
       "1359895             AMOT     chrX  Gene   0.009385   Allosome  \n",
       "1360040            DANT2     chrX  Gene   0.002949   Allosome  \n",
       "1360062           NUS1P1     chrX  Gene   0.001518   Allosome  \n",
       "\n",
       "[1346 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_male"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.389249Z",
     "iopub.status.busy": "2023-03-14T18:11:41.388842Z",
     "iopub.status.idle": "2023-03-14T18:11:41.403471Z",
     "shell.execute_reply": "2023-03-14T18:11:41.403043Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>Type</th>\n",
       "      <th>t</th>\n",
       "      <th>Chrom_Type</th>\n",
       "      <th>Combined XCI status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>CD99|ENSG00000002586.20</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>20.718194</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>PRKCIP1|ENSG00000237682.2</td>\n",
       "      <td>ENSG00000237682</td>\n",
       "      <td>PRKCIP1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>17.897618</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ZBED1|ENSG00000214717.12</td>\n",
       "      <td>ENSG00000214717</td>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>14.600959</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>DHRSX|ENSG00000169084.15</td>\n",
       "      <td>ENSG00000169084</td>\n",
       "      <td>DHRSX</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>13.689327</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000289007|ENSG00000289007.2</td>\n",
       "      <td>ENSG00000289007</td>\n",
       "      <td>ENSG00000289007</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>12.947296</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1341</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>ZC4H2|ENSG00000126970.17</td>\n",
       "      <td>ENSG00000126970</td>\n",
       "      <td>ZC4H2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.011004</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1342</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>ZDHHC15|ENSG00000102383.14</td>\n",
       "      <td>ENSG00000102383</td>\n",
       "      <td>ZDHHC15</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.010772</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1343</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>AMOT|ENSG00000126016.17</td>\n",
       "      <td>ENSG00000126016</td>\n",
       "      <td>AMOT</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.009385</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1344</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>DANT2|ENSG00000235244.6</td>\n",
       "      <td>ENSG00000235244</td>\n",
       "      <td>DANT2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.002949</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1345</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>NUS1P1|ENSG00000235636.1</td>\n",
       "      <td>ENSG00000235636</td>\n",
       "      <td>NUS1P1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>Gene</td>\n",
       "      <td>0.001518</td>\n",
       "      <td>Allosome</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1346 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Tissue                            Feature        ensemblID  \\\n",
       "0         Caudate            CD99|ENSG00000002586.20  ENSG00000002586   \n",
       "1         Caudate          PRKCIP1|ENSG00000237682.2  ENSG00000237682   \n",
       "2         Caudate           ZBED1|ENSG00000214717.12  ENSG00000214717   \n",
       "3         Caudate           DHRSX|ENSG00000169084.15  ENSG00000169084   \n",
       "4         Caudate  ENSG00000289007|ENSG00000289007.2  ENSG00000289007   \n",
       "...           ...                                ...              ...   \n",
       "1341  Hippocampus           ZC4H2|ENSG00000126970.17  ENSG00000126970   \n",
       "1342  Hippocampus         ZDHHC15|ENSG00000102383.14  ENSG00000102383   \n",
       "1343  Hippocampus            AMOT|ENSG00000126016.17  ENSG00000126016   \n",
       "1344  Hippocampus            DANT2|ENSG00000235244.6  ENSG00000235244   \n",
       "1345  Hippocampus           NUS1P1|ENSG00000235636.1  ENSG00000235636   \n",
       "\n",
       "               Symbol seqnames  Type          t Chrom_Type Combined XCI status  \n",
       "0                CD99     chrX  Gene  20.718194   Allosome              escape  \n",
       "1             PRKCIP1     chrX  Gene  17.897618   Allosome             unknown  \n",
       "2               ZBED1     chrX  Gene  14.600959   Allosome              escape  \n",
       "3               DHRSX     chrX  Gene  13.689327   Allosome              escape  \n",
       "4     ENSG00000289007     chrX  Gene  12.947296   Allosome             unknown  \n",
       "...               ...      ...   ...        ...        ...                 ...  \n",
       "1341            ZC4H2     chrX  Gene   0.011004   Allosome            inactive  \n",
       "1342          ZDHHC15     chrX  Gene   0.010772   Allosome            inactive  \n",
       "1343             AMOT     chrX  Gene   0.009385   Allosome            inactive  \n",
       "1344            DANT2     chrX  Gene   0.002949   Allosome             unknown  \n",
       "1345           NUS1P1     chrX  Gene   0.001518   Allosome             unknown  \n",
       "\n",
       "[1346 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_male.merge(xci[[\"ensemblID\", \"Combined XCI status\"]], on=\"ensemblID\", how=\"left\").fillna(\"unknown\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.405980Z",
     "iopub.status.busy": "2023-03-14T18:11:41.405814Z",
     "iopub.status.idle": "2023-03-14T18:11:41.414073Z",
     "shell.execute_reply": "2023-03-14T18:11:41.413678Z"
    }
   },
   "outputs": [],
   "source": [
    "dx = xlinked_male.merge(xci[[\"ensemblID\", \"Combined XCI status\"]], on=\"ensemblID\", how=\"left\").fillna(\"unknown\")\n",
    "dx = dx[(dx[\"Combined XCI status\"] == \"unknown\")].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:41.416682Z",
     "iopub.status.busy": "2023-03-14T18:11:41.416332Z",
     "iopub.status.idle": "2023-03-14T18:11:42.095224Z",
     "shell.execute_reply": "2023-03-14T18:11:42.094616Z"
    }
   },
   "outputs": [],
   "source": [
    "pd.concat([xx_male.merge(xci[[\"ensemblID\", \"Combined XCI status\"]], on=\"ensemblID\"), dx], axis=0)\\\n",
    "  .sort_values([\"Tissue\",  \"Combined XCI status\", \"seqnames\"], ascending=True)\\\n",
    "  .to_csv(\"BrainSeq_male_biased_genes_XCI_status.tsv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Session information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-03-14T18:11:42.098666Z",
     "iopub.status.busy": "2023-03-14T18:11:42.098442Z",
     "iopub.status.idle": "2023-03-14T18:11:43.275502Z",
     "shell.execute_reply": "2023-03-14T18:11:43.275107Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "pandas              1.5.3\n",
       "pyhere              1.0.0\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "asttokens           NA\n",
       "backcall            0.2.0\n",
       "colorama            0.4.6\n",
       "cython_runtime      NA\n",
       "dateutil            2.8.2\n",
       "debugpy             1.6.3\n",
       "decorator           5.1.1\n",
       "entrypoints         0.4\n",
       "executing           1.0.0\n",
       "ipykernel           6.15.2\n",
       "ipython_genutils    0.2.0\n",
       "jaraco              NA\n",
       "jedi                0.18.2\n",
       "jupyter_server      1.18.1\n",
       "more_itertools      9.0.0\n",
       "mpl_toolkits        NA\n",
       "numpy               1.23.5\n",
       "packaging           21.3\n",
       "parso               0.8.3\n",
       "pexpect             4.8.0\n",
       "pickleshare         0.7.5\n",
       "pkg_resources       NA\n",
       "platformdirs        3.0.0\n",
       "prompt_toolkit      3.0.31\n",
       "psutil              5.9.2\n",
       "ptyprocess          0.7.0\n",
       "pure_eval           0.2.2\n",
       "pydev_ipython       NA\n",
       "pydevconsole        NA\n",
       "pydevd              2.8.0\n",
       "pydevd_file_utils   NA\n",
       "pydevd_plugins      NA\n",
       "pydevd_tracing      NA\n",
       "pygments            2.13.0\n",
       "pyparsing           3.0.9\n",
       "pytz                2022.2.1\n",
       "ruamel              NA\n",
       "six                 1.16.0\n",
       "stack_data          0.5.0\n",
       "tornado             6.2\n",
       "traitlets           5.3.0\n",
       "wcwidth             0.2.5\n",
       "zmq                 23.2.1\n",
       "zoneinfo            NA\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.5.0\n",
       "jupyter_client      7.3.5\n",
       "jupyter_core        4.11.1\n",
       "jupyterlab          3.4.6\n",
       "notebook            6.4.12\n",
       "-----\n",
       "Python 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0]\n",
       "Linux-6.2.2-arch2-1-x86_64-with-glibc2.37\n",
       "-----\n",
       "Session information updated at 2023-03-14 14:11\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
