{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract overlapping genes with CMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:21:04.076867Z",
     "iopub.status.busy": "2021-08-06T15:21:04.075376Z",
     "iopub.status.idle": "2021-08-06T15:21:04.384491Z",
     "shell.execute_reply": "2021-08-06T15:21:04.383865Z"
    }
   },
   "outputs": [],
   "source": [
    "import functools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from gtfparse import read_gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:21:04.390514Z",
     "iopub.status.busy": "2021-08-06T15:21:04.388969Z",
     "iopub.status.idle": "2021-08-06T15:21:04.391792Z",
     "shell.execute_reply": "2021-08-06T15:21:04.392777Z"
    }
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_gtf(gtf_file):\n",
    "    return read_gtf(gtf_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:21:04.398280Z",
     "iopub.status.busy": "2021-08-06T15:21:04.397442Z",
     "iopub.status.idle": "2021-08-06T15:21:04.401271Z",
     "shell.execute_reply": "2021-08-06T15:21:04.400399Z"
    }
   },
   "outputs": [],
   "source": [
    "def gene_annotation(gtf_file, feature):\n",
    "    gtf0 = get_gtf(gtf_file)\n",
    "    gtf = gtf0[gtf0[\"feature\"] == feature]\n",
    "    return gtf[[\"gene_id\", \"gene_name\", \"gene_type\", \n",
    "                \"seqname\", \"start\", \"end\", \"strand\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:21:04.408344Z",
     "iopub.status.busy": "2021-08-06T15:21:04.407130Z",
     "iopub.status.idle": "2021-08-06T15:22:18.425270Z",
     "shell.execute_reply": "2021-08-06T15:22:18.426308Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_status', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_status', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    }
   ],
   "source": [
    "gtf_file = '/ceph/genome/human/gencode25/gtf.CHR/_m/gencode.v25.annotation.gtf'\n",
    "gtf_annot = gene_annotation(gtf_file, 'gene')\n",
    "gtf_annot['ensemblID'] = gtf_annot.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Male specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.435597Z",
     "iopub.status.busy": "2021-08-06T15:22:18.434369Z",
     "iopub.status.idle": "2021-08-06T15:22:18.497249Z",
     "shell.execute_reply": "2021-08-06T15:22:18.496019Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000272512.1</td>\n",
       "      <td>RP11-54O7.17</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>995966</td>\n",
       "      <td>998051</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000272512</td>\n",
       "      <td>1386</td>\n",
       "      <td>ENSG00000272512.1</td>\n",
       "      <td>RP11-54O7.17</td>\n",
       "      <td>chr1</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000107404.18</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1335276</td>\n",
       "      <td>1349350</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000107404</td>\n",
       "      <td>4087</td>\n",
       "      <td>ENSG00000107404.18</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id_x   gene_name_x     gene_type_x seqname_x    start  \\\n",
       "0   ENSG00000272512.1  RP11-54O7.17         lincRNA      chr1   995966   \n",
       "1  ENSG00000107404.18          DVL1  protein_coding      chr1  1335276   \n",
       "\n",
       "       end strand        ensemblID  Unnamed: 0           gene_id_y  \\\n",
       "0   998051      -  ENSG00000272512        1386   ENSG00000272512.1   \n",
       "1  1349350      -  ENSG00000107404        4087  ENSG00000107404.18   \n",
       "\n",
       "    gene_name_y seqname_y     gene_type_y  Caudate  DLPFC  Hippocampus  \\\n",
       "0  RP11-54O7.17      chr1         lincRNA        1      0            0   \n",
       "1          DVL1      chr1  protein_coding        1      0            0   \n",
       "\n",
       "   CMC DLPFC  \n",
       "0          0  \n",
       "1          0  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmc_file = \"../../_m/cmc_all_deg_across_tissues_maleSpecific.csv\"\n",
    "df = gtf_annot.merge(pd.read_csv(cmc_file), on='ensemblID')\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping Caudate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.504968Z",
     "iopub.status.busy": "2021-08-06T15:22:18.503614Z",
     "iopub.status.idle": "2021-08-06T15:22:18.531412Z",
     "shell.execute_reply": "2021-08-06T15:22:18.532419Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 64 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSG00000189410.11</td>\n",
       "      <td>SH2D5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20719732</td>\n",
       "      <td>20732837</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000189410</td>\n",
       "      <td>30978</td>\n",
       "      <td>ENSG00000189410.11</td>\n",
       "      <td>SH2D5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>ENSG00000066185.12</td>\n",
       "      <td>ZMYND12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>42430329</td>\n",
       "      <td>42456267</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000066185</td>\n",
       "      <td>64447</td>\n",
       "      <td>ENSG00000066185.12</td>\n",
       "      <td>ZMYND12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>ENSG00000115170.13</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2</td>\n",
       "      <td>157736444</td>\n",
       "      <td>157875862</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000115170</td>\n",
       "      <td>341965</td>\n",
       "      <td>ENSG00000115170.13</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>ENSG00000260400.1</td>\n",
       "      <td>RP11-119F7.5</td>\n",
       "      <td>sense_overlapping</td>\n",
       "      <td>chr10</td>\n",
       "      <td>68698500</td>\n",
       "      <td>68700794</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000260400</td>\n",
       "      <td>1260519</td>\n",
       "      <td>ENSG00000260400.1</td>\n",
       "      <td>RP11-119F7.5</td>\n",
       "      <td>chr10</td>\n",
       "      <td>sense_overlapping</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>ENSG00000139372.14</td>\n",
       "      <td>TDG</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr12</td>\n",
       "      <td>103965804</td>\n",
       "      <td>103988874</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000139372</td>\n",
       "      <td>1584469</td>\n",
       "      <td>ENSG00000139372.14</td>\n",
       "      <td>TDG</td>\n",
       "      <td>chr12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1870</th>\n",
       "      <td>ENSG00000256463.8</td>\n",
       "      <td>SALL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>78980275</td>\n",
       "      <td>79002677</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000256463</td>\n",
       "      <td>2184098</td>\n",
       "      <td>ENSG00000256463.8</td>\n",
       "      <td>SALL3</td>\n",
       "      <td>chr18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2116</th>\n",
       "      <td>ENSG00000100116.16</td>\n",
       "      <td>GCAT</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr22</td>\n",
       "      <td>37807905</td>\n",
       "      <td>37817176</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000100116</td>\n",
       "      <td>2469692</td>\n",
       "      <td>ENSG00000100116.16</td>\n",
       "      <td>GCAT</td>\n",
       "      <td>chr22</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2127</th>\n",
       "      <td>ENSG00000100266.18</td>\n",
       "      <td>PACSIN2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr22</td>\n",
       "      <td>42835412</td>\n",
       "      <td>43015145</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000100266</td>\n",
       "      <td>2479888</td>\n",
       "      <td>ENSG00000100266.18</td>\n",
       "      <td>PACSIN2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2193</th>\n",
       "      <td>ENSG00000134597.14</td>\n",
       "      <td>RBMX2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>130401969</td>\n",
       "      <td>130413343</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000134597</td>\n",
       "      <td>2550700</td>\n",
       "      <td>ENSG00000134597.14</td>\n",
       "      <td>RBMX2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id_x   gene_name_x        gene_type_x seqname_x  \\\n",
       "29    ENSG00000189410.11         SH2D5     protein_coding      chr1   \n",
       "50    ENSG00000066185.12       ZMYND12     protein_coding      chr1   \n",
       "286   ENSG00000115170.13         ACVR1     protein_coding      chr2   \n",
       "1154   ENSG00000260400.1  RP11-119F7.5  sense_overlapping     chr10   \n",
       "1397  ENSG00000139372.14           TDG     protein_coding     chr12   \n",
       "1870   ENSG00000256463.8         SALL3     protein_coding     chr18   \n",
       "2116  ENSG00000100116.16          GCAT     protein_coding     chr22   \n",
       "2127  ENSG00000100266.18       PACSIN2     protein_coding     chr22   \n",
       "2193  ENSG00000134597.14         RBMX2     protein_coding      chrX   \n",
       "\n",
       "          start        end strand        ensemblID  Unnamed: 0  \\\n",
       "29     20719732   20732837      -  ENSG00000189410       30978   \n",
       "50     42430329   42456267      -  ENSG00000066185       64447   \n",
       "286   157736444  157875862      -  ENSG00000115170      341965   \n",
       "1154   68698500   68700794      +  ENSG00000260400     1260519   \n",
       "1397  103965804  103988874      +  ENSG00000139372     1584469   \n",
       "1870   78980275   79002677      +  ENSG00000256463     2184098   \n",
       "2116   37807905   37817176      +  ENSG00000100116     2469692   \n",
       "2127   42835412   43015145      -  ENSG00000100266     2479888   \n",
       "2193  130401969  130413343      +  ENSG00000134597     2550700   \n",
       "\n",
       "               gene_id_y   gene_name_y seqname_y        gene_type_y  Caudate  \\\n",
       "29    ENSG00000189410.11         SH2D5      chr1     protein_coding        1   \n",
       "50    ENSG00000066185.12       ZMYND12      chr1     protein_coding        1   \n",
       "286   ENSG00000115170.13         ACVR1      chr2     protein_coding        1   \n",
       "1154   ENSG00000260400.1  RP11-119F7.5     chr10  sense_overlapping        1   \n",
       "1397  ENSG00000139372.14           TDG     chr12     protein_coding        1   \n",
       "1870   ENSG00000256463.8         SALL3     chr18     protein_coding        1   \n",
       "2116  ENSG00000100116.16          GCAT     chr22     protein_coding        1   \n",
       "2127  ENSG00000100266.18       PACSIN2     chr22     protein_coding        1   \n",
       "2193  ENSG00000134597.14         RBMX2      chrX     protein_coding        1   \n",
       "\n",
       "      DLPFC  Hippocampus  CMC DLPFC  \n",
       "29        0            0          1  \n",
       "50        0            0          1  \n",
       "286       0            0          1  \n",
       "1154      0            0          1  \n",
       "1397      0            0          1  \n",
       "1870      0            0          1  \n",
       "2116      0            0          1  \n",
       "2127      0            0          1  \n",
       "2193      0            0          1  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['Caudate'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping DLPFC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.549373Z",
     "iopub.status.busy": "2021-08-06T15:22:18.538115Z",
     "iopub.status.idle": "2021-08-06T15:22:18.553465Z",
     "shell.execute_reply": "2021-08-06T15:22:18.554478Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ENSG00000171488.14</td>\n",
       "      <td>LRRC8C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>89633072</td>\n",
       "      <td>89769903</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000171488</td>\n",
       "      <td>104025</td>\n",
       "      <td>ENSG00000171488.14</td>\n",
       "      <td>LRRC8C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>ENSG00000231752.5</td>\n",
       "      <td>EMBP1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr1</td>\n",
       "      <td>121519112</td>\n",
       "      <td>121571892</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000231752</td>\n",
       "      <td>128562</td>\n",
       "      <td>ENSG00000231752.5</td>\n",
       "      <td>EMBP1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>ENSG00000156414.18</td>\n",
       "      <td>TDRD9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr14</td>\n",
       "      <td>103928462</td>\n",
       "      <td>104052667</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000156414</td>\n",
       "      <td>1748741</td>\n",
       "      <td>ENSG00000156414.18</td>\n",
       "      <td>TDRD9</td>\n",
       "      <td>chr14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id_x gene_name_x                         gene_type_x  \\\n",
       "80    ENSG00000171488.14      LRRC8C                      protein_coding   \n",
       "104    ENSG00000231752.5       EMBP1  transcribed_unprocessed_pseudogene   \n",
       "1537  ENSG00000156414.18       TDRD9                      protein_coding   \n",
       "\n",
       "     seqname_x      start        end strand        ensemblID  Unnamed: 0  \\\n",
       "80        chr1   89633072   89769903      +  ENSG00000171488      104025   \n",
       "104       chr1  121519112  121571892      +  ENSG00000231752      128562   \n",
       "1537     chr14  103928462  104052667      +  ENSG00000156414     1748741   \n",
       "\n",
       "               gene_id_y gene_name_y seqname_y  \\\n",
       "80    ENSG00000171488.14      LRRC8C      chr1   \n",
       "104    ENSG00000231752.5       EMBP1      chr1   \n",
       "1537  ENSG00000156414.18       TDRD9     chr14   \n",
       "\n",
       "                             gene_type_y  Caudate  DLPFC  Hippocampus  \\\n",
       "80                        protein_coding        0      1            0   \n",
       "104   transcribed_unprocessed_pseudogene        0      1            0   \n",
       "1537                      protein_coding        0      1            0   \n",
       "\n",
       "      CMC DLPFC  \n",
       "80            1  \n",
       "104           1  \n",
       "1537          1  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['DLPFC'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping Hippocampus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.569414Z",
     "iopub.status.busy": "2021-08-06T15:22:18.560317Z",
     "iopub.status.idle": "2021-08-06T15:22:18.574810Z",
     "shell.execute_reply": "2021-08-06T15:22:18.573681Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id_x, gene_name_x, gene_type_x, seqname_x, start, end, strand, ensemblID, Unnamed: 0, gene_id_y, gene_name_y, seqname_y, gene_type_y, Caudate, DLPFC, Hippocampus, CMC DLPFC]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['Hippocampus'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping Caudate & DLPFC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.587621Z",
     "iopub.status.busy": "2021-08-06T15:22:18.586367Z",
     "iopub.status.idle": "2021-08-06T15:22:18.590538Z",
     "shell.execute_reply": "2021-08-06T15:22:18.591561Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id_x, gene_name_x, gene_type_x, seqname_x, start, end, strand, ensemblID, Unnamed: 0, gene_id_y, gene_name_y, seqname_y, gene_type_y, Caudate, DLPFC, Hippocampus, CMC DLPFC]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['Caudate'] == 1) & (df['DLPFC'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Female specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.598315Z",
     "iopub.status.busy": "2021-08-06T15:22:18.597071Z",
     "iopub.status.idle": "2021-08-06T15:22:18.640531Z",
     "shell.execute_reply": "2021-08-06T15:22:18.639340Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000272455.1</td>\n",
       "      <td>RP4-758J18.13</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1409096</td>\n",
       "      <td>1410618</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000272455</td>\n",
       "      <td>4753</td>\n",
       "      <td>ENSG00000272455.1</td>\n",
       "      <td>RP4-758J18.13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142609.17</td>\n",
       "      <td>CFAP74</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1921951</td>\n",
       "      <td>2003837</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000142609</td>\n",
       "      <td>7511</td>\n",
       "      <td>ENSG00000142609.17</td>\n",
       "      <td>CFAP74</td>\n",
       "      <td>chr1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id_x    gene_name_x     gene_type_x seqname_x    start  \\\n",
       "0   ENSG00000272455.1  RP4-758J18.13         lincRNA      chr1  1409096   \n",
       "1  ENSG00000142609.17         CFAP74  protein_coding      chr1  1921951   \n",
       "\n",
       "       end strand        ensemblID  Unnamed: 0           gene_id_y  \\\n",
       "0  1410618      +  ENSG00000272455        4753   ENSG00000272455.1   \n",
       "1  2003837      -  ENSG00000142609        7511  ENSG00000142609.17   \n",
       "\n",
       "     gene_name_y seqname_y     gene_type_y  Caudate  DLPFC  Hippocampus  \\\n",
       "0  RP4-758J18.13      chr1         lincRNA        0      0            0   \n",
       "1         CFAP74      chr1  protein_coding        0      0            0   \n",
       "\n",
       "   CMC DLPFC  \n",
       "0          1  \n",
       "1          1  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmc_file = \"../../_m/cmc_all_deg_across_tissues_femaleSpecific.csv\"\n",
    "df = gtf_annot.merge(pd.read_csv(cmc_file), on='ensemblID')\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping Caudate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.648257Z",
     "iopub.status.busy": "2021-08-06T15:22:18.647040Z",
     "iopub.status.idle": "2021-08-06T15:22:18.665296Z",
     "shell.execute_reply": "2021-08-06T15:22:18.666375Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>ENSG00000249669.8</td>\n",
       "      <td>CARMN</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr5</td>\n",
       "      <td>149406689</td>\n",
       "      <td>149432835</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000249669</td>\n",
       "      <td>770108</td>\n",
       "      <td>ENSG00000249669.8</td>\n",
       "      <td>CARMN</td>\n",
       "      <td>chr5</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>ENSG00000100814.17</td>\n",
       "      <td>CCNB1IP1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr14</td>\n",
       "      <td>20311368</td>\n",
       "      <td>20333312</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000100814</td>\n",
       "      <td>1661524</td>\n",
       "      <td>ENSG00000100814.17</td>\n",
       "      <td>CCNB1IP1</td>\n",
       "      <td>chr14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>ENSG00000269937.1</td>\n",
       "      <td>RP11-20I23.8</td>\n",
       "      <td>antisense</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2561471</td>\n",
       "      <td>2565096</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000269937</td>\n",
       "      <td>1867195</td>\n",
       "      <td>ENSG00000269937.1</td>\n",
       "      <td>RP11-20I23.8</td>\n",
       "      <td>chr16</td>\n",
       "      <td>antisense</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>ENSG00000167703.14</td>\n",
       "      <td>SLC43A2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr17</td>\n",
       "      <td>1569267</td>\n",
       "      <td>1628886</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000167703</td>\n",
       "      <td>1979014</td>\n",
       "      <td>ENSG00000167703.14</td>\n",
       "      <td>SLC43A2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>ROCK1P1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr18</td>\n",
       "      <td>109065</td>\n",
       "      <td>122219</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000263006</td>\n",
       "      <td>2140998</td>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>ROCK1P1</td>\n",
       "      <td>chr18</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id_x   gene_name_x                         gene_type_x  \\\n",
       "208   ENSG00000249669.8         CARMN                             lincRNA   \n",
       "474  ENSG00000100814.17      CCNB1IP1                      protein_coding   \n",
       "529   ENSG00000269937.1  RP11-20I23.8                           antisense   \n",
       "555  ENSG00000167703.14       SLC43A2                      protein_coding   \n",
       "588   ENSG00000263006.6       ROCK1P1  transcribed_unprocessed_pseudogene   \n",
       "\n",
       "    seqname_x      start        end strand        ensemblID  Unnamed: 0  \\\n",
       "208      chr5  149406689  149432835      +  ENSG00000249669      770108   \n",
       "474     chr14   20311368   20333312      -  ENSG00000100814     1661524   \n",
       "529     chr16    2561471    2565096      -  ENSG00000269937     1867195   \n",
       "555     chr17    1569267    1628886      -  ENSG00000167703     1979014   \n",
       "588     chr18     109065     122219      +  ENSG00000263006     2140998   \n",
       "\n",
       "              gene_id_y   gene_name_y seqname_y  \\\n",
       "208   ENSG00000249669.8         CARMN      chr5   \n",
       "474  ENSG00000100814.17      CCNB1IP1     chr14   \n",
       "529   ENSG00000269937.1  RP11-20I23.8     chr16   \n",
       "555  ENSG00000167703.14       SLC43A2     chr17   \n",
       "588   ENSG00000263006.6       ROCK1P1     chr18   \n",
       "\n",
       "                            gene_type_y  Caudate  DLPFC  Hippocampus  \\\n",
       "208                             lincRNA        1      0            0   \n",
       "474                      protein_coding        1      0            0   \n",
       "529                           antisense        1      0            0   \n",
       "555                      protein_coding        1      0            0   \n",
       "588  transcribed_unprocessed_pseudogene        1      0            0   \n",
       "\n",
       "     CMC DLPFC  \n",
       "208          1  \n",
       "474          1  \n",
       "529          1  \n",
       "555          1  \n",
       "588          1  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['Caudate'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping DLPFC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.677943Z",
     "iopub.status.busy": "2021-08-06T15:22:18.672466Z",
     "iopub.status.idle": "2021-08-06T15:22:18.682173Z",
     "shell.execute_reply": "2021-08-06T15:22:18.683201Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id_x, gene_name_x, gene_type_x, seqname_x, start, end, strand, ensemblID, Unnamed: 0, gene_id_y, gene_name_y, seqname_y, gene_type_y, Caudate, DLPFC, Hippocampus, CMC DLPFC]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['DLPFC'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping Hippocampus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.695510Z",
     "iopub.status.busy": "2021-08-06T15:22:18.694292Z",
     "iopub.status.idle": "2021-08-06T15:22:18.698470Z",
     "shell.execute_reply": "2021-08-06T15:22:18.699502Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id_x, gene_name_x, gene_type_x, seqname_x, start, end, strand, ensemblID, Unnamed: 0, gene_id_y, gene_name_y, seqname_y, gene_type_y, Caudate, DLPFC, Hippocampus, CMC DLPFC]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['Hippocampus'] == 1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CMC DLPFC overlapping Caudate & DLPFC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-08-06T15:22:18.712456Z",
     "iopub.status.busy": "2021-08-06T15:22:18.711182Z",
     "iopub.status.idle": "2021-08-06T15:22:18.715461Z",
     "shell.execute_reply": "2021-08-06T15:22:18.716503Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_x</th>\n",
       "      <th>gene_name_x</th>\n",
       "      <th>gene_type_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>gene_name_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>gene_type_y</th>\n",
       "      <th>Caudate</th>\n",
       "      <th>DLPFC</th>\n",
       "      <th>Hippocampus</th>\n",
       "      <th>CMC DLPFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id_x, gene_name_x, gene_type_x, seqname_x, start, end, strand, ensemblID, Unnamed: 0, gene_id_y, gene_name_y, seqname_y, gene_type_y, Caudate, DLPFC, Hippocampus, CMC DLPFC]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df['CMC DLPFC'] == 1) & (df['Caudate'] == 1) & (df['DLPFC'] == 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
