{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential Expression with limma-voom pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({library(SummarizedExperiment)\n",
    "                  library(variancePartition)\n",
    "                  library(BiocParallel)\n",
    "                  library(data.table)\n",
    "                  library(disk.frame)\n",
    "                  library(tidyverse)\n",
    "                  library(memoise)\n",
    "                  library(limma)\n",
    "                  library(edgeR)\n",
    "                  library(glue)\n",
    "                  library(sva)})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simple functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_parallel <- function(feature){\n",
    "    parallel_param = list(\"genes\"=32, \"transcripts\"=16, \n",
    "                          \"exons\"=10, \"junctions\"=12)\n",
    "    param = SnowParam(parallel_param[[feature]], \n",
    "                      \"SOCK\", progressbar=FALSE)\n",
    "    register(param)\n",
    "}\n",
    "\n",
    "\n",
    "load_counts <- function(region, feature){\n",
    "    if(region == 'caudate'){\n",
    "        counts_lt = list(\"genes\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                        \"caudate_brainseq_phase3_hg38_rseGene_merged_n464.rda\"), \n",
    "                         \"transcripts\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                              \"caudate_brainseq_phase3_hg38_rseTx_merged_n464.rda\"),\n",
    "                         \"exons\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                        \"caudate_brainseq_phase3_hg38_rseExon_merged_n464.rda\"),\n",
    "                         \"junctions\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                            \"caudate_brainseq_phase3_hg38_rseJxn_merged_n464.rda\"))\n",
    "        tx_file = \"/ceph/projects/v4_phase3_paper/inputs/counts/_m/transcripts_counts/caudate_counts.txt\"\n",
    "    } else if(region == 'dlpfc'){\n",
    "        counts_lt = list(\"genes\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                        \"dlpfc_ribozero_brainseq_phase2_hg38_rseGene_merged_n453.rda\"), \n",
    "                         \"transcripts\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                              \"dlpfc_ribozero_brainseq_phase2_hg38_rseTx_merged_n453.rda\"),\n",
    "                         \"exons\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                        \"dlpfc_ribozero_brainseq_phase2_hg38_rseExon_merged_n453.rda\"),\n",
    "                         \"junctions\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                            \"dlpfc_ribozero_brainseq_phase2_hg38_rseJxn_merged_n453.rda\"))\n",
    "        tx_file = \"/ceph/projects/v4_phase3_paper/inputs/counts/_m/transcripts_counts/dlpfc_counts.txt\"\n",
    "    } else {\n",
    "        counts_lt = list(\"genes\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                        \"hippo_brainseq_phase2_hg38_rseGene_merged_n447.rda\"), \n",
    "                         \"transcripts\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                              \"hippo_brainseq_phase2_hg38_rseTx_merged_n447.rda\"),\n",
    "                         \"exons\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                        \"hippo_brainseq_phase2_hg38_rseExon_merged_n447.rda\"),\n",
    "                         \"junctions\"=paste0(\"/ceph/projects/v4_phase3_paper/inputs/counts/_m/\",\n",
    "                                            \"hippo_brainseq_phase2_hg38_rseJxn_merged_n447.rda\"))\n",
    "        tx_file = \"/ceph/projects/v4_phase3_paper/inputs/counts/_m/transcripts_counts/hippo_counts.txt\"\n",
    "    }\n",
    "    return(list(\"count_file\"=counts_lt[[feature]], \"tx_file\"=tx_file))\n",
    "}\n",
    "\n",
    "\n",
    "get_model <- function(){\n",
    "    covs = c('Dx', 'Age', 'rRNA_rate', 'totalAssignedGene', 'RIN', \n",
    "             'ERCCsumLogErr', 'overallMapRate', 'mitoRate', 'snpPC1', \n",
    "             'snpPC2', 'snpPC3', 'Sex', 'Region', 'BrNum')\n",
    "    adjust_covs = covs[1:11]\n",
    "    random_effect = covs[14]\n",
    "    formula <- glue(\"~ Sex * Region + \", \n",
    "                    glue_collapse(adjust_covs, sep = \" + \")) %>% \n",
    "        glue(\" + (1|{random_effect})\")\n",
    "    return(formula)\n",
    "}\n",
    "\n",
    "\n",
    "get_null <- function(){\n",
    "    covs = c('Dx', 'Age', 'rRNA_rate', 'totalAssignedGene', 'RIN', \n",
    "             'ERCCsumLogErr', 'overallMapRate', 'mitoRate', 'snpPC1', \n",
    "             'snpPC2', 'snpPC3', 'Sex', 'Region', 'BrNum')\n",
    "    adjust_covs = covs[1:11]\n",
    "    random_effect = covs[14]\n",
    "    null_form <- glue(\"~ \", glue_collapse(adjust_covs[2:11], sep = \" + \")) %>% \n",
    "      glue(\" + (1|{random_effect}) + (1|Dx)\")\n",
    "    return(null_form)\n",
    "}\n",
    "\n",
    "\n",
    "save_volcanoPlot <- function(top, label, feature){\n",
    "    pdf(file=paste0(feature, \"/volcanoPlot_\", label, \".pdf\"), 8, 6)\n",
    "    with(top, plot(logFC, -log10(P.Value), pch=20, cex=0.6))\n",
    "    with(subset(top, adj.P.Val<=0.05), points(logFC, -log10(P.Value),\n",
    "                                              pch=20, col='red', cex=0.6))\n",
    "    with(subset(top, abs(logFC)>0.50), points(logFC, -log10(P.Value),\n",
    "                                              pch=20, col='orange', cex=0.6))\n",
    "    with(subset(top, adj.P.Val<=0.05 & abs(logFC)>0.50),\n",
    "         points(logFC, -log10(P.Value), pch=20, col='green', cex=0.6))\n",
    "    dev.off()\n",
    "}\n",
    "\n",
    "\n",
    "save_MAplot <- function(top, label, feature){\n",
    "    pdf(file=paste0(feature, \"/MAplot_\", label, \".pdf\"), 8, 6)\n",
    "    with(top, plot(AveExpr, logFC, pch=20, cex=0.5))\n",
    "    with(subset(top, adj.P.Val<0.05),\n",
    "         points(AveExpr, logFC, col=\"red\", pch=20, cex=0.5))\n",
    "    dev.off()\n",
    "}\n",
    "\n",
    "\n",
    "extract_de <- function(contrast, label, fit, feature){\n",
    "    top <- topTable(fit, coef=contrast, number=Inf, sort.by=\"P\")\n",
    "    top <- top[order(top$P.Value), ]\n",
    "    top.fdr <- top %>% filter(adj.P.Val<=0.05)\n",
    "    print(paste(\"Comparison for:\", label))\n",
    "    print(paste('There are:', dim(top.fdr)[1], 'DE features!'))\n",
    "    fwrite(top, \n",
    "           file=paste0(feature, \"/diffExpr_\", label, \"_full.txt\"), \n",
    "           sep='\\t', row.names=TRUE)\n",
    "    fwrite(top.fdr, \n",
    "           file=paste0(feature, \"/diffExpr_\", label, \"_FDR05.txt\"), \n",
    "           sep='\\t', row.names=TRUE)\n",
    "    save_volcanoPlot(top, label, feature)\n",
    "    save_MAplot(top, label, feature)\n",
    "}\n",
    "\n",
    "\n",
    "get_juncs <- function(region){\n",
    "    lt = load_counts(region, \"junctions\")\n",
    "    load(lt[['count_file']])\n",
    "    rse_df = rse_jxn\n",
    "    genes = rowData(rse_df) %>% as.data.frame\n",
    "    return(rownames(genes))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cached functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_mds <- function(){\n",
    "    mds_file = \"/ceph/projects/v4_phase3_paper/inputs/genotypes/mds/_m/LIBD_Brain_TopMed.mds\"\n",
    "    mds = data.table::fread(mds_file) %>%\n",
    "        rename_at(.vars = vars(starts_with(\"C\")),\n",
    "                  function(x){sub(\"C\", \"snpPC\", x)}) %>%\n",
    "        mutate_if(is.character, as.factor)\n",
    "    return(mds)\n",
    "}\n",
    "\n",
    "memMDS <- memoise::memoise(get_mds)\n",
    "\n",
    "get_overlapping_juncs <- function(){\n",
    "    common_genes = intersect(intersect(get_juncs(\"caudate\"), get_juncs(\"dlpfc\")), get_juncs(\"hippocampus\"))\n",
    "    return(common_genes)                \n",
    "}\n",
    "\n",
    "memJuncs <- memoise(get_overlapping_juncs)\n",
    "\n",
    "prep_data <- function(feature){\n",
    "    datalist1 = list()\n",
    "    datalist2 = list()\n",
    "    for(region in c('dlpfc', 'caudate', 'hippocampus')){\n",
    "        print(region)\n",
    "        flush.console()\n",
    "        lt = load_counts(region, feature)\n",
    "        load(lt[['count_file']])\n",
    "        tx_file = lt[['tx_file']]\n",
    "        if (exists(\"rse_gene\")){\n",
    "            rse_df = rse_gene\n",
    "        } else if (exists(\"rse_tx\")){\n",
    "            if(region == 'caudate' && exists(\"geneid\")){\n",
    "                counts <- fread(tx_file) %>% filter(gencodeTx %in% geneid) %>%\n",
    "                    column_to_rownames(\"gencodeTx\") %>% \n",
    "                    select(-c(txLength, gencodeID, Symbol, gene_type)) %>% \n",
    "                    select(colnames(rse_tx)) %>% as.matrix \n",
    "                rse_df <- SummarizedExperiment(assays=SimpleList(counts=counts),\n",
    "                                               rowData=annot, colData=colData(rse_tx)) %>%\n",
    "                    as(\"RangedSummarizedExperiment\")\n",
    "            } else {\n",
    "                counts <- fread(tx_file) %>% column_to_rownames(\"transcript_id\") %>%\n",
    "                    select(colnames(rse_tx)) %>% as.matrix\n",
    "                annot <- fread(tx_file) %>% column_to_rownames(\"transcript_id\") %>%\n",
    "                    select(-starts_with(\"R\"))\n",
    "                rse_df <- SummarizedExperiment(assays=SimpleList(counts=counts),\n",
    "                                               rowData=annot, \n",
    "                                               colData=colData(rse_tx)) %>%\n",
    "                    as(\"RangedSummarizedExperiment\")\n",
    "                geneid = rownames(rowData(rse_df))\n",
    "            }\n",
    "        } else if (exists(\"rse_exon\")){\n",
    "            rse_df = rse_exon\n",
    "        } else {\n",
    "            rse_df = rse_jxn\n",
    "        }\n",
    "        keepIndex = which((rse_df$Dx %in% c(\"Control\", \"Schizo\")) & \n",
    "                          rse_df$Age > 17 & rse_df$Race %in% c(\"AA\", \"CAUC\"))\n",
    "        rse_df = rse_df[, keepIndex]\n",
    "        rse_df$Sex <- factor(rse_df$Sex)\n",
    "        rse_df <- jaffelab::merge_rse_metrics(rse_df)\n",
    "        colData(rse_df)$RIN = sapply(colData(rse_df)$RIN,\"[\",1)\n",
    "        colData(rse_df)$ERCCsumLogErr = sapply(colData(rse_df)$ERCCsumLogErr,\"[\",1)\n",
    "        rownames(colData(rse_df)) <- sapply(strsplit(rownames(colData(rse_df)), \"_\"), \"[\", 1)\n",
    "        pheno = colData(rse_df) %>% as.data.frame %>% \n",
    "            inner_join(memMDS(), by=c(\"BrNum\"=\"FID\")) %>% \n",
    "            distinct(RNum, .keep_all = TRUE) \n",
    "        if(feature == \"junctions\"){\n",
    "            datalist1[[region]] <- assays(rse_df)$counts[, pheno$RNum] %>% as.data.frame %>% \n",
    "                rownames_to_column(\"V1\") %>% filter(V1 %in% memJuncs()) %>% \n",
    "                column_to_rownames(\"V1\") %>% t %>% as.data.frame\n",
    "        }else{\n",
    "            datalist1[[region]] <- assays(rse_df)$counts[, pheno$RNum] %>% t %>% \n",
    "                as.data.frame \n",
    "        }\n",
    "        datalist2[[region]] <- pheno[, c('Sex', 'Dx', 'Age', 'rRNA_rate', 'totalAssignedGene', \n",
    "                                         'RIN','ERCCsumLogErr', 'overallMapRate', 'mitoRate', \n",
    "                                         'snpPC1', 'snpPC2', 'snpPC3', 'Region', 'BrNum', 'RNum')]\n",
    "    }\n",
    "    samples <- bind_rows(datalist2) %>% as.data.frame %>% \n",
    "        mutate_if(is.numeric, rescale)\n",
    "    subjects = samples$BrNum\n",
    "    if(feature == \"junctions\"){\n",
    "        genes = rowData(rse_df) %>% as.data.frame %>% rownames_to_column(\"V1\") %>%\n",
    "            filter(V1 %in% memJuncs()) %>% column_to_rownames(\"V1\") \n",
    "    } else {\n",
    "        genes = rowData(rse_df) %>% as.data.frame\n",
    "    }\n",
    "    counts <- bind_rows(datalist1) %>% t %>% as.data.frame\n",
    "    colnames(counts) <- samples$RNum\n",
    "    if(samples %>% has_rownames){\n",
    "        samples = samples %>% select(-RNum)\n",
    "    } else {\n",
    "        samples = samples %>% column_to_rownames(\"RNum\")\n",
    "    }\n",
    "    # Generate DGE list\n",
    "    x <- DGEList(counts=counts[, row.names(samples)], \n",
    "                 genes=genes, samples=samples)\n",
    "    # Filter by expression\n",
    "    design0 <- model.matrix(~Sex*Region, data=x$samples)\n",
    "    keep.x <- filterByExpr(x, design=design0)\n",
    "    x <- x[keep.x, , keep.lib.sizes=FALSE]\n",
    "    print(paste('There are:', sum(keep.x), 'features left!', sep=' '))\n",
    "    # Normalize library size\n",
    "    x <- calcNormFactors(x, method=\"TMM\")\n",
    "    return(x)\n",
    "}\n",
    "\n",
    "memo_prepData <- memoise(prep_data)\n",
    "\n",
    "plot_corrMatrix <- function(feature){\n",
    "    x <- memo_prepData(feature)\n",
    "    covs = c('Dx', 'Age', 'rRNA_rate', 'totalAssignedGene', 'RIN', \n",
    "             'ERCCsumLogErr', 'overallMapRate', 'mitoRate', 'snpPC1', \n",
    "             'snpPC2', 'snpPC3', 'Sex', 'Region', 'BrNum')\n",
    "    form <- paste('~', glue_collapse(covs, sep=\" + \"))\n",
    "    # Compute Canonical Correlation Analysis (CCA)\n",
    "    # between all pairs of variables# returns absolute correlation value\n",
    "    C = canCorPairs(form, x$samples)\n",
    "    # Plot correlation matrix\n",
    "    plotCorrMatrix( C )\n",
    "}\n",
    "\n",
    "memo_corrMatrix <- memoise(plot_corrMatrix)\n",
    "\n",
    "get_voom <- function(feature){\n",
    "    memo_corrMatrix(feature)\n",
    "    x <- memo_prepData(feature)\n",
    "    vobjDream = voomWithDreamWeights(x, get_model(), \n",
    "                                     x$samples, plot=TRUE)\n",
    "    return(vobjDream)\n",
    "}\n",
    "\n",
    "memo_voom <- memoise(get_voom)\n",
    "\n",
    "fit_dream <- function(feature){\n",
    "    x <- memo_prepData(feature)\n",
    "    vobj <- memo_voom(feature)\n",
    "    # define and then cbind contrasts\n",
    "    L1 = getContrast(vobj, get_model(), x$samples, \n",
    "                     c(\"RegionDLPFC\", \"RegionHIPPO\"))\n",
    "    L2 = getContrast(vobj, get_model(), x$samples, \n",
    "                     c(\"SexM:RegionDLPFC\", \"SexM:RegionHIPPO\"))\n",
    "    # combine contrasts\n",
    "    L = cbind(L1, L2)     \n",
    "    # Visualize contrasts\n",
    "    plotContrasts(L) \n",
    "    fit = dream(vobj, get_model(), x$samples, L=L)\n",
    "    return(fit)\n",
    "}\n",
    "\n",
    "memo_dream <- memoise(fit_dream)\n",
    "\n",
    "cal_res <- function(feature){\n",
    "    ### Calculate residuals\n",
    "    vobj <- memo_voom(feature)\n",
    "    residList <- fitVarPartModel(vobj, get_null(), vobj$targets, fxn=residuals)\n",
    "    do.call(rbind, residList) %>% as.data.frame %>% \n",
    "        rownames_to_column %>% rename(Geneid=rowname) %>% \n",
    "        fwrite(file=paste0(feature, '/residualized_expression.tsv'), sep='\\t')\n",
    "}\n",
    "\n",
    "memo_res <- memoise(cal_res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for(feature in c('genes', 'transcripts', 'junctions', 'exons')){\n",
    "    flush.console()\n",
    "    dir.create(feature)\n",
    "    # Set parallel parameters\n",
    "    print(\"Set parallel!\")\n",
    "    set_parallel(feature)\n",
    "    # Prepare data\n",
    "    print(\"Prepare data!\")\n",
    "    x <- memo_prepData(feature)\n",
    "    save(x, file=paste0(feature, \"/normData.RData\"))\n",
    "    # Preform voom\n",
    "    print(\"Preform voom!\")\n",
    "    v <- memo_voom(feature)\n",
    "    save(v, file=paste0(feature,'/voom_dream.RData'))\n",
    "    # Fit model and apply eBayes\n",
    "    print(\"Fit model!\")\n",
    "    fit = memo_dream(feature)\n",
    "    # Save differential expression\n",
    "    extract_de(\"SexM\", \"sex\", fit)\n",
    "    extract_de(\"SexM:RegionDLPFC\", \"CvD_sex\", fit)\n",
    "    extract_de(\"SexM:RegionHIPPO\", \"CvH_sex\", fit)\n",
    "    extract_de(\"L2\", \"DvH_sex\", fit)\n",
    "    # Calculate residuals\n",
    "    print(\"Generated residualized expression!\")\n",
    "    memo_res(feature)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Repreducibility Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.time()\n",
    "proc.time()\n",
    "options(width = 120)\n",
    "sessioninfo::session_info()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
