{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract male bias genes on the X chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_deg(tissue):\n",
    "    fn = \"../../../../differential_expression/%s/\" % tissue +\\\n",
    "    \"metrics_summary/_m/chrom_annotation_genes.txt\"\n",
    "    return pd.read_csv(fn, sep='\\t').loc[:, [\"gene_id\", \"seqname\", \"Symbol\", \n",
    "                                                         \"t\", \"adj.P.Val\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1058, 7)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame()\n",
    "for tissue in [\"caudate\", \"dlpfc\", \"hippocampus\"]:\n",
    "    dt = get_deg(tissue)\n",
    "    dt[\"Tissue\"] = tissue\n",
    "    df = pd.concat([df, dt], axis=0)\n",
    "df[\"ensemblID\"] = df.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start position</th>\n",
       "      <th>End position</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Combined XCI status</th>\n",
       "      <th>ensemblID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.8</td>\n",
       "      <td>X</td>\n",
       "      <td>192989</td>\n",
       "      <td>220023</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "      <td>ENSG00000182378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>ENSG00000178605.8</td>\n",
       "      <td>X</td>\n",
       "      <td>220025</td>\n",
       "      <td>230886</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "      <td>ENSG00000178605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene name            Gene ID Chr  Start position  End position  \\\n",
       "0    PLCXD1  ENSG00000182378.8   X          192989        220023   \n",
       "1    GTPBP6  ENSG00000178605.8   X          220025        230886   \n",
       "\n",
       "  Transcript type Combined XCI status        ensemblID  \n",
       "0  protein_coding              escape  ENSG00000182378  \n",
       "1  protein_coding              escape  ENSG00000178605  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xci = pd.read_csv(\"../../_h/xci_status_hg19.txt\", sep='\\t')\n",
    "xci[\"ensemblID\"] = xci[\"Gene ID\"].str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "xci.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Combined XCI status\n",
       "escape       99\n",
       "inactive    431\n",
       "variable    101\n",
       "dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xci.groupby(\"Combined XCI status\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>t</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start position</th>\n",
       "      <th>End position</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Combined XCI status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ENSG00000182378.13_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>6.988151</td>\n",
       "      <td>4.239261e-09</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000182378</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.8</td>\n",
       "      <td>X</td>\n",
       "      <td>192989</td>\n",
       "      <td>220023</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ENSG00000182378.13_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>6.107645</td>\n",
       "      <td>9.826634e-07</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000182378</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.8</td>\n",
       "      <td>X</td>\n",
       "      <td>192989</td>\n",
       "      <td>220023</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000182378.13_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>4.772588</td>\n",
       "      <td>7.966110e-04</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>ENSG00000182378</td>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.8</td>\n",
       "      <td>X</td>\n",
       "      <td>192989</td>\n",
       "      <td>220023</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>ENSG00000002586.18_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>CD99</td>\n",
       "      <td>4.264277</td>\n",
       "      <td>3.645847e-03</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>ENSG00000002586.13</td>\n",
       "      <td>X</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2659350</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>ENSG00000002586.18_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>CD99</td>\n",
       "      <td>3.901722</td>\n",
       "      <td>1.477514e-02</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>ENSG00000002586.13</td>\n",
       "      <td>X</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2659350</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>ENSG00000002586.18_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>CD99</td>\n",
       "      <td>4.073952</td>\n",
       "      <td>1.481474e-02</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>ENSG00000002586</td>\n",
       "      <td>CD99</td>\n",
       "      <td>ENSG00000002586.13</td>\n",
       "      <td>X</td>\n",
       "      <td>2609220</td>\n",
       "      <td>2659350</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ENSG00000169093.15_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>ASMTL</td>\n",
       "      <td>3.831986</td>\n",
       "      <td>1.701108e-02</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000169093</td>\n",
       "      <td>ASMTL</td>\n",
       "      <td>ENSG00000169093.10</td>\n",
       "      <td>X</td>\n",
       "      <td>1522032</td>\n",
       "      <td>1572655</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ENSG00000178605.13_PAR_Y</td>\n",
       "      <td>chrY</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>3.599197</td>\n",
       "      <td>2.847852e-02</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000178605</td>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>ENSG00000178605.8</td>\n",
       "      <td>X</td>\n",
       "      <td>220025</td>\n",
       "      <td>230886</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>escape</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id seqname  Symbol         t     adj.P.Val  \\\n",
       "55  ENSG00000182378.13_PAR_Y    chrY  PLCXD1  6.988151  4.239261e-09   \n",
       "56  ENSG00000182378.13_PAR_Y    chrY  PLCXD1  6.107645  9.826634e-07   \n",
       "57  ENSG00000182378.13_PAR_Y    chrY  PLCXD1  4.772588  7.966110e-04   \n",
       "72  ENSG00000002586.18_PAR_Y    chrY    CD99  4.264277  3.645847e-03   \n",
       "73  ENSG00000002586.18_PAR_Y    chrY    CD99  3.901722  1.477514e-02   \n",
       "74  ENSG00000002586.18_PAR_Y    chrY    CD99  4.073952  1.481474e-02   \n",
       "80  ENSG00000169093.15_PAR_Y    chrY   ASMTL  3.831986  1.701108e-02   \n",
       "83  ENSG00000178605.13_PAR_Y    chrY  GTPBP6  3.599197  2.847852e-02   \n",
       "\n",
       "         Tissue        ensemblID Gene name             Gene ID Chr  \\\n",
       "55      caudate  ENSG00000182378    PLCXD1   ENSG00000182378.8   X   \n",
       "56        dlpfc  ENSG00000182378    PLCXD1   ENSG00000182378.8   X   \n",
       "57  hippocampus  ENSG00000182378    PLCXD1   ENSG00000182378.8   X   \n",
       "72      caudate  ENSG00000002586      CD99  ENSG00000002586.13   X   \n",
       "73        dlpfc  ENSG00000002586      CD99  ENSG00000002586.13   X   \n",
       "74  hippocampus  ENSG00000002586      CD99  ENSG00000002586.13   X   \n",
       "80        dlpfc  ENSG00000169093     ASMTL  ENSG00000169093.10   X   \n",
       "83        dlpfc  ENSG00000178605    GTPBP6   ENSG00000178605.8   X   \n",
       "\n",
       "    Start position  End position Transcript type Combined XCI status  \n",
       "55          192989        220023  protein_coding              escape  \n",
       "56          192989        220023  protein_coding              escape  \n",
       "57          192989        220023  protein_coding              escape  \n",
       "72         2609220       2659350  protein_coding              escape  \n",
       "73         2609220       2659350  protein_coding              escape  \n",
       "74         2609220       2659350  protein_coding              escape  \n",
       "80         1522032       1572655  protein_coding              escape  \n",
       "83          220025        230886  protein_coding              escape  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt = df.merge(xci[(xci[\"Combined XCI status\"] == \"escape\")], on=\"ensemblID\")\n",
    "tt[(tt['t'] > 0)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Escaped genes are also located on the PAR regions of the Y chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "xlinked = df[(df['seqname'] == 'chrX')].copy()\n",
    "xlinked_male = xlinked[(xlinked[\"t\"] > 0)].copy()\n",
    "xlinked_female = xlinked[(xlinked[\"t\"] < 0)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "caudate        45\n",
       "dlpfc          60\n",
       "hippocampus    31\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked.groupby(\"Tissue\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "caudate         3\n",
       "dlpfc          18\n",
       "hippocampus     1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_male.groupby(\"Tissue\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tissue\n",
       "caudate        42\n",
       "dlpfc          42\n",
       "hippocampus    30\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_female.groupby(\"Tissue\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>t</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>ensemblID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>ENSG00000213468.4</td>\n",
       "      <td>chrX</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4.689886</td>\n",
       "      <td>0.000705</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000213468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>ENSG00000186675.6</td>\n",
       "      <td>chrX</td>\n",
       "      <td>MAGEE2</td>\n",
       "      <td>4.613375</td>\n",
       "      <td>0.000956</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000186675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>ENSG00000102001.12</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CACNA1F</td>\n",
       "      <td>4.608831</td>\n",
       "      <td>0.000963</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000102001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>ENSG00000172465.13</td>\n",
       "      <td>chrX</td>\n",
       "      <td>TCEAL1</td>\n",
       "      <td>4.402338</td>\n",
       "      <td>0.003347</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000172465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>ENSG00000236064.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.364580</td>\n",
       "      <td>0.003866</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000236064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>ENSG00000277883.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NLRP3P1</td>\n",
       "      <td>4.155232</td>\n",
       "      <td>0.007473</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000277883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>ENSG00000184515.10</td>\n",
       "      <td>chrX</td>\n",
       "      <td>BEX5</td>\n",
       "      <td>4.126387</td>\n",
       "      <td>0.008105</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000184515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>ENSG00000204071.10</td>\n",
       "      <td>chrX</td>\n",
       "      <td>TCEAL6</td>\n",
       "      <td>3.836015</td>\n",
       "      <td>0.017011</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000204071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>ENSG00000147155.10</td>\n",
       "      <td>chrX</td>\n",
       "      <td>EBP</td>\n",
       "      <td>3.751150</td>\n",
       "      <td>0.020360</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000147155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>ENSG00000232119.7</td>\n",
       "      <td>chrX</td>\n",
       "      <td>MCTS1</td>\n",
       "      <td>3.632172</td>\n",
       "      <td>0.026693</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000232119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>ENSG00000198932.12</td>\n",
       "      <td>chrX</td>\n",
       "      <td>GPRASP1</td>\n",
       "      <td>3.611593</td>\n",
       "      <td>0.027517</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000198932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>ENSG00000186675.6</td>\n",
       "      <td>chrX</td>\n",
       "      <td>MAGEE2</td>\n",
       "      <td>3.555431</td>\n",
       "      <td>0.031388</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000186675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>ENSG00000102054.17</td>\n",
       "      <td>chrX</td>\n",
       "      <td>RBBP7</td>\n",
       "      <td>3.530376</td>\n",
       "      <td>0.032470</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000102054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>ENSG00000278530.4</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CHMP1B2P</td>\n",
       "      <td>3.485690</td>\n",
       "      <td>0.035151</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000278530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>ENSG00000133169.5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>BEX1</td>\n",
       "      <td>3.454265</td>\n",
       "      <td>0.037068</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000133169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>ENSG00000184905.8</td>\n",
       "      <td>chrX</td>\n",
       "      <td>TCEAL2</td>\n",
       "      <td>3.428241</td>\n",
       "      <td>0.039058</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000184905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENSG00000184867.13</td>\n",
       "      <td>chrX</td>\n",
       "      <td>ARMCX2</td>\n",
       "      <td>3.417176</td>\n",
       "      <td>0.039832</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000184867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>ENSG00000102401.19</td>\n",
       "      <td>chrX</td>\n",
       "      <td>ARMCX3</td>\n",
       "      <td>3.328593</td>\n",
       "      <td>0.045811</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000102401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>ENSG00000133134.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>BEX2</td>\n",
       "      <td>3.325992</td>\n",
       "      <td>0.045811</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000133134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>ENSG00000224204.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>PHEX-AS1</td>\n",
       "      <td>3.311985</td>\n",
       "      <td>0.046573</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000224204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>ENSG00000178947.8</td>\n",
       "      <td>chrX</td>\n",
       "      <td>SMIM10L2A</td>\n",
       "      <td>3.297549</td>\n",
       "      <td>0.047002</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000178947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>ENSG00000147124.12</td>\n",
       "      <td>chrX</td>\n",
       "      <td>ZNF41</td>\n",
       "      <td>4.105295</td>\n",
       "      <td>0.013322</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>ENSG00000147124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id seqname     Symbol         t  adj.P.Val       Tissue  \\\n",
       "126   ENSG00000213468.4    chrX      FIRRE  4.689886   0.000705      caudate   \n",
       "132   ENSG00000186675.6    chrX     MAGEE2  4.613375   0.000956      caudate   \n",
       "133  ENSG00000102001.12    chrX    CACNA1F  4.608831   0.000963      caudate   \n",
       "97   ENSG00000172465.13    chrX     TCEAL1  4.402338   0.003347        dlpfc   \n",
       "99    ENSG00000236064.1    chrX        NaN  4.364580   0.003866        dlpfc   \n",
       "125   ENSG00000277883.1    chrX    NLRP3P1  4.155232   0.007473        dlpfc   \n",
       "130  ENSG00000184515.10    chrX       BEX5  4.126387   0.008105        dlpfc   \n",
       "199  ENSG00000204071.10    chrX     TCEAL6  3.836015   0.017011        dlpfc   \n",
       "232  ENSG00000147155.10    chrX        EBP  3.751150   0.020360        dlpfc   \n",
       "277   ENSG00000232119.7    chrX      MCTS1  3.632172   0.026693        dlpfc   \n",
       "291  ENSG00000198932.12    chrX    GPRASP1  3.611593   0.027517        dlpfc   \n",
       "314   ENSG00000186675.6    chrX     MAGEE2  3.555431   0.031388        dlpfc   \n",
       "330  ENSG00000102054.17    chrX      RBBP7  3.530376   0.032470        dlpfc   \n",
       "361   ENSG00000278530.4    chrX   CHMP1B2P  3.485690   0.035151        dlpfc   \n",
       "384   ENSG00000133169.5    chrX       BEX1  3.454265   0.037068        dlpfc   \n",
       "399   ENSG00000184905.8    chrX     TCEAL2  3.428241   0.039058        dlpfc   \n",
       "408  ENSG00000184867.13    chrX     ARMCX2  3.417176   0.039832        dlpfc   \n",
       "483  ENSG00000102401.19    chrX     ARMCX3  3.328593   0.045811        dlpfc   \n",
       "486  ENSG00000133134.11    chrX       BEX2  3.325992   0.045811        dlpfc   \n",
       "505   ENSG00000224204.1    chrX   PHEX-AS1  3.311985   0.046573        dlpfc   \n",
       "528   ENSG00000178947.8    chrX  SMIM10L2A  3.297549   0.047002        dlpfc   \n",
       "85   ENSG00000147124.12    chrX      ZNF41  4.105295   0.013322  hippocampus   \n",
       "\n",
       "           ensemblID  \n",
       "126  ENSG00000213468  \n",
       "132  ENSG00000186675  \n",
       "133  ENSG00000102001  \n",
       "97   ENSG00000172465  \n",
       "99   ENSG00000236064  \n",
       "125  ENSG00000277883  \n",
       "130  ENSG00000184515  \n",
       "199  ENSG00000204071  \n",
       "232  ENSG00000147155  \n",
       "277  ENSG00000232119  \n",
       "291  ENSG00000198932  \n",
       "314  ENSG00000186675  \n",
       "330  ENSG00000102054  \n",
       "361  ENSG00000278530  \n",
       "384  ENSG00000133169  \n",
       "399  ENSG00000184905  \n",
       "408  ENSG00000184867  \n",
       "483  ENSG00000102401  \n",
       "486  ENSG00000133134  \n",
       "505  ENSG00000224204  \n",
       "528  ENSG00000178947  \n",
       "85   ENSG00000147124  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xlinked_male"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>t</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>Combined XCI status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000213468.4</td>\n",
       "      <td>chrX</td>\n",
       "      <td>FIRRE</td>\n",
       "      <td>4.689886</td>\n",
       "      <td>0.000705</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000213468</td>\n",
       "      <td>variable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000102001.12</td>\n",
       "      <td>chrX</td>\n",
       "      <td>CACNA1F</td>\n",
       "      <td>4.608831</td>\n",
       "      <td>0.000963</td>\n",
       "      <td>caudate</td>\n",
       "      <td>ENSG00000102001</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000172465.13</td>\n",
       "      <td>chrX</td>\n",
       "      <td>TCEAL1</td>\n",
       "      <td>4.402338</td>\n",
       "      <td>0.003347</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000172465</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000236064.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.364580</td>\n",
       "      <td>0.003866</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000236064</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000232119.7</td>\n",
       "      <td>chrX</td>\n",
       "      <td>MCTS1</td>\n",
       "      <td>3.632172</td>\n",
       "      <td>0.026693</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000232119</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000198932.12</td>\n",
       "      <td>chrX</td>\n",
       "      <td>GPRASP1</td>\n",
       "      <td>3.611593</td>\n",
       "      <td>0.027517</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000198932</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000102054.17</td>\n",
       "      <td>chrX</td>\n",
       "      <td>RBBP7</td>\n",
       "      <td>3.530376</td>\n",
       "      <td>0.032470</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000102054</td>\n",
       "      <td>variable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000184905.8</td>\n",
       "      <td>chrX</td>\n",
       "      <td>TCEAL2</td>\n",
       "      <td>3.428241</td>\n",
       "      <td>0.039058</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000184905</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000184867.13</td>\n",
       "      <td>chrX</td>\n",
       "      <td>ARMCX2</td>\n",
       "      <td>3.417176</td>\n",
       "      <td>0.039832</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000184867</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000102401.19</td>\n",
       "      <td>chrX</td>\n",
       "      <td>ARMCX3</td>\n",
       "      <td>3.328593</td>\n",
       "      <td>0.045811</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000102401</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000133134.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>BEX2</td>\n",
       "      <td>3.325992</td>\n",
       "      <td>0.045811</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000133134</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000224204.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>PHEX-AS1</td>\n",
       "      <td>3.311985</td>\n",
       "      <td>0.046573</td>\n",
       "      <td>dlpfc</td>\n",
       "      <td>ENSG00000224204</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000147124.12</td>\n",
       "      <td>chrX</td>\n",
       "      <td>ZNF41</td>\n",
       "      <td>4.105295</td>\n",
       "      <td>0.013322</td>\n",
       "      <td>hippocampus</td>\n",
       "      <td>ENSG00000147124</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id seqname    Symbol         t  adj.P.Val       Tissue  \\\n",
       "0    ENSG00000213468.4    chrX     FIRRE  4.689886   0.000705      caudate   \n",
       "1   ENSG00000102001.12    chrX   CACNA1F  4.608831   0.000963      caudate   \n",
       "2   ENSG00000172465.13    chrX    TCEAL1  4.402338   0.003347        dlpfc   \n",
       "3    ENSG00000236064.1    chrX       NaN  4.364580   0.003866        dlpfc   \n",
       "4    ENSG00000232119.7    chrX     MCTS1  3.632172   0.026693        dlpfc   \n",
       "5   ENSG00000198932.12    chrX   GPRASP1  3.611593   0.027517        dlpfc   \n",
       "6   ENSG00000102054.17    chrX     RBBP7  3.530376   0.032470        dlpfc   \n",
       "7    ENSG00000184905.8    chrX    TCEAL2  3.428241   0.039058        dlpfc   \n",
       "8   ENSG00000184867.13    chrX    ARMCX2  3.417176   0.039832        dlpfc   \n",
       "9   ENSG00000102401.19    chrX    ARMCX3  3.328593   0.045811        dlpfc   \n",
       "10  ENSG00000133134.11    chrX      BEX2  3.325992   0.045811        dlpfc   \n",
       "11   ENSG00000224204.1    chrX  PHEX-AS1  3.311985   0.046573        dlpfc   \n",
       "12  ENSG00000147124.12    chrX     ZNF41  4.105295   0.013322  hippocampus   \n",
       "\n",
       "          ensemblID Combined XCI status  \n",
       "0   ENSG00000213468            variable  \n",
       "1   ENSG00000102001            inactive  \n",
       "2   ENSG00000172465            inactive  \n",
       "3   ENSG00000236064            inactive  \n",
       "4   ENSG00000232119            inactive  \n",
       "5   ENSG00000198932            inactive  \n",
       "6   ENSG00000102054            variable  \n",
       "7   ENSG00000184905            inactive  \n",
       "8   ENSG00000184867            inactive  \n",
       "9   ENSG00000102401            inactive  \n",
       "10  ENSG00000133134            inactive  \n",
       "11  ENSG00000224204            inactive  \n",
       "12  ENSG00000147124            inactive  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xci[\"ensemblID\"] = xci[\"Gene ID\"].str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "xlinked_male[\"ensemblID\"] = xlinked_male.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "xlinked_male.merge(xci[[\"ensemblID\", \"Combined XCI status\"]], on=\"ensemblID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
