{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting eQTLs, increase font sizes\n",
    "\n",
    "### Kynon Jade Benjamin and ApuÃ£ Paquola"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import functools\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from plotnine import *\n",
    "from pandas_plink import read_plink\n",
    "from warnings import filterwarnings\n",
    "from matplotlib.cbook import mplDeprecation\n",
    "\n",
    "filterwarnings(\"ignore\",category=mplDeprecation)\n",
    "filterwarnings('ignore', category=UserWarning, module='plotnine.*')\n",
    "filterwarnings('ignore', category=DeprecationWarning, module='plotnine.*')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue = \"hippocampus\"; feature = \"genes\"\n",
    "config = {\n",
    "    'biomart_file': '../_h/biomart.csv',\n",
    "    'residual_expression_file': \"../../../../prep_eqtl_analysis/%s/%s/covariates/\" % (tissue, feature)+\\\n",
    "    \"residualized_expression/_m/%s_residualized_expression.csv\" % feature,\n",
    "    'phenotype_file': '/ceph/projects/v4_phase3_paper/inputs/phenotypes/_m/merged_phenotypes.csv',\n",
    "    'plink_file_prefix': '/ceph/projects/v4_phase3_paper/inputs/genotypes/_m/LIBD_Brain_TopMed',\n",
    "    'eqtl_output_file': '../../../summary_table/_m/Brainseq_sex_interacting_4features_3regions.eFeatures.txt.gz',\n",
    "    'gwas_snp_file': '/ceph/projects/v4_phase3_paper/inputs/sz_gwas/pgc2_clozuk/map_phase3/_m/libd_hg38_pgc2sz_snps_p5e_minus8.tsv'\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def tissue_map(tissue):\n",
    "    return {\"caudate\": \"Caudate\", \"dlpfc\": \"DLPFC\",\n",
    "            \"hippocampus\": \"Hippocampus\"}[tissue]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def feature_map(feature):\n",
    "    return {\"genes\": \"Gene\", \"transcripts\": \"Transcript\",\n",
    "            \"exons\": \"Exon\", \"junctions\": \"Junction\"}[feature]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_biomart_df():\n",
    "    biomart = pd.read_csv(config['biomart_file'], index_col=0)\n",
    "    biomart['description'] = biomart['description'].str.replace('\\[Source.*$','', regex=True)\n",
    "    return biomart\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_residual_expression_df():\n",
    "    return pd.read_csv(config['residual_expression_file'], index_col=0).transpose()\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_pheno_df():\n",
    "    return pd.read_csv(config['phenotype_file']).set_index(\"BrNum\").loc[:, [\"RNum\", \"Sex\", \"Dx\"]]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_expression_and_pheno_df():\n",
    "    return pd.merge(get_pheno_df(), get_residual_expression_df(),\n",
    "                    left_index=True, right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gene_id_df():\n",
    "    return pd.DataFrame({'gene_id': get_residual_expression_df().columns,\n",
    "                         'ensembl_gene_id': get_residual_expression_df().columns.str.replace('\\..+$','', regex=True)})\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def gene_info_from_symbol(gene_symbol):\n",
    "    return get_biomart_df()[get_biomart_df()['external_gene_name']==gene_symbol]\\\n",
    "        .merge(get_gene_id_df(), on='ensembl_gene_id', how='left')\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def gene_id_from_symbol(gene_symbol):\n",
    "    df = gene_info_from_symbol(gene_symbol)\n",
    "    assert df.shape[0] == 1\n",
    "    return df[['gene_id']].iloc[0].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotype and eQTL functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def letter_snp(number, a0, a1):\n",
    "    '''\n",
    "    Example:\n",
    "    letter_snp(0, 'A', 'G') is 'AA'\n",
    "    letter_snp(1, 'A', 'G') is 'AG'\n",
    "    letter_snp(2, 'A', 'G') is 'GG'\n",
    "    '''\n",
    "    if np.isnan(number):\n",
    "        return np.nan\n",
    "    if len(a0)==1 and len(a1)==1:\n",
    "        sep = ''\n",
    "    else:\n",
    "        sep = ' '\n",
    "    return sep.join(sorted([a0]*int(number) + [a1]*(2-int(number))))\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_plink_tuple():\n",
    "    '''\n",
    "    Usage: (bim, fam, bed) = get_plink_tuple()\n",
    "    '''\n",
    "    return read_plink(config['plink_file_prefix'])\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_eFeature_df():\n",
    "    eqtl_df = pd.read_csv(config[\"eqtl_output_file\"], sep='\\t')\n",
    "    return eqtl_df[(eqtl_df[\"Type\"] == feature_map(feature)) & \n",
    "                   (eqtl_df[\"Tissue\"] == tissue_map(tissue))]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_snps():\n",
    "    return pd.read_csv(config['gwas_snp_file'], sep='\\t', index_col=0)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_risk_allele(snp_id):\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['OR'].iloc[0] > 1:\n",
    "        ra = gwas_snp['A1'].iloc[0]\n",
    "    else:\n",
    "        ra = gwas_snp['A2'].iloc[0]\n",
    "    return ra\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the same as in the plink files.\n",
    "    \n",
    "    Example: \n",
    "    get_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             0              GG    0\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             2              AA    2\\nAA\n",
    "    Br5182             2              AA    2\\nAA\n",
    "    '''\n",
    "    (bim, fam, bed) = get_plink_tuple()\n",
    "    brain_ids = list(set(get_expression_and_pheno_df().index).intersection(set(fam['fid'])))\n",
    "    snp_info = bim[bim['snp']==snp_id]\n",
    "    snp_pos = snp_info.iloc[0]['i']\n",
    "    fam_pos = list(fam.drop_duplicates(subset=\"fid\").set_index('fid').loc[brain_ids]['i'])\n",
    "    dfsnp = (pd.DataFrame(bed[[snp_pos]].compute()[:,fam_pos], \n",
    "                          columns=brain_ids, index=[snp_id + '_num'])\n",
    "             .transpose().dropna())\n",
    "    my_letter_snp = functools.partial(letter_snp, a0=snp_info.iloc[0]['a0'], a1=snp_info.iloc[0]['a1'])\n",
    "    dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']].astype('int')\n",
    "    dfsnp[snp_id + '_letter'] = dfsnp[snp_id + '_num'].apply(my_letter_snp)\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_ordered_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the number of risk alleles according to GWAS.\n",
    "    \n",
    "    Example: \n",
    "    get_gwas_ordered_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             2              GG    2\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             0              AA    0\\nAA\n",
    "    Br5182             0              AA    0\\nAA\n",
    "    '''\n",
    "    pgc = get_gwas_snps()\n",
    "    dfsnp = get_snp_df(snp_id).copy()\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['pgc2_a1_same_as_our_counted'].iloc[0]:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            pass\n",
    "        else:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "    else:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "        else:\n",
    "            pass\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func):\n",
    "    pheno_columns = list(get_pheno_df().columns)\n",
    "    expr_df = get_expression_and_pheno_df()[pheno_columns + [gene_id]]\n",
    "    snp_df =  snp_df_func(snp_id)\n",
    "    return expr_df.merge(snp_df, left_index=True, right_index=True)\n",
    "    \n",
    "\n",
    "def simple_snp_expression_plot_impl(snp_id, gene_id, snp_df_func):\n",
    "    df = get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func)\n",
    "    y0 = df[gene_id].quantile(.01) - 0.26\n",
    "    y1 = df[gene_id].quantile(.99) + 0.26\n",
    "    p = ggplot(df, aes(x=snp_id, y=gene_id, fill='Sex')) \\\n",
    "    + geom_boxplot(alpha=0.4, outlier_alpha=0) \\\n",
    "    + geom_jitter(position=position_jitterdodge(jitter_width=0.25),\n",
    "                  stroke=0, alpha=0.6) \\\n",
    "    + ylim(y0, y1) \\\n",
    "    + theme_bw(base_size=15) \\\n",
    "    + theme(panel_grid=element_blank(), \n",
    "            axis_title=element_text(face=\"bold\"))\n",
    "    return p\n",
    "    \n",
    "\n",
    "def simple_snp_expression_plot(snp_id, gene_id):\n",
    "    return simple_snp_expression_plot_impl(snp_id, gene_id, get_snp_df)\n",
    "\n",
    "\n",
    "def simple_gwas_ordered_snp_expression_plot(snp_id, gene_id):\n",
    "    return simple_snp_expression_plot_impl(snp_id, gene_id, get_gwas_ordered_snp_df)\n",
    "\n",
    "\n",
    "def get_gene_symbol(gene_id, biomart=get_biomart_df()):\n",
    "    ensge = re.sub('\\..+$','', gene_id)\n",
    "    ggg = biomart[biomart['ensembl_gene_id']==ensge]\n",
    "    if ggg.shape[0]==0:\n",
    "        return '', ''\n",
    "    gs = ggg['external_gene_name'].values[0]\n",
    "    de = ggg['description'].values[0]\n",
    "    if type(de)!=str:\n",
    "        de = ''\n",
    "    de = re.sub('\\[Source:.*$','',de)\n",
    "    return gs, de\n",
    "\n",
    "\n",
    "def get_gwas_snp(snp_id):\n",
    "    gwas = get_gwas_snps()\n",
    "    r = gwas[gwas['our_snp_id']==snp_id]\n",
    "    assert len(r) == 1\n",
    "    return r\n",
    "\n",
    "\n",
    "def gwas_annotation(snp_id):\n",
    "    return 'SZ GWAS pvalue: %.1e' % get_gwas_snp(snp_id).iloc[0]['P']\n",
    "\n",
    "\n",
    "def eqtl_annotation(snp_id, gene_id):\n",
    "    eqtl_df = get_eFeature_df()\n",
    "    r = eqtl_df[(eqtl_df['variant_id']==snp_id) & (eqtl_df['gene_id']==gene_id)]\n",
    "    assert len(r)==1\n",
    "    return 'eQTL adjusted p-value: %.1e' % r.iloc[0]['BF']\n",
    "\n",
    "\n",
    "def risk_allele_annotation(snp_id):\n",
    "    return 'SZ risk allele: %s' % get_risk_allele(snp_id)\n",
    "\n",
    "\n",
    "def annotated_eqtl_plot(snp_id, gene_id):\n",
    "    p = simple_snp_expression_plot(snp_id, gene_id)\n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join([gene_symbol,\n",
    "                     eqtl_annotation(snp_id, gene_id)\n",
    "                     ])\n",
    "    p += ggtitle(title) + ylab('Residualized Expression') \n",
    "    return p\n",
    "\n",
    "\n",
    "def gwas_annotated_eqtl_plot(snp_id, gene_id):\n",
    "    p = simple_gwas_ordered_snp_expression_plot(snp_id, gene_id)\n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join([gene_symbol,\n",
    "                     eqtl_annotation(snp_id, gene_id),\n",
    "                     gwas_annotation(snp_id),\n",
    "                     risk_allele_annotation(snp_id)\n",
    "                     ])\n",
    "    p += ggtitle(title) + ylab('Residualized Expression') \n",
    "    return p\n",
    "\n",
    "\n",
    "def save_plot(p, fn):\n",
    "    for ext in ['png', 'pdf', 'svg']:\n",
    "        p.save(fn + '.' + ext)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot eQTLs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRD2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, gene_id, gencodeID, slope, statistic, pval_nominal, BF, eigenMT_BH, TESTS, Type, Tissue]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_eFeature_df()[(get_eFeature_df()[\"gene_id\"] == gene_id_from_symbol('DRD2'))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top 5 eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55266</th>\n",
       "      <td>chrX:100404152:G:A</td>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>0.497189</td>\n",
       "      <td>17.628291</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.004420</td>\n",
       "      <td>0.478076</td>\n",
       "      <td>323</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Hippocampus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55267</th>\n",
       "      <td>chr1:196610909:G:A</td>\n",
       "      <td>ENSG00000000971.15</td>\n",
       "      <td>ENSG00000000971.15</td>\n",
       "      <td>-0.476843</td>\n",
       "      <td>-11.561317</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.001964</td>\n",
       "      <td>0.443931</td>\n",
       "      <td>151</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Hippocampus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55268</th>\n",
       "      <td>chr1:24888756:T:C</td>\n",
       "      <td>ENSG00000001461.16</td>\n",
       "      <td>ENSG00000001461.16</td>\n",
       "      <td>-0.311761</td>\n",
       "      <td>-11.705099</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.022525</td>\n",
       "      <td>0.634349</td>\n",
       "      <td>536</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Hippocampus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55269</th>\n",
       "      <td>chr7:42749666:G:C</td>\n",
       "      <td>ENSG00000002746.14</td>\n",
       "      <td>ENSG00000002746.14</td>\n",
       "      <td>0.314297</td>\n",
       "      <td>5.959216</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.033827</td>\n",
       "      <td>0.663881</td>\n",
       "      <td>440</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Hippocampus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55270</th>\n",
       "      <td>chr17:38495906:AT:A</td>\n",
       "      <td>ENSG00000002834.17</td>\n",
       "      <td>ENSG00000002834.17</td>\n",
       "      <td>0.477555</td>\n",
       "      <td>11.402260</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.015118</td>\n",
       "      <td>0.598413</td>\n",
       "      <td>608</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Hippocampus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                variant_id             gene_id           gencodeID     slope  \\\n",
       "55266   chrX:100404152:G:A   ENSG00000000005.5   ENSG00000000005.5  0.497189   \n",
       "55267   chr1:196610909:G:A  ENSG00000000971.15  ENSG00000000971.15 -0.476843   \n",
       "55268    chr1:24888756:T:C  ENSG00000001461.16  ENSG00000001461.16 -0.311761   \n",
       "55269    chr7:42749666:G:C  ENSG00000002746.14  ENSG00000002746.14  0.314297   \n",
       "55270  chr17:38495906:AT:A  ENSG00000002834.17  ENSG00000002834.17  0.477555   \n",
       "\n",
       "       statistic  pval_nominal        BF  eigenMT_BH  TESTS  Type       Tissue  \n",
       "55266  17.628291      0.000014  0.004420    0.478076    323  Gene  Hippocampus  \n",
       "55267 -11.561317      0.000013  0.001964    0.443931    151  Gene  Hippocampus  \n",
       "55268 -11.705099      0.000042  0.022525    0.634349    536  Gene  Hippocampus  \n",
       "55269   5.959216      0.000077  0.033827    0.663881    440  Gene  Hippocampus  \n",
       "55270  11.402260      0.000025  0.015118    0.598413    608  Gene  Hippocampus  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_df = get_eFeature_df()\n",
    "eqtl_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_5 = eqtl_df.sort_values('pval_nominal').reset_index(drop=True).head(5)\n",
    "for x in top_5.itertuples():\n",
    "    filename = \"top_%d_eqtl_%s\" % (x.Index, tissue)\n",
    "    p = annotated_eqtl_plot(x.variant_id, x.gene_id)\n",
    "    print(filename, x.Index, x.variant_id, x.gene_id)\n",
    "    print(p)\n",
    "    save_plot(p, filename)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top 5 X-linked genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_5_x = eqtl_df[eqtl_df['variant_id'].str.contains(\"chrX\")].sort_values(\"pval_nominal\").reset_index(drop=True).head(5)\n",
    "for x  in top_5_x.itertuples():\n",
    "    filename = \"top_%d_eqtl_xlinked_%s\" % (x.Index, tissue)\n",
    "    p = annotated_eqtl_plot(x.variant_id, x.gene_id)\n",
    "    print(filename, x.Index, x.variant_id, x.gene_id)\n",
    "    print(p)\n",
    "    save_plot(p, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top 5 eQTL with GWAS significant index SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_eqtl_df = eqtl_df.merge(get_gwas_snps(), left_on = 'variant_id', \n",
    "                             right_on = 'our_snp_id', suffixes=['','_gwas'])\n",
    "print(gwas_eqtl_df.shape)\n",
    "gwas_eqtl_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_gwas_eqtl_df = gwas_eqtl_df[(gwas_eqtl_df['is_index_snp'])].sort_values(['BF', 'P'])\n",
    "print(top_gwas_eqtl_df.shape)\n",
    "top_gwas_eqtl_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_gwas_eqtl_df = gwas_eqtl_df.sort_values(['BF', 'P']).reset_index(drop=True)\n",
    "print(top_gwas_eqtl_df.shape)\n",
    "top_gwas_eqtl_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_5_gwas = top_gwas_eqtl_df.head(5)\n",
    "for x  in top_5_gwas.itertuples():\n",
    "    filename = \"top_%d_eqtl_in_gwas_significant_snps_%s\" % (x.Index, tissue)\n",
    "    p = gwas_annotated_eqtl_plot(x.variant_id, x.gene_id)\n",
    "    print(filename, x.Index, x.variant_id, x.gene_id)\n",
    "    print(p)\n",
    "    save_plot(p, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
