{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examine overlaps with published data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:24.513776Z",
     "iopub.status.busy": "2021-09-14T13:25:24.512021Z",
     "iopub.status.idle": "2021-09-14T13:25:24.814431Z",
     "shell.execute_reply": "2021-09-14T13:25:24.815425Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Public si-eQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:24.828805Z",
     "iopub.status.busy": "2021-09-14T13:25:24.828139Z",
     "iopub.status.idle": "2021-09-14T13:25:24.834371Z",
     "shell.execute_reply": "2021-09-14T13:25:24.834848Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shen et al.:\n",
      "21\n",
      "Kukurba et al.:\n",
      "6\n",
      "Yao et al.:\n",
      "14\n",
      "Total of Yao + Kukurba:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shen = [\"GDAP2\", \"AIM2\", \"SLAMF6\", \"RLF\", \"ATG4C\", \"FUT7\",\n",
    "        \"TMEM218\", \"C11orf74\", \"RAB35\", \"TMEM5\", \"HNRNPK\",\n",
    "        \"CDCA3\", \"ERCC5\", \"GJB6\", \"SNTB2\", \"SPNS3\", \n",
    "        \"XAF1\", \"RBBP8\", \"RUFY4\", \"CA2\", \"RAPGEF1\"]\n",
    "print(\"Shen et al.:\")\n",
    "print(len(shen))\n",
    "\n",
    "kukurba = [\"NOD2\", \"WDR36\", \"BSCL2\", \"MAP7D3\", \"RHOXF1\", \"DNAH1\"]\n",
    "print(\"Kukurba et al.:\")\n",
    "print(len(kukurba))\n",
    "\n",
    "yao = [\"NOD2\", \"HLA-DRB5\", \"HLA-DRB5\", \"KIAA0586\", \"PPP2R5A\", \n",
    "       \"TSNAXIP1\", \"MUT\", \"GRIK2\", \"C15orf37\", \"LIMA1\", \"IL6ST\", \n",
    "       \"HCG8\", \"BLOC1S3\", \"NKX3-1\", \"CXorf23\"]\n",
    "print(\"Yao et al.:\")\n",
    "print(len(np.unique(yao)))\n",
    "print(\"Total of Yao + Kukurba:\")\n",
    "len(set(yao) | set(kukurba))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load BrainSeq si-eQTL results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interacting variant-gene pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:24.840485Z",
     "iopub.status.busy": "2021-09-14T13:25:24.839886Z",
     "iopub.status.idle": "2021-09-14T13:25:25.191203Z",
     "shell.execute_reply": "2021-09-14T13:25:25.190140Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(974, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15826</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>ENSG00000185386.14</td>\n",
       "      <td>ENSG00000185386.14</td>\n",
       "      <td>chr22:50563003:G:C</td>\n",
       "      <td>chr22</td>\n",
       "      <td>0.049299</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000185386</td>\n",
       "      <td>MAPK11</td>\n",
       "      <td>5600.0</td>\n",
       "      <td>mitogen-activated protein kinase 11 [Source:HG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15827</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>ENSG00000267251.2</td>\n",
       "      <td>ENSG00000267251.2</td>\n",
       "      <td>chr18:80244356:T:C</td>\n",
       "      <td>chr18</td>\n",
       "      <td>0.049375</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000267251</td>\n",
       "      <td>AC139100.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Tissue             gene_id           gencodeID  \\\n",
       "15826  Hippocampus  ENSG00000185386.14  ENSG00000185386.14   \n",
       "15827  Hippocampus   ENSG00000267251.2   ENSG00000267251.2   \n",
       "\n",
       "               variant_id seqnames      lfsr  Type  ensembl_gene_id  \\\n",
       "15826  chr22:50563003:G:C    chr22  0.049299  Gene  ENSG00000185386   \n",
       "15827  chr18:80244356:T:C    chr18  0.049375  Gene  ENSG00000267251   \n",
       "\n",
       "      external_gene_name  entrezgene  \\\n",
       "15826             MAPK11      5600.0   \n",
       "15827         AC139100.1         NaN   \n",
       "\n",
       "                                             description  \n",
       "15826  mitogen-activated protein kinase 11 [Source:HG...  \n",
       "15827                                                NaN  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs0 = pd.read_csv(\"../../summary_table/_m/BrainSeq_sexGenotypes_4features_3regions.txt.gz\", sep='\\t')\n",
    "bs0[\"ensembl_gene_id\"] = bs0.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "biomart = pd.read_csv(\"../../../residualized_expression/final_plot_eqtls/_h/biomart.csv\", index_col=0)\n",
    "bs = bs0.merge(biomart, on=\"ensembl_gene_id\").drop_duplicates(subset=\"gene_id\")\n",
    "print(bs.shape)\n",
    "bs.tail(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.207270Z",
     "iopub.status.busy": "2021-09-14T13:25:25.205547Z",
     "iopub.status.idle": "2021-09-14T13:25:25.211530Z",
     "shell.execute_reply": "2021-09-14T13:25:25.210450Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4670</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000125703.14</td>\n",
       "      <td>ENSG00000125703.14</td>\n",
       "      <td>chr1:63060301:G:A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.049896</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000125703</td>\n",
       "      <td>ATG4C</td>\n",
       "      <td>84938.0</td>\n",
       "      <td>autophagy related 4C cysteine peptidase [Sourc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Tissue             gene_id           gencodeID         variant_id  \\\n",
       "4670  Caudate  ENSG00000125703.14  ENSG00000125703.14  chr1:63060301:G:A   \n",
       "\n",
       "     seqnames      lfsr  Type  ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "4670     chr1  0.049896  Gene  ENSG00000125703              ATG4C     84938.0   \n",
       "\n",
       "                                            description  \n",
       "4670  autophagy related 4C cysteine peptidase [Sourc...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(shen))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.223492Z",
     "iopub.status.busy": "2021-09-14T13:25:25.221787Z",
     "iopub.status.idle": "2021-09-14T13:25:25.227580Z",
     "shell.execute_reply": "2021-09-14T13:25:25.226472Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Tissue, gene_id, gencodeID, variant_id, seqnames, lfsr, Type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(kukurba))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.239485Z",
     "iopub.status.busy": "2021-09-14T13:25:25.237816Z",
     "iopub.status.idle": "2021-09-14T13:25:25.244136Z",
     "shell.execute_reply": "2021-09-14T13:25:25.243040Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Tissue, gene_id, gencodeID, variant_id, seqnames, lfsr, Type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(yao))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.259846Z",
     "iopub.status.busy": "2021-09-14T13:25:25.258226Z",
     "iopub.status.idle": "2021-09-14T13:25:25.263915Z",
     "shell.execute_reply": "2021-09-14T13:25:25.262835Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4670</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000125703.14</td>\n",
       "      <td>ENSG00000125703.14</td>\n",
       "      <td>chr1:63060301:G:A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.049896</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000125703</td>\n",
       "      <td>ATG4C</td>\n",
       "      <td>84938.0</td>\n",
       "      <td>autophagy related 4C cysteine peptidase [Sourc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Tissue             gene_id           gencodeID         variant_id  \\\n",
       "4670  Caudate  ENSG00000125703.14  ENSG00000125703.14  chr1:63060301:G:A   \n",
       "\n",
       "     seqnames      lfsr  Type  ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "4670     chr1  0.049896  Gene  ENSG00000125703              ATG4C     84938.0   \n",
       "\n",
       "                                            description  \n",
       "4670  autophagy related 4C cysteine peptidase [Sourc...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(shen+kukurba+yao))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### eigenMT corrected p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.271457Z",
     "iopub.status.busy": "2021-09-14T13:25:25.270181Z",
     "iopub.status.idle": "2021-09-14T13:25:25.574287Z",
     "shell.execute_reply": "2021-09-14T13:25:25.573195Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5177, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7:42936690:A:C</td>\n",
       "      <td>ENSG00000002746.14</td>\n",
       "      <td>ENSG00000002746.14</td>\n",
       "      <td>0.436502</td>\n",
       "      <td>7.820971</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.010399</td>\n",
       "      <td>0.445692</td>\n",
       "      <td>445</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000002746</td>\n",
       "      <td>HECW1</td>\n",
       "      <td>23072.0</td>\n",
       "      <td>HECT, C2 and WW domain containing E3 ubiquitin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17:48075934:C:T</td>\n",
       "      <td>ENSG00000002919.14</td>\n",
       "      <td>ENSG00000002919.14</td>\n",
       "      <td>-0.536903</td>\n",
       "      <td>-7.959488</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.037424</td>\n",
       "      <td>0.549367</td>\n",
       "      <td>294</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000002919</td>\n",
       "      <td>SNX11</td>\n",
       "      <td>29916.0</td>\n",
       "      <td>sorting nexin 11 [Source:HGNC Symbol;Acc:HGNC:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           variant_id             gene_id           gencodeID     slope  \\\n",
       "0   chr7:42936690:A:C  ENSG00000002746.14  ENSG00000002746.14  0.436502   \n",
       "1  chr17:48075934:C:T  ENSG00000002919.14  ENSG00000002919.14 -0.536903   \n",
       "\n",
       "   statistic  pval_nominal        BF  eigenMT_BH  TESTS  Type  \\\n",
       "0   7.820971      0.000023  0.010399    0.445692    445  Gene   \n",
       "1  -7.959488      0.000127  0.037424    0.549367    294  Gene   \n",
       "\n",
       "   ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "0  ENSG00000002746              HECW1     23072.0   \n",
       "1  ENSG00000002919              SNX11     29916.0   \n",
       "\n",
       "                                         description  \n",
       "0  HECT, C2 and WW domain containing E3 ubiquitin...  \n",
       "1  sorting nexin 11 [Source:HGNC Symbol;Acc:HGNC:...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0 = pd.read_csv(\"../../../../summary_table/_m/Brainseq_sex_interacting_4features_3regions.eFeatures.txt.gz\", sep='\\t')\n",
    "df = df0[(df0[\"Type\"] == \"Gene\")].drop_duplicates(subset=\"gene_id\").drop(\"Tissue\", axis=1)\n",
    "df[\"ensembl_gene_id\"] = df.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "dft = df.merge(biomart, on=\"ensembl_gene_id\")\n",
    "print(dft.shape)\n",
    "dft.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.582241Z",
     "iopub.status.busy": "2021-09-14T13:25:25.580929Z",
     "iopub.status.idle": "2021-09-14T13:25:25.586410Z",
     "shell.execute_reply": "2021-09-14T13:25:25.585246Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "568"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Not all genes significant via mashr are also significant with eigenMT\n",
    "len(set(bs.gene_id) & set(dft.gene_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.607665Z",
     "iopub.status.busy": "2021-09-14T13:25:25.592501Z",
     "iopub.status.idle": "2021-09-14T13:25:25.610825Z",
     "shell.execute_reply": "2021-09-14T13:25:25.611849Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>chr1:63052156:A:G</td>\n",
       "      <td>ENSG00000125703.14</td>\n",
       "      <td>ENSG00000125703.14</td>\n",
       "      <td>0.210967</td>\n",
       "      <td>10.344946</td>\n",
       "      <td>7.923880e-05</td>\n",
       "      <td>0.037955</td>\n",
       "      <td>0.551098</td>\n",
       "      <td>479</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000125703</td>\n",
       "      <td>ATG4C</td>\n",
       "      <td>84938.0</td>\n",
       "      <td>autophagy related 4C cysteine peptidase [Sourc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>chr16:68729086:G:C</td>\n",
       "      <td>ENSG00000168807.16</td>\n",
       "      <td>ENSG00000168807.16</td>\n",
       "      <td>0.378104</td>\n",
       "      <td>11.451744</td>\n",
       "      <td>1.090640e-04</td>\n",
       "      <td>0.033483</td>\n",
       "      <td>0.541553</td>\n",
       "      <td>307</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000168807</td>\n",
       "      <td>SNTB2</td>\n",
       "      <td>6645.0</td>\n",
       "      <td>syntrophin beta 2 [Source:HGNC Symbol;Acc:HGNC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>chr13:20258404:TC:T</td>\n",
       "      <td>ENSG00000121742.16</td>\n",
       "      <td>ENSG00000121742.16</td>\n",
       "      <td>0.593176</td>\n",
       "      <td>10.551991</td>\n",
       "      <td>8.335770e-07</td>\n",
       "      <td>0.000542</td>\n",
       "      <td>0.317735</td>\n",
       "      <td>650</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000121742</td>\n",
       "      <td>GJB6</td>\n",
       "      <td>10804.0</td>\n",
       "      <td>gap junction protein beta 6 [Source:HGNC Symbo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2804</th>\n",
       "      <td>chr11:36347842:A:G</td>\n",
       "      <td>ENSG00000166352.15</td>\n",
       "      <td>ENSG00000166352.15</td>\n",
       "      <td>0.518875</td>\n",
       "      <td>8.165337</td>\n",
       "      <td>1.671190e-05</td>\n",
       "      <td>0.009492</td>\n",
       "      <td>0.539284</td>\n",
       "      <td>568</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000166352</td>\n",
       "      <td>C11orf74</td>\n",
       "      <td>119710.0</td>\n",
       "      <td>chromosome 11 open reading frame 74 [Source:HG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3861</th>\n",
       "      <td>chr8:84969351:T:C</td>\n",
       "      <td>ENSG00000104267.9</td>\n",
       "      <td>ENSG00000104267.9</td>\n",
       "      <td>-0.230131</td>\n",
       "      <td>-11.113888</td>\n",
       "      <td>3.085070e-04</td>\n",
       "      <td>0.040723</td>\n",
       "      <td>0.693507</td>\n",
       "      <td>132</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000104267</td>\n",
       "      <td>CA2</td>\n",
       "      <td>760.0</td>\n",
       "      <td>carbonic anhydrase 2 [Source:HGNC Symbol;Acc:H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4091</th>\n",
       "      <td>chr17:6627584:T:C</td>\n",
       "      <td>ENSG00000132530.16</td>\n",
       "      <td>ENSG00000132530.16</td>\n",
       "      <td>-0.515449</td>\n",
       "      <td>-14.299219</td>\n",
       "      <td>5.135140e-05</td>\n",
       "      <td>0.044214</td>\n",
       "      <td>0.698000</td>\n",
       "      <td>861</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000132530</td>\n",
       "      <td>XAF1</td>\n",
       "      <td>54739.0</td>\n",
       "      <td>XIAP associated factor 1 [Source:HGNC Symbol;A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               variant_id             gene_id           gencodeID     slope  \\\n",
       "486     chr1:63052156:A:G  ENSG00000125703.14  ENSG00000125703.14  0.210967   \n",
       "1029   chr16:68729086:G:C  ENSG00000168807.16  ENSG00000168807.16  0.378104   \n",
       "2399  chr13:20258404:TC:T  ENSG00000121742.16  ENSG00000121742.16  0.593176   \n",
       "2804   chr11:36347842:A:G  ENSG00000166352.15  ENSG00000166352.15  0.518875   \n",
       "3861    chr8:84969351:T:C   ENSG00000104267.9   ENSG00000104267.9 -0.230131   \n",
       "4091    chr17:6627584:T:C  ENSG00000132530.16  ENSG00000132530.16 -0.515449   \n",
       "\n",
       "      statistic  pval_nominal        BF  eigenMT_BH  TESTS  Type  \\\n",
       "486   10.344946  7.923880e-05  0.037955    0.551098    479  Gene   \n",
       "1029  11.451744  1.090640e-04  0.033483    0.541553    307  Gene   \n",
       "2399  10.551991  8.335770e-07  0.000542    0.317735    650  Gene   \n",
       "2804   8.165337  1.671190e-05  0.009492    0.539284    568  Gene   \n",
       "3861 -11.113888  3.085070e-04  0.040723    0.693507    132  Gene   \n",
       "4091 -14.299219  5.135140e-05  0.044214    0.698000    861  Gene   \n",
       "\n",
       "      ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "486   ENSG00000125703              ATG4C     84938.0   \n",
       "1029  ENSG00000168807              SNTB2      6645.0   \n",
       "2399  ENSG00000121742               GJB6     10804.0   \n",
       "2804  ENSG00000166352           C11orf74    119710.0   \n",
       "3861  ENSG00000104267                CA2       760.0   \n",
       "4091  ENSG00000132530               XAF1     54739.0   \n",
       "\n",
       "                                            description  \n",
       "486   autophagy related 4C cysteine peptidase [Sourc...  \n",
       "1029  syntrophin beta 2 [Source:HGNC Symbol;Acc:HGNC...  \n",
       "2399  gap junction protein beta 6 [Source:HGNC Symbo...  \n",
       "2804  chromosome 11 open reading frame 74 [Source:HG...  \n",
       "3861  carbonic anhydrase 2 [Source:HGNC Symbol;Acc:H...  \n",
       "4091  XIAP associated factor 1 [Source:HGNC Symbol;A...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft[(dft['external_gene_name'].isin(shen))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.619226Z",
     "iopub.status.busy": "2021-09-14T13:25:25.617995Z",
     "iopub.status.idle": "2021-09-14T13:25:25.631006Z",
     "shell.execute_reply": "2021-09-14T13:25:25.630486Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, gene_id, gencodeID, slope, statistic, pval_nominal, BF, eigenMT_BH, TESTS, Type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft[(dft['external_gene_name'].isin(shen)) & (dft[\"eigenMT_BH\"] < 0.25)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.642346Z",
     "iopub.status.busy": "2021-09-14T13:25:25.641758Z",
     "iopub.status.idle": "2021-09-14T13:25:25.644791Z",
     "shell.execute_reply": "2021-09-14T13:25:25.644295Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, gene_id, gencodeID, slope, statistic, pval_nominal, BF, eigenMT_BH, TESTS, Type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft[(dft['external_gene_name'].isin(kukurba)) & (dft[\"eigenMT_BH\"] < 0.25)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.655610Z",
     "iopub.status.busy": "2021-09-14T13:25:25.655020Z",
     "iopub.status.idle": "2021-09-14T13:25:25.657576Z",
     "shell.execute_reply": "2021-09-14T13:25:25.658043Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, gene_id, gencodeID, slope, statistic, pval_nominal, BF, eigenMT_BH, TESTS, Type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft[(dft['external_gene_name'].isin(yao)) & (dft[\"eigenMT_BH\"] < 0.25)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GTEx comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:25.662483Z",
     "iopub.status.busy": "2021-09-14T13:25:25.661914Z",
     "iopub.status.idle": "2021-09-14T13:25:27.198165Z",
     "shell.execute_reply": "2021-09-14T13:25:27.199212Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal_sb</th>\n",
       "      <th>slope_sb</th>\n",
       "      <th>slope_se_sb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000241860.6</td>\n",
       "      <td>RP11-34P13.13</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr1_14677_G_A_b38</td>\n",
       "      <td>rs201327123</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.051635</td>\n",
       "      <td>0.847114</td>\n",
       "      <td>0.055080</td>\n",
       "      <td>0.285537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>chr1_64764_C_T_b38</td>\n",
       "      <td>rs769952832</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.061102</td>\n",
       "      <td>0.316881</td>\n",
       "      <td>0.222928</td>\n",
       "      <td>0.222511</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ensembl_gene_id   hugo_gene_id               gene_type  \\\n",
       "0  ENSG00000241860.6  RP11-34P13.13    processed_transcript   \n",
       "1  ENSG00000227232.5         WASH7P  unprocessed_pseudogene   \n",
       "\n",
       "           variant_id        rs_id                Tissue       maf  \\\n",
       "0  chr1_14677_G_A_b38  rs201327123  Adipose_Subcutaneous  0.051635   \n",
       "1  chr1_64764_C_T_b38  rs769952832  Adipose_Subcutaneous  0.061102   \n",
       "\n",
       "   pval_nominal_sb  slope_sb  slope_se_sb  \n",
       "0         0.847114  0.055080     0.285537  \n",
       "1         0.316881  0.222928     0.222511  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex = pd.read_csv(\"/ceph/users/jbenja13/projects/sex_sz_ria/input/gtex_v8/\"+\\\n",
    "                   \"_m/GTEx_Analysis_v8_sbeQTLs/GTEx_Analysis_v8_sbeQTLs.txt\", \n",
    "                   sep='\\t')\n",
    "gtex.iloc[0:2, 0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.210306Z",
     "iopub.status.busy": "2021-09-14T13:25:27.208985Z",
     "iopub.status.idle": "2021-09-14T13:25:27.214355Z",
     "shell.execute_reply": "2021-09-14T13:25:27.213259Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numtested</th>\n",
       "      <th>pvals.corrected</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.847114</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.022302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.316881</td>\n",
       "      <td>0.981254</td>\n",
       "      <td>0.003978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   numtested  pvals.corrected      qval  pval_nominal_f\n",
       "0          1         0.847114  1.000000        0.022302\n",
       "1          1         0.316881  0.981254        0.003978"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex.iloc[0:2, 10:14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.222129Z",
     "iopub.status.busy": "2021-09-14T13:25:27.220855Z",
     "iopub.status.idle": "2021-09-14T13:25:27.379714Z",
     "shell.execute_reply": "2021-09-14T13:25:27.380761Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>pvals.corrected</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62155</th>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>VIM</td>\n",
       "      <td>Brain_Amygdala</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.012836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116842</th>\n",
       "      <td>ENSG00000160818.16</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>Brain_Nucleus_accumbens_basal_ganglia</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121904</th>\n",
       "      <td>ENSG00000141562.17</td>\n",
       "      <td>NARF</td>\n",
       "      <td>Brain_Nucleus_accumbens_basal_ganglia</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122123</th>\n",
       "      <td>ENSG00000267174.5</td>\n",
       "      <td>CTC-510F12.4</td>\n",
       "      <td>Brain_Nucleus_accumbens_basal_ganglia</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.198445</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ensembl_gene_id  hugo_gene_id  \\\n",
       "62155   ENSG00000026025.15           VIM   \n",
       "116842  ENSG00000160818.16       GPATCH4   \n",
       "121904  ENSG00000141562.17          NARF   \n",
       "122123   ENSG00000267174.5  CTC-510F12.4   \n",
       "\n",
       "                                       Tissue  pvals.corrected      qval  \n",
       "62155                          Brain_Amygdala         0.000004  0.012836  \n",
       "116842  Brain_Nucleus_accumbens_basal_ganglia         0.000088  0.198445  \n",
       "121904  Brain_Nucleus_accumbens_basal_ganglia         0.000056  0.198445  \n",
       "122123  Brain_Nucleus_accumbens_basal_ganglia         0.000083  0.198445  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## qval threshold equal to number of published sb-eQTL\n",
    "gtex[(gtex['qval'] < 0.25) & (gtex[\"Tissue\"].str.contains(\"Brain\"))]\\\n",
    "    .loc[:, [\"ensembl_gene_id\", \"hugo_gene_id\", \"Tissue\", \"pvals.corrected\", 'qval']].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.387597Z",
     "iopub.status.busy": "2021-09-14T13:25:27.386354Z",
     "iopub.status.idle": "2021-09-14T13:25:27.527269Z",
     "shell.execute_reply": "2021-09-14T13:25:27.526114Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>pvals.corrected</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>362961</th>\n",
       "      <td>ENSG00000221571.3</td>\n",
       "      <td>RNU6ATAC35P</td>\n",
       "      <td>Whole_Blood</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.139762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365043</th>\n",
       "      <td>ENSG00000196743.8</td>\n",
       "      <td>GM2A</td>\n",
       "      <td>Whole_Blood</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.116825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367164</th>\n",
       "      <td>ENSG00000148459.15</td>\n",
       "      <td>PDSS1</td>\n",
       "      <td>Whole_Blood</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.139762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ensembl_gene_id hugo_gene_id       Tissue  pvals.corrected  \\\n",
       "362961   ENSG00000221571.3  RNU6ATAC35P  Whole_Blood         0.000039   \n",
       "365043   ENSG00000196743.8         GM2A  Whole_Blood         0.000011   \n",
       "367164  ENSG00000148459.15        PDSS1  Whole_Blood         0.000027   \n",
       "\n",
       "            qval  \n",
       "362961  0.139762  \n",
       "365043  0.116825  \n",
       "367164  0.139762  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## qval threshold equal to number of published sb-eQTL\n",
    "gtex[(gtex['qval'] < 0.25) & (gtex[\"Tissue\"].str.contains(\"Whole\"))]\\\n",
    "    .loc[:, [\"ensembl_gene_id\", \"hugo_gene_id\", \"Tissue\", \"pvals.corrected\", 'qval']].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.533029Z",
     "iopub.status.busy": "2021-09-14T13:25:27.531755Z",
     "iopub.status.idle": "2021-09-14T13:25:27.537849Z",
     "shell.execute_reply": "2021-09-14T13:25:27.536739Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(369, 22)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_sig = gtex[(gtex['qval'] < 0.25)]\n",
    "gtex_sig.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.564806Z",
     "iopub.status.busy": "2021-09-14T13:25:27.563563Z",
     "iopub.status.idle": "2021-09-14T13:25:27.568461Z",
     "shell.execute_reply": "2021-09-14T13:25:27.569512Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal_sb</th>\n",
       "      <th>slope_sb</th>\n",
       "      <th>slope_se_sb</th>\n",
       "      <th>...</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_f</th>\n",
       "      <th>slope_f</th>\n",
       "      <th>slope_se_f</th>\n",
       "      <th>pval_nominal_m</th>\n",
       "      <th>slope_m</th>\n",
       "      <th>slope_se_m</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1096</th>\n",
       "      <td>ENSG00000076356.6</td>\n",
       "      <td>PLXNA2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_208030492_G_A_b38</td>\n",
       "      <td>rs3811383</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.123924</td>\n",
       "      <td>5.391600e-05</td>\n",
       "      <td>0.338278</td>\n",
       "      <td>0.083064</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121068</td>\n",
       "      <td>1.718880e-08</td>\n",
       "      <td>0.456729</td>\n",
       "      <td>0.075705</td>\n",
       "      <td>9.155700e-01</td>\n",
       "      <td>0.009739</td>\n",
       "      <td>0.091682</td>\n",
       "      <td>2.747400e-05</td>\n",
       "      <td>0.171830</td>\n",
       "      <td>0.040604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5262</th>\n",
       "      <td>ENSG00000170632.13</td>\n",
       "      <td>ARMC10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7_103076937_C_T_b38</td>\n",
       "      <td>rs6958836</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.169535</td>\n",
       "      <td>5.011130e-05</td>\n",
       "      <td>0.357403</td>\n",
       "      <td>0.087384</td>\n",
       "      <td>...</td>\n",
       "      <td>0.192900</td>\n",
       "      <td>4.933240e-01</td>\n",
       "      <td>-0.054539</td>\n",
       "      <td>0.079379</td>\n",
       "      <td>3.219220e-07</td>\n",
       "      <td>-0.429800</td>\n",
       "      <td>0.079545</td>\n",
       "      <td>8.797530e-08</td>\n",
       "      <td>-0.216374</td>\n",
       "      <td>0.039857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5644</th>\n",
       "      <td>ENSG00000120907.17</td>\n",
       "      <td>ADRA1A</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr8_26839198_G_A_b38</td>\n",
       "      <td>rs117380715</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.216867</td>\n",
       "      <td>1.045890e-05</td>\n",
       "      <td>-0.323552</td>\n",
       "      <td>0.072676</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084548</td>\n",
       "      <td>4.637410e-18</td>\n",
       "      <td>-0.779707</td>\n",
       "      <td>0.076596</td>\n",
       "      <td>3.976660e-10</td>\n",
       "      <td>-0.469672</td>\n",
       "      <td>0.069091</td>\n",
       "      <td>5.637370e-52</td>\n",
       "      <td>-0.568916</td>\n",
       "      <td>0.033334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6414</th>\n",
       "      <td>ENSG00000136830.11</td>\n",
       "      <td>FAM129B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr9_127584339_G_A_b38</td>\n",
       "      <td>rs10739693</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.304647</td>\n",
       "      <td>7.387010e-07</td>\n",
       "      <td>-0.283660</td>\n",
       "      <td>0.056579</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004976</td>\n",
       "      <td>1.978000e-06</td>\n",
       "      <td>-0.333315</td>\n",
       "      <td>0.066772</td>\n",
       "      <td>1.653380e-01</td>\n",
       "      <td>-0.082625</td>\n",
       "      <td>0.059205</td>\n",
       "      <td>1.393160e-08</td>\n",
       "      <td>-0.168762</td>\n",
       "      <td>0.029260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7220</th>\n",
       "      <td>ENSG00000166787.3</td>\n",
       "      <td>SAA3P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr11_18269355_T_C_b38</td>\n",
       "      <td>rs34068567</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.278830</td>\n",
       "      <td>2.207290e-05</td>\n",
       "      <td>0.323030</td>\n",
       "      <td>0.075427</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074347</td>\n",
       "      <td>6.409400e-08</td>\n",
       "      <td>0.453034</td>\n",
       "      <td>0.078725</td>\n",
       "      <td>3.138000e-01</td>\n",
       "      <td>0.063002</td>\n",
       "      <td>0.062292</td>\n",
       "      <td>2.433600e-08</td>\n",
       "      <td>0.211910</td>\n",
       "      <td>0.037395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8540</th>\n",
       "      <td>ENSG00000183463.5</td>\n",
       "      <td>URAD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr13_27990205_T_A_b38</td>\n",
       "      <td>rs7335293</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>9.078700e-09</td>\n",
       "      <td>-0.444892</td>\n",
       "      <td>0.076123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>9.982650e-21</td>\n",
       "      <td>-0.887723</td>\n",
       "      <td>0.078738</td>\n",
       "      <td>1.892290e-09</td>\n",
       "      <td>-0.457733</td>\n",
       "      <td>0.070571</td>\n",
       "      <td>3.077310e-53</td>\n",
       "      <td>-0.640604</td>\n",
       "      <td>0.036976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9191</th>\n",
       "      <td>ENSG00000282651.2</td>\n",
       "      <td>IGHV5-10-1</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>chr14_106114510_A_G_b38</td>\n",
       "      <td>rs4573838</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.419105</td>\n",
       "      <td>2.025150e-05</td>\n",
       "      <td>-0.406760</td>\n",
       "      <td>0.094541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074347</td>\n",
       "      <td>5.248710e-12</td>\n",
       "      <td>-0.682629</td>\n",
       "      <td>0.089412</td>\n",
       "      <td>2.805060e-03</td>\n",
       "      <td>-0.289091</td>\n",
       "      <td>0.094806</td>\n",
       "      <td>3.458420e-21</td>\n",
       "      <td>-0.445408</td>\n",
       "      <td>0.045073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14611</th>\n",
       "      <td>ENSG00000143933.16</td>\n",
       "      <td>CALM2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2_46225349_C_T_b38</td>\n",
       "      <td>rs12477148</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.072495</td>\n",
       "      <td>4.497930e-05</td>\n",
       "      <td>-0.480557</td>\n",
       "      <td>0.116471</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161955</td>\n",
       "      <td>4.747150e-04</td>\n",
       "      <td>-0.491287</td>\n",
       "      <td>0.134732</td>\n",
       "      <td>9.165740e-01</td>\n",
       "      <td>0.013043</td>\n",
       "      <td>0.124116</td>\n",
       "      <td>2.197750e-05</td>\n",
       "      <td>-0.246023</td>\n",
       "      <td>0.057281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15082</th>\n",
       "      <td>ENSG00000144410.4</td>\n",
       "      <td>CPO</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2_206822186_C_T_b38</td>\n",
       "      <td>rs12470278</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.097015</td>\n",
       "      <td>3.204120e-05</td>\n",
       "      <td>0.682291</td>\n",
       "      <td>0.162191</td>\n",
       "      <td>...</td>\n",
       "      <td>0.115370</td>\n",
       "      <td>1.165430e-01</td>\n",
       "      <td>0.280837</td>\n",
       "      <td>0.176978</td>\n",
       "      <td>4.431060e-06</td>\n",
       "      <td>-0.558002</td>\n",
       "      <td>0.113158</td>\n",
       "      <td>7.896400e-06</td>\n",
       "      <td>-0.320288</td>\n",
       "      <td>0.070745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17452</th>\n",
       "      <td>ENSG00000211698.2</td>\n",
       "      <td>TRGV4</td>\n",
       "      <td>TR_V_gene</td>\n",
       "      <td>chr7_38361995_A_C_b38</td>\n",
       "      <td>rs10233345</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.335821</td>\n",
       "      <td>6.438100e-05</td>\n",
       "      <td>0.427491</td>\n",
       "      <td>0.105837</td>\n",
       "      <td>...</td>\n",
       "      <td>0.139089</td>\n",
       "      <td>7.011230e-06</td>\n",
       "      <td>-0.481758</td>\n",
       "      <td>0.100091</td>\n",
       "      <td>8.857670e-15</td>\n",
       "      <td>-1.068840</td>\n",
       "      <td>0.112111</td>\n",
       "      <td>1.630390e-49</td>\n",
       "      <td>-0.838766</td>\n",
       "      <td>0.049090</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_gene_id hugo_gene_id                           gene_type  \\\n",
       "1096    ENSG00000076356.6       PLXNA2                      protein_coding   \n",
       "5262   ENSG00000170632.13       ARMC10                      protein_coding   \n",
       "5644   ENSG00000120907.17       ADRA1A                      protein_coding   \n",
       "6414   ENSG00000136830.11      FAM129B                      protein_coding   \n",
       "7220    ENSG00000166787.3        SAA3P  transcribed_unprocessed_pseudogene   \n",
       "8540    ENSG00000183463.5         URAD                      protein_coding   \n",
       "9191    ENSG00000282651.2   IGHV5-10-1                           IG_V_gene   \n",
       "14611  ENSG00000143933.16        CALM2                      protein_coding   \n",
       "15082   ENSG00000144410.4          CPO                      protein_coding   \n",
       "17452   ENSG00000211698.2        TRGV4                           TR_V_gene   \n",
       "\n",
       "                    variant_id        rs_id                    Tissue  \\\n",
       "1096    chr1_208030492_G_A_b38    rs3811383      Adipose_Subcutaneous   \n",
       "5262    chr7_103076937_C_T_b38    rs6958836      Adipose_Subcutaneous   \n",
       "5644     chr8_26839198_G_A_b38  rs117380715      Adipose_Subcutaneous   \n",
       "6414    chr9_127584339_G_A_b38   rs10739693      Adipose_Subcutaneous   \n",
       "7220    chr11_18269355_T_C_b38   rs34068567      Adipose_Subcutaneous   \n",
       "8540    chr13_27990205_T_A_b38    rs7335293      Adipose_Subcutaneous   \n",
       "9191   chr14_106114510_A_G_b38    rs4573838      Adipose_Subcutaneous   \n",
       "14611    chr2_46225349_C_T_b38   rs12477148  Adipose_Visceral_Omentum   \n",
       "15082   chr2_206822186_C_T_b38   rs12470278  Adipose_Visceral_Omentum   \n",
       "17452    chr7_38361995_A_C_b38   rs10233345  Adipose_Visceral_Omentum   \n",
       "\n",
       "            maf  pval_nominal_sb  slope_sb  slope_se_sb  ...      qval  \\\n",
       "1096   0.123924     5.391600e-05  0.338278     0.083064  ...  0.121068   \n",
       "5262   0.169535     5.011130e-05  0.357403     0.087384  ...  0.192900   \n",
       "5644   0.216867     1.045890e-05 -0.323552     0.072676  ...  0.084548   \n",
       "6414   0.304647     7.387010e-07 -0.283660     0.056579  ...  0.004976   \n",
       "7220   0.278830     2.207290e-05  0.323030     0.075427  ...  0.074347   \n",
       "8540   0.500000     9.078700e-09 -0.444892     0.076123  ...  0.000122   \n",
       "9191   0.419105     2.025150e-05 -0.406760     0.094541  ...  0.074347   \n",
       "14611  0.072495     4.497930e-05 -0.480557     0.116471  ...  0.161955   \n",
       "15082  0.097015     3.204120e-05  0.682291     0.162191  ...  0.115370   \n",
       "17452  0.335821     6.438100e-05  0.427491     0.105837  ...  0.139089   \n",
       "\n",
       "       pval_nominal_f   slope_f  slope_se_f  pval_nominal_m   slope_m  \\\n",
       "1096     1.718880e-08  0.456729    0.075705    9.155700e-01  0.009739   \n",
       "5262     4.933240e-01 -0.054539    0.079379    3.219220e-07 -0.429800   \n",
       "5644     4.637410e-18 -0.779707    0.076596    3.976660e-10 -0.469672   \n",
       "6414     1.978000e-06 -0.333315    0.066772    1.653380e-01 -0.082625   \n",
       "7220     6.409400e-08  0.453034    0.078725    3.138000e-01  0.063002   \n",
       "8540     9.982650e-21 -0.887723    0.078738    1.892290e-09 -0.457733   \n",
       "9191     5.248710e-12 -0.682629    0.089412    2.805060e-03 -0.289091   \n",
       "14611    4.747150e-04 -0.491287    0.134732    9.165740e-01  0.013043   \n",
       "15082    1.165430e-01  0.280837    0.176978    4.431060e-06 -0.558002   \n",
       "17452    7.011230e-06 -0.481758    0.100091    8.857670e-15 -1.068840   \n",
       "\n",
       "       slope_se_m  pval_nominal     slope  slope_se  \n",
       "1096     0.091682  2.747400e-05  0.171830  0.040604  \n",
       "5262     0.079545  8.797530e-08 -0.216374  0.039857  \n",
       "5644     0.069091  5.637370e-52 -0.568916  0.033334  \n",
       "6414     0.059205  1.393160e-08 -0.168762  0.029260  \n",
       "7220     0.062292  2.433600e-08  0.211910  0.037395  \n",
       "8540     0.070571  3.077310e-53 -0.640604  0.036976  \n",
       "9191     0.094806  3.458420e-21 -0.445408  0.045073  \n",
       "14611    0.124116  2.197750e-05 -0.246023  0.057281  \n",
       "15082    0.113158  7.896400e-06 -0.320288  0.070745  \n",
       "17452    0.112111  1.630390e-49 -0.838766  0.049090  \n",
       "\n",
       "[10 rows x 22 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_sig.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mashr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.575625Z",
     "iopub.status.busy": "2021-09-14T13:25:27.574397Z",
     "iopub.status.idle": "2021-09-14T13:25:27.594920Z",
     "shell.execute_reply": "2021-09-14T13:25:27.595950Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2638</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000105695.14</td>\n",
       "      <td>ENSG00000105695.14</td>\n",
       "      <td>chr19:35065453:C:T</td>\n",
       "      <td>chr19</td>\n",
       "      <td>0.044690</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000105695</td>\n",
       "      <td>MAG</td>\n",
       "      <td>4099.0</td>\n",
       "      <td>myelin associated glycoprotein [Source:HGNC Sy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3163</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000109775.10</td>\n",
       "      <td>ENSG00000109775.10</td>\n",
       "      <td>chr4:185462033:A:G</td>\n",
       "      <td>chr4</td>\n",
       "      <td>0.034580</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000109775</td>\n",
       "      <td>UFSP2</td>\n",
       "      <td>55325.0</td>\n",
       "      <td>UFM1 specific peptidase 2 [Source:HGNC Symbol;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4518</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>chr4:42387723:C:A</td>\n",
       "      <td>chr4</td>\n",
       "      <td>0.013706</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000124406</td>\n",
       "      <td>ATP8A1</td>\n",
       "      <td>10396.0</td>\n",
       "      <td>ATPase phospholipid transporting 8A1 [Source:H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000132199.18</td>\n",
       "      <td>ENSG00000132199.18</td>\n",
       "      <td>chr18:381997:C:G</td>\n",
       "      <td>chr18</td>\n",
       "      <td>0.044197</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000132199</td>\n",
       "      <td>ENOSF1</td>\n",
       "      <td>55556.0</td>\n",
       "      <td>enolase superfamily member 1 [Source:HGNC Symb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5635</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000136830.11</td>\n",
       "      <td>ENSG00000136830.11</td>\n",
       "      <td>chr9:127456523:G:A</td>\n",
       "      <td>chr9</td>\n",
       "      <td>0.044628</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000136830</td>\n",
       "      <td>FAM129B</td>\n",
       "      <td>64855.0</td>\n",
       "      <td>family with sequence similarity 129 member B [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6605</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000143933.16</td>\n",
       "      <td>ENSG00000143933.16</td>\n",
       "      <td>chr2:47607174:T:C</td>\n",
       "      <td>chr2</td>\n",
       "      <td>0.022399</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000143933</td>\n",
       "      <td>CALM2</td>\n",
       "      <td>805.0</td>\n",
       "      <td>calmodulin 2 [Source:HGNC Symbol;Acc:HGNC:1445]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6687</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000146530.11</td>\n",
       "      <td>ENSG00000146530.11</td>\n",
       "      <td>chr7:12170840:G:A</td>\n",
       "      <td>chr7</td>\n",
       "      <td>0.023059</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000146530</td>\n",
       "      <td>VWDE</td>\n",
       "      <td>221806.0</td>\n",
       "      <td>von Willebrand factor D and EGF domains [Sourc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10405</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000181264.8</td>\n",
       "      <td>ENSG00000181264.8</td>\n",
       "      <td>chr11:120171373:G:A</td>\n",
       "      <td>chr11</td>\n",
       "      <td>0.002548</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000181264</td>\n",
       "      <td>TMEM136</td>\n",
       "      <td>219902.0</td>\n",
       "      <td>transmembrane protein 136 [Source:HGNC Symbol;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11623</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000189067.12</td>\n",
       "      <td>ENSG00000189067.12</td>\n",
       "      <td>chr16:11300271:A:G</td>\n",
       "      <td>chr16</td>\n",
       "      <td>0.007562</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000189067</td>\n",
       "      <td>LITAF</td>\n",
       "      <td>9516.0</td>\n",
       "      <td>lipopolysaccharide induced TNF factor [Source:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12455</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000204248.10</td>\n",
       "      <td>ENSG00000204248.10</td>\n",
       "      <td>chr6:33170900:G:A</td>\n",
       "      <td>chr6</td>\n",
       "      <td>0.045996</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000204248</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>1302.0</td>\n",
       "      <td>collagen type XI alpha 2 chain [Source:HGNC Sy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14958</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000235098.8</td>\n",
       "      <td>ENSG00000235098.8</td>\n",
       "      <td>chr1:1425750:T:C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>0.006797</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000235098</td>\n",
       "      <td>ANKRD65</td>\n",
       "      <td>441869.0</td>\n",
       "      <td>ankyrin repeat domain 65 [Source:HGNC Symbol;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15443</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000267056.2</td>\n",
       "      <td>ENSG00000267056.2</td>\n",
       "      <td>chr19:16385981:G:A</td>\n",
       "      <td>chr19</td>\n",
       "      <td>0.040125</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000267056</td>\n",
       "      <td>AC005336.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15478</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000268362.5</td>\n",
       "      <td>ENSG00000268362.5</td>\n",
       "      <td>chr19:23910181:A:G</td>\n",
       "      <td>chr19</td>\n",
       "      <td>0.042637</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000268362</td>\n",
       "      <td>AC092279.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15565</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>chr22:25901440:A:C</td>\n",
       "      <td>chr22</td>\n",
       "      <td>0.004803</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000272977</td>\n",
       "      <td>AL008721.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Tissue             gene_id           gencodeID           variant_id  \\\n",
       "2638   Caudate  ENSG00000105695.14  ENSG00000105695.14   chr19:35065453:C:T   \n",
       "3163   Caudate  ENSG00000109775.10  ENSG00000109775.10   chr4:185462033:A:G   \n",
       "4518   Caudate  ENSG00000124406.16  ENSG00000124406.16    chr4:42387723:C:A   \n",
       "5118   Caudate  ENSG00000132199.18  ENSG00000132199.18     chr18:381997:C:G   \n",
       "5635   Caudate  ENSG00000136830.11  ENSG00000136830.11   chr9:127456523:G:A   \n",
       "6605   Caudate  ENSG00000143933.16  ENSG00000143933.16    chr2:47607174:T:C   \n",
       "6687   Caudate  ENSG00000146530.11  ENSG00000146530.11    chr7:12170840:G:A   \n",
       "10405  Caudate   ENSG00000181264.8   ENSG00000181264.8  chr11:120171373:G:A   \n",
       "11623  Caudate  ENSG00000189067.12  ENSG00000189067.12   chr16:11300271:A:G   \n",
       "12455  Caudate  ENSG00000204248.10  ENSG00000204248.10    chr6:33170900:G:A   \n",
       "14958  Caudate   ENSG00000235098.8   ENSG00000235098.8     chr1:1425750:T:C   \n",
       "15443  Caudate   ENSG00000267056.2   ENSG00000267056.2   chr19:16385981:G:A   \n",
       "15478  Caudate   ENSG00000268362.5   ENSG00000268362.5   chr19:23910181:A:G   \n",
       "15565  Caudate   ENSG00000272977.1   ENSG00000272977.1   chr22:25901440:A:C   \n",
       "\n",
       "      seqnames      lfsr  Type  ensembl_gene_id external_gene_name  \\\n",
       "2638     chr19  0.044690  Gene  ENSG00000105695                MAG   \n",
       "3163      chr4  0.034580  Gene  ENSG00000109775              UFSP2   \n",
       "4518      chr4  0.013706  Gene  ENSG00000124406             ATP8A1   \n",
       "5118     chr18  0.044197  Gene  ENSG00000132199             ENOSF1   \n",
       "5635      chr9  0.044628  Gene  ENSG00000136830            FAM129B   \n",
       "6605      chr2  0.022399  Gene  ENSG00000143933              CALM2   \n",
       "6687      chr7  0.023059  Gene  ENSG00000146530               VWDE   \n",
       "10405    chr11  0.002548  Gene  ENSG00000181264            TMEM136   \n",
       "11623    chr16  0.007562  Gene  ENSG00000189067              LITAF   \n",
       "12455     chr6  0.045996  Gene  ENSG00000204248            COL11A2   \n",
       "14958     chr1  0.006797  Gene  ENSG00000235098            ANKRD65   \n",
       "15443    chr19  0.040125  Gene  ENSG00000267056         AC005336.1   \n",
       "15478    chr19  0.042637  Gene  ENSG00000268362         AC092279.1   \n",
       "15565    chr22  0.004803  Gene  ENSG00000272977         AL008721.2   \n",
       "\n",
       "       entrezgene                                        description  \n",
       "2638       4099.0  myelin associated glycoprotein [Source:HGNC Sy...  \n",
       "3163      55325.0  UFM1 specific peptidase 2 [Source:HGNC Symbol;...  \n",
       "4518      10396.0  ATPase phospholipid transporting 8A1 [Source:H...  \n",
       "5118      55556.0  enolase superfamily member 1 [Source:HGNC Symb...  \n",
       "5635      64855.0  family with sequence similarity 129 member B [...  \n",
       "6605        805.0    calmodulin 2 [Source:HGNC Symbol;Acc:HGNC:1445]  \n",
       "6687     221806.0  von Willebrand factor D and EGF domains [Sourc...  \n",
       "10405    219902.0  transmembrane protein 136 [Source:HGNC Symbol;...  \n",
       "11623      9516.0  lipopolysaccharide induced TNF factor [Source:...  \n",
       "12455      1302.0  collagen type XI alpha 2 chain [Source:HGNC Sy...  \n",
       "14958    441869.0  ankyrin repeat domain 65 [Source:HGNC Symbol;A...  \n",
       "15443         NaN                                                NaN  \n",
       "15478         NaN                                                NaN  \n",
       "15565         NaN                                                NaN  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap = bs[(bs['gene_id'].isin(gtex_sig.ensembl_gene_id))].drop_duplicates()\n",
    "print(gtex_overlap.shape)\n",
    "gtex_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.601921Z",
     "iopub.status.busy": "2021-09-14T13:25:27.600692Z",
     "iopub.status.idle": "2021-09-14T13:25:27.605862Z",
     "shell.execute_reply": "2021-09-14T13:25:27.604779Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.4373716632443532"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap.shape[0]/bs.shape[0] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.635733Z",
     "iopub.status.busy": "2021-09-14T13:25:27.634512Z",
     "iopub.status.idle": "2021-09-14T13:25:27.638946Z",
     "shell.execute_reply": "2021-09-14T13:25:27.639983Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal_sb</th>\n",
       "      <th>slope_sb</th>\n",
       "      <th>slope_se_sb</th>\n",
       "      <th>...</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_f</th>\n",
       "      <th>slope_f</th>\n",
       "      <th>slope_se_f</th>\n",
       "      <th>pval_nominal_m</th>\n",
       "      <th>slope_m</th>\n",
       "      <th>slope_se_m</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6414</th>\n",
       "      <td>ENSG00000136830.11</td>\n",
       "      <td>FAM129B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr9_127584339_G_A_b38</td>\n",
       "      <td>rs10739693</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.304647</td>\n",
       "      <td>7.387010e-07</td>\n",
       "      <td>-0.283660</td>\n",
       "      <td>0.056579</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004976</td>\n",
       "      <td>1.978000e-06</td>\n",
       "      <td>-0.333315</td>\n",
       "      <td>0.066772</td>\n",
       "      <td>1.653380e-01</td>\n",
       "      <td>-0.082625</td>\n",
       "      <td>0.059205</td>\n",
       "      <td>1.393160e-08</td>\n",
       "      <td>-0.168762</td>\n",
       "      <td>0.029260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14611</th>\n",
       "      <td>ENSG00000143933.16</td>\n",
       "      <td>CALM2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2_46225349_C_T_b38</td>\n",
       "      <td>rs12477148</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.072495</td>\n",
       "      <td>4.497930e-05</td>\n",
       "      <td>-0.480557</td>\n",
       "      <td>0.116471</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161955</td>\n",
       "      <td>4.747150e-04</td>\n",
       "      <td>-0.491287</td>\n",
       "      <td>0.134732</td>\n",
       "      <td>9.165740e-01</td>\n",
       "      <td>0.013043</td>\n",
       "      <td>0.124116</td>\n",
       "      <td>2.197750e-05</td>\n",
       "      <td>-0.246023</td>\n",
       "      <td>0.057281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45230</th>\n",
       "      <td>ENSG00000181264.8</td>\n",
       "      <td>TMEM136</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr11_120493228_T_C_b38</td>\n",
       "      <td>rs4938809</td>\n",
       "      <td>Artery_Coronary</td>\n",
       "      <td>0.403756</td>\n",
       "      <td>1.832690e-05</td>\n",
       "      <td>0.332900</td>\n",
       "      <td>0.075523</td>\n",
       "      <td>...</td>\n",
       "      <td>0.098269</td>\n",
       "      <td>6.159840e-08</td>\n",
       "      <td>0.485968</td>\n",
       "      <td>0.074741</td>\n",
       "      <td>1.990100e-01</td>\n",
       "      <td>0.113737</td>\n",
       "      <td>0.087221</td>\n",
       "      <td>8.491350e-08</td>\n",
       "      <td>0.231096</td>\n",
       "      <td>0.041304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135656</th>\n",
       "      <td>ENSG00000235098.8</td>\n",
       "      <td>ANKRD65</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_1419214_A_G_b38</td>\n",
       "      <td>rs3766165</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.131313</td>\n",
       "      <td>7.063370e-08</td>\n",
       "      <td>0.499559</td>\n",
       "      <td>0.090555</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>7.568940e-06</td>\n",
       "      <td>-0.439928</td>\n",
       "      <td>0.091937</td>\n",
       "      <td>1.443540e-11</td>\n",
       "      <td>-0.879777</td>\n",
       "      <td>0.111990</td>\n",
       "      <td>3.281960e-30</td>\n",
       "      <td>-0.726153</td>\n",
       "      <td>0.057266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137724</th>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>ATP8A1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr4_42653682_A_T_b38</td>\n",
       "      <td>rs17448575</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.354430</td>\n",
       "      <td>2.577750e-04</td>\n",
       "      <td>0.246240</td>\n",
       "      <td>0.066638</td>\n",
       "      <td>...</td>\n",
       "      <td>0.090512</td>\n",
       "      <td>2.989190e-07</td>\n",
       "      <td>-0.323359</td>\n",
       "      <td>0.057862</td>\n",
       "      <td>1.095640e-13</td>\n",
       "      <td>-0.613169</td>\n",
       "      <td>0.068668</td>\n",
       "      <td>1.462140e-26</td>\n",
       "      <td>-0.370315</td>\n",
       "      <td>0.031672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137946</th>\n",
       "      <td>ENSG00000109775.10</td>\n",
       "      <td>UFSP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr4_185425919_G_C_b38</td>\n",
       "      <td>rs11132303</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.309343</td>\n",
       "      <td>4.674540e-07</td>\n",
       "      <td>-0.348749</td>\n",
       "      <td>0.067806</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>9.942900e-10</td>\n",
       "      <td>0.432381</td>\n",
       "      <td>0.062550</td>\n",
       "      <td>7.750080e-18</td>\n",
       "      <td>0.758610</td>\n",
       "      <td>0.068668</td>\n",
       "      <td>2.014910e-57</td>\n",
       "      <td>0.706155</td>\n",
       "      <td>0.035862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138621</th>\n",
       "      <td>ENSG00000204248.10</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr6_33081200_G_GA_b38</td>\n",
       "      <td>rs113353922</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.118687</td>\n",
       "      <td>6.199050e-04</td>\n",
       "      <td>0.418594</td>\n",
       "      <td>0.121102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.102197</td>\n",
       "      <td>2.304030e-01</td>\n",
       "      <td>-0.125417</td>\n",
       "      <td>0.103786</td>\n",
       "      <td>2.720980e-04</td>\n",
       "      <td>-0.553034</td>\n",
       "      <td>0.145261</td>\n",
       "      <td>4.482050e-06</td>\n",
       "      <td>-0.293079</td>\n",
       "      <td>0.062803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138995</th>\n",
       "      <td>ENSG00000146530.11</td>\n",
       "      <td>VWDE</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7_12312811_T_C_b38</td>\n",
       "      <td>rs4721084</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.452020</td>\n",
       "      <td>1.741340e-06</td>\n",
       "      <td>-0.301523</td>\n",
       "      <td>0.061908</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000744</td>\n",
       "      <td>5.408710e-12</td>\n",
       "      <td>-0.446355</td>\n",
       "      <td>0.055294</td>\n",
       "      <td>1.501580e-03</td>\n",
       "      <td>-0.237816</td>\n",
       "      <td>0.072364</td>\n",
       "      <td>2.765010e-20</td>\n",
       "      <td>-0.319694</td>\n",
       "      <td>0.032375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142471</th>\n",
       "      <td>ENSG00000189067.12</td>\n",
       "      <td>LITAF</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr16_11551157_A_T_b38</td>\n",
       "      <td>rs11644920</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.339646</td>\n",
       "      <td>1.196880e-04</td>\n",
       "      <td>0.247997</td>\n",
       "      <td>0.063687</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035705</td>\n",
       "      <td>3.307190e-04</td>\n",
       "      <td>-0.171901</td>\n",
       "      <td>0.045841</td>\n",
       "      <td>2.925070e-07</td>\n",
       "      <td>-0.385154</td>\n",
       "      <td>0.068855</td>\n",
       "      <td>9.495520e-22</td>\n",
       "      <td>-0.323697</td>\n",
       "      <td>0.031398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143690</th>\n",
       "      <td>ENSG00000267056.2</td>\n",
       "      <td>AC005336.4</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr19_15934939_G_A_b38</td>\n",
       "      <td>rs12985091</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.478535</td>\n",
       "      <td>7.228630e-05</td>\n",
       "      <td>-0.373013</td>\n",
       "      <td>0.092781</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015797</td>\n",
       "      <td>1.961980e-07</td>\n",
       "      <td>-0.587505</td>\n",
       "      <td>0.103261</td>\n",
       "      <td>1.935340e-03</td>\n",
       "      <td>-0.245596</td>\n",
       "      <td>0.076638</td>\n",
       "      <td>1.176010e-12</td>\n",
       "      <td>-0.331468</td>\n",
       "      <td>0.044801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143783</th>\n",
       "      <td>ENSG00000268362.5</td>\n",
       "      <td>CTD-2017D11.1</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>chr19_24059854_C_T_b38</td>\n",
       "      <td>rs17206087</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.170455</td>\n",
       "      <td>2.481440e-03</td>\n",
       "      <td>-0.260946</td>\n",
       "      <td>0.085573</td>\n",
       "      <td>...</td>\n",
       "      <td>0.155454</td>\n",
       "      <td>1.501750e-17</td>\n",
       "      <td>0.858155</td>\n",
       "      <td>0.078738</td>\n",
       "      <td>8.474450e-17</td>\n",
       "      <td>0.981955</td>\n",
       "      <td>0.093415</td>\n",
       "      <td>1.509120e-67</td>\n",
       "      <td>0.972900</td>\n",
       "      <td>0.043631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143823</th>\n",
       "      <td>ENSG00000105695.14</td>\n",
       "      <td>MAG</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr19_35309759_G_T_b38</td>\n",
       "      <td>rs10411704</td>\n",
       "      <td>Breast_Mammary_Tissue</td>\n",
       "      <td>0.179293</td>\n",
       "      <td>3.726460e-06</td>\n",
       "      <td>-0.621945</td>\n",
       "      <td>0.132132</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001446</td>\n",
       "      <td>8.536830e-14</td>\n",
       "      <td>-1.030720</td>\n",
       "      <td>0.114726</td>\n",
       "      <td>1.809540e-04</td>\n",
       "      <td>-0.520457</td>\n",
       "      <td>0.132593</td>\n",
       "      <td>8.545960e-22</td>\n",
       "      <td>-0.693095</td>\n",
       "      <td>0.067157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263405</th>\n",
       "      <td>ENSG00000132199.18</td>\n",
       "      <td>ENOSF1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18_712037_G_A_b38</td>\n",
       "      <td>rs2612103</td>\n",
       "      <td>Muscle_Skeletal</td>\n",
       "      <td>0.443343</td>\n",
       "      <td>2.579330e-05</td>\n",
       "      <td>-0.266917</td>\n",
       "      <td>0.062968</td>\n",
       "      <td>...</td>\n",
       "      <td>0.123976</td>\n",
       "      <td>1.058820e-29</td>\n",
       "      <td>-0.787918</td>\n",
       "      <td>0.056652</td>\n",
       "      <td>4.149110e-16</td>\n",
       "      <td>-0.508209</td>\n",
       "      <td>0.056303</td>\n",
       "      <td>5.746020e-65</td>\n",
       "      <td>-0.613437</td>\n",
       "      <td>0.031989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338770</th>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>CTA-390C10.10</td>\n",
       "      <td>sense_intronic</td>\n",
       "      <td>chr22_25459662_G_A_b38</td>\n",
       "      <td>rs6004655</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>0.167401</td>\n",
       "      <td>3.787520e-05</td>\n",
       "      <td>0.413128</td>\n",
       "      <td>0.097856</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212883</td>\n",
       "      <td>7.931210e-14</td>\n",
       "      <td>-1.129030</td>\n",
       "      <td>0.107327</td>\n",
       "      <td>5.155260e-16</td>\n",
       "      <td>-1.276590</td>\n",
       "      <td>0.104690</td>\n",
       "      <td>4.896640e-51</td>\n",
       "      <td>-1.193670</td>\n",
       "      <td>0.056659</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           ensembl_gene_id   hugo_gene_id             gene_type  \\\n",
       "6414    ENSG00000136830.11        FAM129B        protein_coding   \n",
       "14611   ENSG00000143933.16          CALM2        protein_coding   \n",
       "45230    ENSG00000181264.8        TMEM136        protein_coding   \n",
       "135656   ENSG00000235098.8        ANKRD65        protein_coding   \n",
       "137724  ENSG00000124406.16         ATP8A1        protein_coding   \n",
       "137946  ENSG00000109775.10          UFSP2        protein_coding   \n",
       "138621  ENSG00000204248.10        COL11A2        protein_coding   \n",
       "138995  ENSG00000146530.11           VWDE        protein_coding   \n",
       "142471  ENSG00000189067.12          LITAF        protein_coding   \n",
       "143690   ENSG00000267056.2     AC005336.4  processed_pseudogene   \n",
       "143783   ENSG00000268362.5  CTD-2017D11.1               lincRNA   \n",
       "143823  ENSG00000105695.14            MAG        protein_coding   \n",
       "263405  ENSG00000132199.18         ENOSF1        protein_coding   \n",
       "338770   ENSG00000272977.1  CTA-390C10.10        sense_intronic   \n",
       "\n",
       "                     variant_id        rs_id                    Tissue  \\\n",
       "6414     chr9_127584339_G_A_b38   rs10739693      Adipose_Subcutaneous   \n",
       "14611     chr2_46225349_C_T_b38   rs12477148  Adipose_Visceral_Omentum   \n",
       "45230   chr11_120493228_T_C_b38    rs4938809           Artery_Coronary   \n",
       "135656     chr1_1419214_A_G_b38    rs3766165     Breast_Mammary_Tissue   \n",
       "137724    chr4_42653682_A_T_b38   rs17448575     Breast_Mammary_Tissue   \n",
       "137946   chr4_185425919_G_C_b38   rs11132303     Breast_Mammary_Tissue   \n",
       "138621   chr6_33081200_G_GA_b38  rs113353922     Breast_Mammary_Tissue   \n",
       "138995    chr7_12312811_T_C_b38    rs4721084     Breast_Mammary_Tissue   \n",
       "142471   chr16_11551157_A_T_b38   rs11644920     Breast_Mammary_Tissue   \n",
       "143690   chr19_15934939_G_A_b38   rs12985091     Breast_Mammary_Tissue   \n",
       "143783   chr19_24059854_C_T_b38   rs17206087     Breast_Mammary_Tissue   \n",
       "143823   chr19_35309759_G_T_b38   rs10411704     Breast_Mammary_Tissue   \n",
       "263405     chr18_712037_G_A_b38    rs2612103           Muscle_Skeletal   \n",
       "338770   chr22_25459662_G_A_b38    rs6004655                    Spleen   \n",
       "\n",
       "             maf  pval_nominal_sb  slope_sb  slope_se_sb  ...      qval  \\\n",
       "6414    0.304647     7.387010e-07 -0.283660     0.056579  ...  0.004976   \n",
       "14611   0.072495     4.497930e-05 -0.480557     0.116471  ...  0.161955   \n",
       "45230   0.403756     1.832690e-05  0.332900     0.075523  ...  0.098269   \n",
       "135656  0.131313     7.063370e-08  0.499559     0.090555  ...  0.000111   \n",
       "137724  0.354430     2.577750e-04  0.246240     0.066638  ...  0.090512   \n",
       "137946  0.309343     4.674540e-07 -0.348749     0.067806  ...  0.000399   \n",
       "138621  0.118687     6.199050e-04  0.418594     0.121102  ...  0.102197   \n",
       "138995  0.452020     1.741340e-06 -0.301523     0.061908  ...  0.000744   \n",
       "142471  0.339646     1.196880e-04  0.247997     0.063687  ...  0.035705   \n",
       "143690  0.478535     7.228630e-05 -0.373013     0.092781  ...  0.015797   \n",
       "143783  0.170455     2.481440e-03 -0.260946     0.085573  ...  0.155454   \n",
       "143823  0.179293     3.726460e-06 -0.621945     0.132132  ...  0.001446   \n",
       "263405  0.443343     2.579330e-05 -0.266917     0.062968  ...  0.123976   \n",
       "338770  0.167401     3.787520e-05  0.413128     0.097856  ...  0.212883   \n",
       "\n",
       "        pval_nominal_f   slope_f  slope_se_f  pval_nominal_m   slope_m  \\\n",
       "6414      1.978000e-06 -0.333315    0.066772    1.653380e-01 -0.082625   \n",
       "14611     4.747150e-04 -0.491287    0.134732    9.165740e-01  0.013043   \n",
       "45230     6.159840e-08  0.485968    0.074741    1.990100e-01  0.113737   \n",
       "135656    7.568940e-06 -0.439928    0.091937    1.443540e-11 -0.879777   \n",
       "137724    2.989190e-07 -0.323359    0.057862    1.095640e-13 -0.613169   \n",
       "137946    9.942900e-10  0.432381    0.062550    7.750080e-18  0.758610   \n",
       "138621    2.304030e-01 -0.125417    0.103786    2.720980e-04 -0.553034   \n",
       "138995    5.408710e-12 -0.446355    0.055294    1.501580e-03 -0.237816   \n",
       "142471    3.307190e-04 -0.171901    0.045841    2.925070e-07 -0.385154   \n",
       "143690    1.961980e-07 -0.587505    0.103261    1.935340e-03 -0.245596   \n",
       "143783    1.501750e-17  0.858155    0.078738    8.474450e-17  0.981955   \n",
       "143823    8.536830e-14 -1.030720    0.114726    1.809540e-04 -0.520457   \n",
       "263405    1.058820e-29 -0.787918    0.056652    4.149110e-16 -0.508209   \n",
       "338770    7.931210e-14 -1.129030    0.107327    5.155260e-16 -1.276590   \n",
       "\n",
       "        slope_se_m  pval_nominal     slope  slope_se  \n",
       "6414      0.059205  1.393160e-08 -0.168762  0.029260  \n",
       "14611     0.124116  2.197750e-05 -0.246023  0.057281  \n",
       "45230     0.087221  8.491350e-08  0.231096  0.041304  \n",
       "135656    0.111990  3.281960e-30 -0.726153  0.057266  \n",
       "137724    0.068668  1.462140e-26 -0.370315  0.031672  \n",
       "137946    0.068668  2.014910e-57  0.706155  0.035862  \n",
       "138621    0.145261  4.482050e-06 -0.293079  0.062803  \n",
       "138995    0.072364  2.765010e-20 -0.319694  0.032375  \n",
       "142471    0.068855  9.495520e-22 -0.323697  0.031398  \n",
       "143690    0.076638  1.176010e-12 -0.331468  0.044801  \n",
       "143783    0.093415  1.509120e-67  0.972900  0.043631  \n",
       "143823    0.132593  8.545960e-22 -0.693095  0.067157  \n",
       "263405    0.056303  5.746020e-65 -0.613437  0.031989  \n",
       "338770    0.104690  4.896640e-51 -1.193670  0.056659  \n",
       "\n",
       "[14 rows x 22 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_sig[(gtex_sig['ensembl_gene_id'].isin(bs.gene_id))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.646814Z",
     "iopub.status.busy": "2021-09-14T13:25:27.645566Z",
     "iopub.status.idle": "2021-09-14T13:25:27.651940Z",
     "shell.execute_reply": "2021-09-14T13:25:27.650861Z"
    }
   },
   "outputs": [],
   "source": [
    "gtex_sig[(gtex_sig['ensembl_gene_id'].isin(bs.gene_id))]\\\n",
    "    .to_csv(\"siEQTL_gtex_comparison.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### eigenMT corrected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.659669Z",
     "iopub.status.busy": "2021-09-14T13:25:27.658417Z",
     "iopub.status.idle": "2021-09-14T13:25:27.679014Z",
     "shell.execute_reply": "2021-09-14T13:25:27.680063Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>chr4:42392391:T:G</td>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>-0.291917</td>\n",
       "      <td>-13.197929</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>0.362373</td>\n",
       "      <td>502</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000124406</td>\n",
       "      <td>ATP8A1</td>\n",
       "      <td>10396.0</td>\n",
       "      <td>ATPase phospholipid transporting 8A1 [Source:H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4271</th>\n",
       "      <td>chr8:11603602:A:G</td>\n",
       "      <td>ENSG00000154328.15</td>\n",
       "      <td>ENSG00000154328.15</td>\n",
       "      <td>0.380078</td>\n",
       "      <td>14.183350</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002130</td>\n",
       "      <td>0.458802</td>\n",
       "      <td>888</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000154328</td>\n",
       "      <td>NEIL2</td>\n",
       "      <td>252969.0</td>\n",
       "      <td>nei like DNA glycosylase 2 [Source:HGNC Symbol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>chr1:89159687:G:A</td>\n",
       "      <td>ENSG00000117226.11</td>\n",
       "      <td>ENSG00000117226.11</td>\n",
       "      <td>0.431197</td>\n",
       "      <td>15.848961</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.002374</td>\n",
       "      <td>0.362373</td>\n",
       "      <td>266</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000117226</td>\n",
       "      <td>GBP3</td>\n",
       "      <td>2635.0</td>\n",
       "      <td>guanylate binding protein 3 [Source:HGNC Symbo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2556</th>\n",
       "      <td>chr8:63463308:T:C</td>\n",
       "      <td>ENSG00000137563.11</td>\n",
       "      <td>ENSG00000137563.11</td>\n",
       "      <td>0.527946</td>\n",
       "      <td>10.192544</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.002460</td>\n",
       "      <td>0.438136</td>\n",
       "      <td>335</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000137563</td>\n",
       "      <td>GGH</td>\n",
       "      <td>8836.0</td>\n",
       "      <td>gamma-glutamyl hydrolase [Source:HGNC Symbol;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3627</th>\n",
       "      <td>chr22:25059120:A:C</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>0.626299</td>\n",
       "      <td>16.616479</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.002482</td>\n",
       "      <td>0.438136</td>\n",
       "      <td>744</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000272977</td>\n",
       "      <td>AL008721.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              variant_id             gene_id           gencodeID     slope  \\\n",
       "473    chr4:42392391:T:G  ENSG00000124406.16  ENSG00000124406.16 -0.291917   \n",
       "4271   chr8:11603602:A:G  ENSG00000154328.15  ENSG00000154328.15  0.380078   \n",
       "406    chr1:89159687:G:A  ENSG00000117226.11  ENSG00000117226.11  0.431197   \n",
       "2556   chr8:63463308:T:C  ENSG00000137563.11  ENSG00000137563.11  0.527946   \n",
       "3627  chr22:25059120:A:C   ENSG00000272977.1   ENSG00000272977.1  0.626299   \n",
       "\n",
       "      statistic  pval_nominal        BF  eigenMT_BH  TESTS  Type  \\\n",
       "473  -13.197929      0.000004  0.002091    0.362373    502  Gene   \n",
       "4271  14.183350      0.000002  0.002130    0.458802    888  Gene   \n",
       "406   15.848961      0.000009  0.002374    0.362373    266  Gene   \n",
       "2556  10.192544      0.000007  0.002460    0.438136    335  Gene   \n",
       "3627  16.616479      0.000003  0.002482    0.438136    744  Gene   \n",
       "\n",
       "      ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "473   ENSG00000124406             ATP8A1     10396.0   \n",
       "4271  ENSG00000154328              NEIL2    252969.0   \n",
       "406   ENSG00000117226               GBP3      2635.0   \n",
       "2556  ENSG00000137563                GGH      8836.0   \n",
       "3627  ENSG00000272977         AL008721.2         NaN   \n",
       "\n",
       "                                            description  \n",
       "473   ATPase phospholipid transporting 8A1 [Source:H...  \n",
       "4271  nei like DNA glycosylase 2 [Source:HGNC Symbol...  \n",
       "406   guanylate binding protein 3 [Source:HGNC Symbo...  \n",
       "2556  gamma-glutamyl hydrolase [Source:HGNC Symbol;A...  \n",
       "3627                                                NaN  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap = dft[(dft['gene_id'].isin(gtex_sig.ensembl_gene_id)) & (dft[\"BF\"] < 0.05)].drop_duplicates()\n",
    "print(gtex_overlap.shape)\n",
    "gtex_overlap.sort_values(\"BF\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.687968Z",
     "iopub.status.busy": "2021-09-14T13:25:27.686747Z",
     "iopub.status.idle": "2021-09-14T13:25:27.705678Z",
     "shell.execute_reply": "2021-09-14T13:25:27.706736Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>slope</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>BF</th>\n",
       "      <th>eigenMT_BH</th>\n",
       "      <th>TESTS</th>\n",
       "      <th>Type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>chr4:42392391:T:G</td>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>ENSG00000124406.16</td>\n",
       "      <td>-0.291917</td>\n",
       "      <td>-13.197929</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>0.362373</td>\n",
       "      <td>502</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000124406</td>\n",
       "      <td>ATP8A1</td>\n",
       "      <td>10396.0</td>\n",
       "      <td>ATPase phospholipid transporting 8A1 [Source:H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4271</th>\n",
       "      <td>chr8:11603602:A:G</td>\n",
       "      <td>ENSG00000154328.15</td>\n",
       "      <td>ENSG00000154328.15</td>\n",
       "      <td>0.380078</td>\n",
       "      <td>14.183350</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002130</td>\n",
       "      <td>0.458802</td>\n",
       "      <td>888</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000154328</td>\n",
       "      <td>NEIL2</td>\n",
       "      <td>252969.0</td>\n",
       "      <td>nei like DNA glycosylase 2 [Source:HGNC Symbol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>chr1:89159687:G:A</td>\n",
       "      <td>ENSG00000117226.11</td>\n",
       "      <td>ENSG00000117226.11</td>\n",
       "      <td>0.431197</td>\n",
       "      <td>15.848961</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.002374</td>\n",
       "      <td>0.362373</td>\n",
       "      <td>266</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000117226</td>\n",
       "      <td>GBP3</td>\n",
       "      <td>2635.0</td>\n",
       "      <td>guanylate binding protein 3 [Source:HGNC Symbo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2556</th>\n",
       "      <td>chr8:63463308:T:C</td>\n",
       "      <td>ENSG00000137563.11</td>\n",
       "      <td>ENSG00000137563.11</td>\n",
       "      <td>0.527946</td>\n",
       "      <td>10.192544</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.002460</td>\n",
       "      <td>0.438136</td>\n",
       "      <td>335</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000137563</td>\n",
       "      <td>GGH</td>\n",
       "      <td>8836.0</td>\n",
       "      <td>gamma-glutamyl hydrolase [Source:HGNC Symbol;A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3627</th>\n",
       "      <td>chr22:25059120:A:C</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>0.626299</td>\n",
       "      <td>16.616479</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.002482</td>\n",
       "      <td>0.438136</td>\n",
       "      <td>744</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000272977</td>\n",
       "      <td>AL008721.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              variant_id             gene_id           gencodeID     slope  \\\n",
       "473    chr4:42392391:T:G  ENSG00000124406.16  ENSG00000124406.16 -0.291917   \n",
       "4271   chr8:11603602:A:G  ENSG00000154328.15  ENSG00000154328.15  0.380078   \n",
       "406    chr1:89159687:G:A  ENSG00000117226.11  ENSG00000117226.11  0.431197   \n",
       "2556   chr8:63463308:T:C  ENSG00000137563.11  ENSG00000137563.11  0.527946   \n",
       "3627  chr22:25059120:A:C   ENSG00000272977.1   ENSG00000272977.1  0.626299   \n",
       "\n",
       "      statistic  pval_nominal        BF  eigenMT_BH  TESTS  Type  \\\n",
       "473  -13.197929      0.000004  0.002091    0.362373    502  Gene   \n",
       "4271  14.183350      0.000002  0.002130    0.458802    888  Gene   \n",
       "406   15.848961      0.000009  0.002374    0.362373    266  Gene   \n",
       "2556  10.192544      0.000007  0.002460    0.438136    335  Gene   \n",
       "3627  16.616479      0.000003  0.002482    0.438136    744  Gene   \n",
       "\n",
       "      ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "473   ENSG00000124406             ATP8A1     10396.0   \n",
       "4271  ENSG00000154328              NEIL2    252969.0   \n",
       "406   ENSG00000117226               GBP3      2635.0   \n",
       "2556  ENSG00000137563                GGH      8836.0   \n",
       "3627  ENSG00000272977         AL008721.2         NaN   \n",
       "\n",
       "                                            description  \n",
       "473   ATPase phospholipid transporting 8A1 [Source:H...  \n",
       "4271  nei like DNA glycosylase 2 [Source:HGNC Symbol...  \n",
       "406   guanylate binding protein 3 [Source:HGNC Symbo...  \n",
       "2556  gamma-glutamyl hydrolase [Source:HGNC Symbol;A...  \n",
       "3627                                                NaN  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap = dft[(dft['gene_id'].isin(gtex_sig.ensembl_gene_id))].drop_duplicates()\n",
    "print(gtex_overlap.shape)\n",
    "gtex_overlap.sort_values(\"BF\").head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-14T13:25:27.712678Z",
     "iopub.status.busy": "2021-09-14T13:25:27.711428Z",
     "iopub.status.idle": "2021-09-14T13:25:27.716665Z",
     "shell.execute_reply": "2021-09-14T13:25:27.715554Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.1010237589337455"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap.shape[0]/dft.shape[0] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
