{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examine overlaps with published data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:25.851084Z",
     "iopub.status.busy": "2023-06-08T13:10:25.849129Z",
     "iopub.status.idle": "2023-06-08T13:10:32.916655Z",
     "shell.execute_reply": "2023-06-08T13:10:32.914765Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import session_info\n",
    "from pyhere import here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Public si-eQTL analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:32.927501Z",
     "iopub.status.busy": "2023-06-08T13:10:32.925882Z",
     "iopub.status.idle": "2023-06-08T13:10:32.956387Z",
     "shell.execute_reply": "2023-06-08T13:10:32.955059Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shen et al.:\n",
      "21\n",
      "Kukurba et al.:\n",
      "6\n",
      "Yao et al.:\n",
      "14\n",
      "Total of Yao + Kukurba:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shen = [\"GDAP2\", \"AIM2\", \"SLAMF6\", \"RLF\", \"ATG4C\", \"FUT7\",\n",
    "        \"TMEM218\", \"C11orf74\", \"RAB35\", \"TMEM5\", \"HNRNPK\",\n",
    "        \"CDCA3\", \"ERCC5\", \"GJB6\", \"SNTB2\", \"SPNS3\", \n",
    "        \"XAF1\", \"RBBP8\", \"RUFY4\", \"CA2\", \"RAPGEF1\"]\n",
    "print(\"Shen et al.:\")\n",
    "print(len(shen))\n",
    "\n",
    "kukurba = [\"NOD2\", \"WDR36\", \"BSCL2\", \"MAP7D3\", \"RHOXF1\", \"DNAH1\"]\n",
    "print(\"Kukurba et al.:\")\n",
    "print(len(kukurba))\n",
    "\n",
    "yao = [\"NOD2\", \"HLA-DRB5\", \"HLA-DRB5\", \"KIAA0586\", \"PPP2R5A\", \n",
    "       \"TSNAXIP1\", \"MUT\", \"GRIK2\", \"C15orf37\", \"LIMA1\", \"IL6ST\", \n",
    "       \"HCG8\", \"BLOC1S3\", \"NKX3-1\", \"CXorf23\"]\n",
    "print(\"Yao et al.:\")\n",
    "print(len(np.unique(yao)))\n",
    "print(\"Total of Yao + Kukurba:\")\n",
    "len(set(yao) | set(kukurba))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load BrainSeq si-eQTL results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interacting variant-gene pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:33.039849Z",
     "iopub.status.busy": "2023-06-08T13:10:33.039572Z",
     "iopub.status.idle": "2023-06-08T13:10:33.961384Z",
     "shell.execute_reply": "2023-06-08T13:10:33.959453Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(692, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8285</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000270605.1</td>\n",
       "      <td>chr1:28102893:G:C</td>\n",
       "      <td>ENSG00000270605.1</td>\n",
       "      <td>ENSG00000270605</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28239509</td>\n",
       "      <td>28241453</td>\n",
       "      <td>0.049873</td>\n",
       "      <td>-0.261352</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000270605</td>\n",
       "      <td>AL353622.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8286</th>\n",
       "      <td>DLPFC</td>\n",
       "      <td>ENSG00000187498.16</td>\n",
       "      <td>chr13:109650494:C:T</td>\n",
       "      <td>ENSG00000187498.16</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>chr13</td>\n",
       "      <td>110148963</td>\n",
       "      <td>110307157</td>\n",
       "      <td>0.048863</td>\n",
       "      <td>0.193807</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000187498</td>\n",
       "      <td>COL4A1</td>\n",
       "      <td>1282.0</td>\n",
       "      <td>collagen type IV alpha 1 chain [Source:HGNC Sy...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       region             gene_id           variant_id          gencode_id  \\\n",
       "8285  Caudate   ENSG00000270605.1    chr1:28102893:G:C   ENSG00000270605.1   \n",
       "8286    DLPFC  ENSG00000187498.16  chr13:109650494:C:T  ENSG00000187498.16   \n",
       "\n",
       "            gene_name seqnames      start        end      lfsr  \\\n",
       "8285  ENSG00000270605     chr1   28239509   28241453  0.049873   \n",
       "8286           COL4A1    chr13  110148963  110307157  0.048863   \n",
       "\n",
       "      posterior_mean feature_type  ensembl_gene_id external_gene_name  \\\n",
       "8285       -0.261352         Gene  ENSG00000270605         AL353622.1   \n",
       "8286        0.193807         Gene  ENSG00000187498             COL4A1   \n",
       "\n",
       "      entrezgene                                        description  \n",
       "8285         NaN                                                NaN  \n",
       "8286      1282.0  collagen type IV alpha 1 chain [Source:HGNC Sy...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs0 = pd.read_csv(\"../../_m/BrainSeq_sexGenotypes_4features_3regions.txt.gz\", sep='\\t')\n",
    "bs0[\"ensembl_gene_id\"] = bs0.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "biomart = pd.read_csv(\"../_h/biomart.csv\", index_col=0)\n",
    "bs = bs0.merge(biomart, on=\"ensembl_gene_id\").drop_duplicates(subset=\"gene_id\")\n",
    "print(bs.shape)\n",
    "bs.tail(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:33.968215Z",
     "iopub.status.busy": "2023-06-08T13:10:33.967765Z",
     "iopub.status.idle": "2023-06-08T13:10:34.008060Z",
     "shell.execute_reply": "2023-06-08T13:10:34.006276Z"
    }
   },
   "outputs": [],
   "source": [
    "bs[(bs['external_gene_name'].isin(shen))].to_csv(\"siEQTL_Shen_comparison.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:34.017412Z",
     "iopub.status.busy": "2023-06-08T13:10:34.016733Z",
     "iopub.status.idle": "2023-06-08T13:10:34.038060Z",
     "shell.execute_reply": "2023-06-08T13:10:34.035703Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [region, gene_id, variant_id, gencode_id, gene_name, seqnames, start, end, lfsr, posterior_mean, feature_type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(kukurba))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:34.045112Z",
     "iopub.status.busy": "2023-06-08T13:10:34.044086Z",
     "iopub.status.idle": "2023-06-08T13:10:34.058441Z",
     "shell.execute_reply": "2023-06-08T13:10:34.057410Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [region, gene_id, variant_id, gencode_id, gene_name, seqnames, start, end, lfsr, posterior_mean, feature_type, ensembl_gene_id, external_gene_name, entrezgene, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(yao))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:34.063827Z",
     "iopub.status.busy": "2023-06-08T13:10:34.063336Z",
     "iopub.status.idle": "2023-06-08T13:10:34.089981Z",
     "shell.execute_reply": "2023-06-08T13:10:34.089154Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5787</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000125703.15</td>\n",
       "      <td>chr1:63060301:G:A</td>\n",
       "      <td>ENSG00000125703.15</td>\n",
       "      <td>ATG4C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>62784132</td>\n",
       "      <td>62865516</td>\n",
       "      <td>0.019839</td>\n",
       "      <td>0.094136</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000125703</td>\n",
       "      <td>ATG4C</td>\n",
       "      <td>84938.0</td>\n",
       "      <td>autophagy related 4C cysteine peptidase [Sourc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8130</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000104267.10</td>\n",
       "      <td>chr8:84966439:A:T</td>\n",
       "      <td>ENSG00000104267.10</td>\n",
       "      <td>CA2</td>\n",
       "      <td>chr8</td>\n",
       "      <td>85463968</td>\n",
       "      <td>85481493</td>\n",
       "      <td>0.045578</td>\n",
       "      <td>0.289557</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000104267</td>\n",
       "      <td>CA2</td>\n",
       "      <td>760.0</td>\n",
       "      <td>carbonic anhydrase 2 [Source:HGNC Symbol;Acc:H...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       region             gene_id         variant_id          gencode_id  \\\n",
       "5787  Caudate  ENSG00000125703.15  chr1:63060301:G:A  ENSG00000125703.15   \n",
       "8130  Caudate  ENSG00000104267.10  chr8:84966439:A:T  ENSG00000104267.10   \n",
       "\n",
       "     gene_name seqnames     start       end      lfsr  posterior_mean  \\\n",
       "5787     ATG4C     chr1  62784132  62865516  0.019839        0.094136   \n",
       "8130       CA2     chr8  85463968  85481493  0.045578        0.289557   \n",
       "\n",
       "     feature_type  ensembl_gene_id external_gene_name  entrezgene  \\\n",
       "5787         Gene  ENSG00000125703              ATG4C     84938.0   \n",
       "8130         Gene  ENSG00000104267                CA2       760.0   \n",
       "\n",
       "                                            description  \n",
       "5787  autophagy related 4C cysteine peptidase [Sourc...  \n",
       "8130  carbonic anhydrase 2 [Source:HGNC Symbol;Acc:H...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs[(bs['external_gene_name'].isin(shen+kukurba+yao))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GTEx comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:34.095990Z",
     "iopub.status.busy": "2023-06-08T13:10:34.095070Z",
     "iopub.status.idle": "2023-06-08T13:10:35.843739Z",
     "shell.execute_reply": "2023-06-08T13:10:35.842956Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal_sb</th>\n",
       "      <th>slope_sb</th>\n",
       "      <th>slope_se_sb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000241860.6</td>\n",
       "      <td>RP11-34P13.13</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>chr1_14677_G_A_b38</td>\n",
       "      <td>rs201327123</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.051635</td>\n",
       "      <td>0.847114</td>\n",
       "      <td>0.055080</td>\n",
       "      <td>0.285537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>chr1_64764_C_T_b38</td>\n",
       "      <td>rs769952832</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.061102</td>\n",
       "      <td>0.316881</td>\n",
       "      <td>0.222928</td>\n",
       "      <td>0.222511</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ensembl_gene_id   hugo_gene_id               gene_type  \\\n",
       "0  ENSG00000241860.6  RP11-34P13.13    processed_transcript   \n",
       "1  ENSG00000227232.5         WASH7P  unprocessed_pseudogene   \n",
       "\n",
       "           variant_id        rs_id                Tissue       maf  \\\n",
       "0  chr1_14677_G_A_b38  rs201327123  Adipose_Subcutaneous  0.051635   \n",
       "1  chr1_64764_C_T_b38  rs769952832  Adipose_Subcutaneous  0.061102   \n",
       "\n",
       "   pval_nominal_sb  slope_sb  slope_se_sb  \n",
       "0         0.847114  0.055080     0.285537  \n",
       "1         0.316881  0.222928     0.222511  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex = pd.read_csv(here(\"input/public_results/gtex_results/_m\",\n",
    "                        \"GTEx_Analysis_v8_sbeQTLs/GTEx_Analysis_v8_sbeQTLs.txt\"), \n",
    "                   sep='\\t')\n",
    "gtex.iloc[0:2, 0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:35.846798Z",
     "iopub.status.busy": "2023-06-08T13:10:35.846544Z",
     "iopub.status.idle": "2023-06-08T13:10:35.853533Z",
     "shell.execute_reply": "2023-06-08T13:10:35.852920Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numtested</th>\n",
       "      <th>pvals.corrected</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.847114</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.022302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.316881</td>\n",
       "      <td>0.981254</td>\n",
       "      <td>0.003978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   numtested  pvals.corrected      qval  pval_nominal_f\n",
       "0          1         0.847114  1.000000        0.022302\n",
       "1          1         0.316881  0.981254        0.003978"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex.iloc[0:2, 10:14]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:35.856800Z",
     "iopub.status.busy": "2023-06-08T13:10:35.856613Z",
     "iopub.status.idle": "2023-06-08T13:10:35.976919Z",
     "shell.execute_reply": "2023-06-08T13:10:35.976272Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>pvals.corrected</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62155</th>\n",
       "      <td>ENSG00000026025.15</td>\n",
       "      <td>VIM</td>\n",
       "      <td>Brain_Amygdala</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.012836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116842</th>\n",
       "      <td>ENSG00000160818.16</td>\n",
       "      <td>GPATCH4</td>\n",
       "      <td>Brain_Nucleus_accumbens_basal_ganglia</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121904</th>\n",
       "      <td>ENSG00000141562.17</td>\n",
       "      <td>NARF</td>\n",
       "      <td>Brain_Nucleus_accumbens_basal_ganglia</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.198445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122123</th>\n",
       "      <td>ENSG00000267174.5</td>\n",
       "      <td>CTC-510F12.4</td>\n",
       "      <td>Brain_Nucleus_accumbens_basal_ganglia</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.198445</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ensembl_gene_id  hugo_gene_id  \\\n",
       "62155   ENSG00000026025.15           VIM   \n",
       "116842  ENSG00000160818.16       GPATCH4   \n",
       "121904  ENSG00000141562.17          NARF   \n",
       "122123   ENSG00000267174.5  CTC-510F12.4   \n",
       "\n",
       "                                       Tissue  pvals.corrected      qval  \n",
       "62155                          Brain_Amygdala         0.000004  0.012836  \n",
       "116842  Brain_Nucleus_accumbens_basal_ganglia         0.000088  0.198445  \n",
       "121904  Brain_Nucleus_accumbens_basal_ganglia         0.000056  0.198445  \n",
       "122123  Brain_Nucleus_accumbens_basal_ganglia         0.000083  0.198445  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## qval threshold equal to number of published sb-eQTL\n",
    "gtex[(gtex['qval'] < 0.25) & (gtex[\"Tissue\"].str.contains(\"Brain\"))]\\\n",
    "    .loc[:, [\"ensembl_gene_id\", \"hugo_gene_id\", \"Tissue\", \"pvals.corrected\", 'qval']].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:35.979703Z",
     "iopub.status.busy": "2023-06-08T13:10:35.979461Z",
     "iopub.status.idle": "2023-06-08T13:10:36.083594Z",
     "shell.execute_reply": "2023-06-08T13:10:36.082933Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>pvals.corrected</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>362961</th>\n",
       "      <td>ENSG00000221571.3</td>\n",
       "      <td>RNU6ATAC35P</td>\n",
       "      <td>Whole_Blood</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.139762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365043</th>\n",
       "      <td>ENSG00000196743.8</td>\n",
       "      <td>GM2A</td>\n",
       "      <td>Whole_Blood</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.116825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367164</th>\n",
       "      <td>ENSG00000148459.15</td>\n",
       "      <td>PDSS1</td>\n",
       "      <td>Whole_Blood</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.139762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ensembl_gene_id hugo_gene_id       Tissue  pvals.corrected  \\\n",
       "362961   ENSG00000221571.3  RNU6ATAC35P  Whole_Blood         0.000039   \n",
       "365043   ENSG00000196743.8         GM2A  Whole_Blood         0.000011   \n",
       "367164  ENSG00000148459.15        PDSS1  Whole_Blood         0.000027   \n",
       "\n",
       "            qval  \n",
       "362961  0.139762  \n",
       "365043  0.116825  \n",
       "367164  0.139762  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## qval threshold equal to number of published sb-eQTL\n",
    "gtex[(gtex['qval'] < 0.25) & (gtex[\"Tissue\"].str.contains(\"Whole\"))]\\\n",
    "    .loc[:, [\"ensembl_gene_id\", \"hugo_gene_id\", \"Tissue\", \"pvals.corrected\", 'qval']].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.086390Z",
     "iopub.status.busy": "2023-06-08T13:10:36.086147Z",
     "iopub.status.idle": "2023-06-08T13:10:36.091689Z",
     "shell.execute_reply": "2023-06-08T13:10:36.091085Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(369, 22)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_sig = gtex[(gtex['qval'] < 0.25)]\n",
    "gtex_sig.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.094218Z",
     "iopub.status.busy": "2023-06-08T13:10:36.093984Z",
     "iopub.status.idle": "2023-06-08T13:10:36.112567Z",
     "shell.execute_reply": "2023-06-08T13:10:36.111949Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal_sb</th>\n",
       "      <th>slope_sb</th>\n",
       "      <th>slope_se_sb</th>\n",
       "      <th>...</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_f</th>\n",
       "      <th>slope_f</th>\n",
       "      <th>slope_se_f</th>\n",
       "      <th>pval_nominal_m</th>\n",
       "      <th>slope_m</th>\n",
       "      <th>slope_se_m</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1096</th>\n",
       "      <td>ENSG00000076356.6</td>\n",
       "      <td>PLXNA2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_208030492_G_A_b38</td>\n",
       "      <td>rs3811383</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.123924</td>\n",
       "      <td>5.391600e-05</td>\n",
       "      <td>0.338278</td>\n",
       "      <td>0.083064</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121068</td>\n",
       "      <td>1.718880e-08</td>\n",
       "      <td>0.456729</td>\n",
       "      <td>0.075705</td>\n",
       "      <td>9.155700e-01</td>\n",
       "      <td>0.009739</td>\n",
       "      <td>0.091682</td>\n",
       "      <td>2.747400e-05</td>\n",
       "      <td>0.171830</td>\n",
       "      <td>0.040604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5262</th>\n",
       "      <td>ENSG00000170632.13</td>\n",
       "      <td>ARMC10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7_103076937_C_T_b38</td>\n",
       "      <td>rs6958836</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.169535</td>\n",
       "      <td>5.011130e-05</td>\n",
       "      <td>0.357403</td>\n",
       "      <td>0.087384</td>\n",
       "      <td>...</td>\n",
       "      <td>0.192900</td>\n",
       "      <td>4.933240e-01</td>\n",
       "      <td>-0.054539</td>\n",
       "      <td>0.079379</td>\n",
       "      <td>3.219220e-07</td>\n",
       "      <td>-0.429800</td>\n",
       "      <td>0.079545</td>\n",
       "      <td>8.797530e-08</td>\n",
       "      <td>-0.216374</td>\n",
       "      <td>0.039857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5644</th>\n",
       "      <td>ENSG00000120907.17</td>\n",
       "      <td>ADRA1A</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr8_26839198_G_A_b38</td>\n",
       "      <td>rs117380715</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.216867</td>\n",
       "      <td>1.045890e-05</td>\n",
       "      <td>-0.323552</td>\n",
       "      <td>0.072676</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084548</td>\n",
       "      <td>4.637410e-18</td>\n",
       "      <td>-0.779707</td>\n",
       "      <td>0.076596</td>\n",
       "      <td>3.976660e-10</td>\n",
       "      <td>-0.469672</td>\n",
       "      <td>0.069091</td>\n",
       "      <td>5.637370e-52</td>\n",
       "      <td>-0.568916</td>\n",
       "      <td>0.033334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6414</th>\n",
       "      <td>ENSG00000136830.11</td>\n",
       "      <td>FAM129B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr9_127584339_G_A_b38</td>\n",
       "      <td>rs10739693</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.304647</td>\n",
       "      <td>7.387010e-07</td>\n",
       "      <td>-0.283660</td>\n",
       "      <td>0.056579</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004976</td>\n",
       "      <td>1.978000e-06</td>\n",
       "      <td>-0.333315</td>\n",
       "      <td>0.066772</td>\n",
       "      <td>1.653380e-01</td>\n",
       "      <td>-0.082625</td>\n",
       "      <td>0.059205</td>\n",
       "      <td>1.393160e-08</td>\n",
       "      <td>-0.168762</td>\n",
       "      <td>0.029260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7220</th>\n",
       "      <td>ENSG00000166787.3</td>\n",
       "      <td>SAA3P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr11_18269355_T_C_b38</td>\n",
       "      <td>rs34068567</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.278830</td>\n",
       "      <td>2.207290e-05</td>\n",
       "      <td>0.323030</td>\n",
       "      <td>0.075427</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074347</td>\n",
       "      <td>6.409400e-08</td>\n",
       "      <td>0.453034</td>\n",
       "      <td>0.078725</td>\n",
       "      <td>3.138000e-01</td>\n",
       "      <td>0.063002</td>\n",
       "      <td>0.062292</td>\n",
       "      <td>2.433600e-08</td>\n",
       "      <td>0.211910</td>\n",
       "      <td>0.037395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8540</th>\n",
       "      <td>ENSG00000183463.5</td>\n",
       "      <td>URAD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr13_27990205_T_A_b38</td>\n",
       "      <td>rs7335293</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>9.078700e-09</td>\n",
       "      <td>-0.444892</td>\n",
       "      <td>0.076123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>9.982650e-21</td>\n",
       "      <td>-0.887723</td>\n",
       "      <td>0.078738</td>\n",
       "      <td>1.892290e-09</td>\n",
       "      <td>-0.457733</td>\n",
       "      <td>0.070571</td>\n",
       "      <td>3.077310e-53</td>\n",
       "      <td>-0.640604</td>\n",
       "      <td>0.036976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9191</th>\n",
       "      <td>ENSG00000282651.2</td>\n",
       "      <td>IGHV5-10-1</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>chr14_106114510_A_G_b38</td>\n",
       "      <td>rs4573838</td>\n",
       "      <td>Adipose_Subcutaneous</td>\n",
       "      <td>0.419105</td>\n",
       "      <td>2.025150e-05</td>\n",
       "      <td>-0.406760</td>\n",
       "      <td>0.094541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074347</td>\n",
       "      <td>5.248710e-12</td>\n",
       "      <td>-0.682629</td>\n",
       "      <td>0.089412</td>\n",
       "      <td>2.805060e-03</td>\n",
       "      <td>-0.289091</td>\n",
       "      <td>0.094806</td>\n",
       "      <td>3.458420e-21</td>\n",
       "      <td>-0.445408</td>\n",
       "      <td>0.045073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14611</th>\n",
       "      <td>ENSG00000143933.16</td>\n",
       "      <td>CALM2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2_46225349_C_T_b38</td>\n",
       "      <td>rs12477148</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.072495</td>\n",
       "      <td>4.497930e-05</td>\n",
       "      <td>-0.480557</td>\n",
       "      <td>0.116471</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161955</td>\n",
       "      <td>4.747150e-04</td>\n",
       "      <td>-0.491287</td>\n",
       "      <td>0.134732</td>\n",
       "      <td>9.165740e-01</td>\n",
       "      <td>0.013043</td>\n",
       "      <td>0.124116</td>\n",
       "      <td>2.197750e-05</td>\n",
       "      <td>-0.246023</td>\n",
       "      <td>0.057281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15082</th>\n",
       "      <td>ENSG00000144410.4</td>\n",
       "      <td>CPO</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2_206822186_C_T_b38</td>\n",
       "      <td>rs12470278</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.097015</td>\n",
       "      <td>3.204120e-05</td>\n",
       "      <td>0.682291</td>\n",
       "      <td>0.162191</td>\n",
       "      <td>...</td>\n",
       "      <td>0.115370</td>\n",
       "      <td>1.165430e-01</td>\n",
       "      <td>0.280837</td>\n",
       "      <td>0.176978</td>\n",
       "      <td>4.431060e-06</td>\n",
       "      <td>-0.558002</td>\n",
       "      <td>0.113158</td>\n",
       "      <td>7.896400e-06</td>\n",
       "      <td>-0.320288</td>\n",
       "      <td>0.070745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17452</th>\n",
       "      <td>ENSG00000211698.2</td>\n",
       "      <td>TRGV4</td>\n",
       "      <td>TR_V_gene</td>\n",
       "      <td>chr7_38361995_A_C_b38</td>\n",
       "      <td>rs10233345</td>\n",
       "      <td>Adipose_Visceral_Omentum</td>\n",
       "      <td>0.335821</td>\n",
       "      <td>6.438100e-05</td>\n",
       "      <td>0.427491</td>\n",
       "      <td>0.105837</td>\n",
       "      <td>...</td>\n",
       "      <td>0.139089</td>\n",
       "      <td>7.011230e-06</td>\n",
       "      <td>-0.481758</td>\n",
       "      <td>0.100091</td>\n",
       "      <td>8.857670e-15</td>\n",
       "      <td>-1.068840</td>\n",
       "      <td>0.112111</td>\n",
       "      <td>1.630390e-49</td>\n",
       "      <td>-0.838766</td>\n",
       "      <td>0.049090</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_gene_id hugo_gene_id                           gene_type  \\\n",
       "1096    ENSG00000076356.6       PLXNA2                      protein_coding   \n",
       "5262   ENSG00000170632.13       ARMC10                      protein_coding   \n",
       "5644   ENSG00000120907.17       ADRA1A                      protein_coding   \n",
       "6414   ENSG00000136830.11      FAM129B                      protein_coding   \n",
       "7220    ENSG00000166787.3        SAA3P  transcribed_unprocessed_pseudogene   \n",
       "8540    ENSG00000183463.5         URAD                      protein_coding   \n",
       "9191    ENSG00000282651.2   IGHV5-10-1                           IG_V_gene   \n",
       "14611  ENSG00000143933.16        CALM2                      protein_coding   \n",
       "15082   ENSG00000144410.4          CPO                      protein_coding   \n",
       "17452   ENSG00000211698.2        TRGV4                           TR_V_gene   \n",
       "\n",
       "                    variant_id        rs_id                    Tissue  \\\n",
       "1096    chr1_208030492_G_A_b38    rs3811383      Adipose_Subcutaneous   \n",
       "5262    chr7_103076937_C_T_b38    rs6958836      Adipose_Subcutaneous   \n",
       "5644     chr8_26839198_G_A_b38  rs117380715      Adipose_Subcutaneous   \n",
       "6414    chr9_127584339_G_A_b38   rs10739693      Adipose_Subcutaneous   \n",
       "7220    chr11_18269355_T_C_b38   rs34068567      Adipose_Subcutaneous   \n",
       "8540    chr13_27990205_T_A_b38    rs7335293      Adipose_Subcutaneous   \n",
       "9191   chr14_106114510_A_G_b38    rs4573838      Adipose_Subcutaneous   \n",
       "14611    chr2_46225349_C_T_b38   rs12477148  Adipose_Visceral_Omentum   \n",
       "15082   chr2_206822186_C_T_b38   rs12470278  Adipose_Visceral_Omentum   \n",
       "17452    chr7_38361995_A_C_b38   rs10233345  Adipose_Visceral_Omentum   \n",
       "\n",
       "            maf  pval_nominal_sb  slope_sb  slope_se_sb  ...      qval  \\\n",
       "1096   0.123924     5.391600e-05  0.338278     0.083064  ...  0.121068   \n",
       "5262   0.169535     5.011130e-05  0.357403     0.087384  ...  0.192900   \n",
       "5644   0.216867     1.045890e-05 -0.323552     0.072676  ...  0.084548   \n",
       "6414   0.304647     7.387010e-07 -0.283660     0.056579  ...  0.004976   \n",
       "7220   0.278830     2.207290e-05  0.323030     0.075427  ...  0.074347   \n",
       "8540   0.500000     9.078700e-09 -0.444892     0.076123  ...  0.000122   \n",
       "9191   0.419105     2.025150e-05 -0.406760     0.094541  ...  0.074347   \n",
       "14611  0.072495     4.497930e-05 -0.480557     0.116471  ...  0.161955   \n",
       "15082  0.097015     3.204120e-05  0.682291     0.162191  ...  0.115370   \n",
       "17452  0.335821     6.438100e-05  0.427491     0.105837  ...  0.139089   \n",
       "\n",
       "       pval_nominal_f   slope_f  slope_se_f  pval_nominal_m   slope_m  \\\n",
       "1096     1.718880e-08  0.456729    0.075705    9.155700e-01  0.009739   \n",
       "5262     4.933240e-01 -0.054539    0.079379    3.219220e-07 -0.429800   \n",
       "5644     4.637410e-18 -0.779707    0.076596    3.976660e-10 -0.469672   \n",
       "6414     1.978000e-06 -0.333315    0.066772    1.653380e-01 -0.082625   \n",
       "7220     6.409400e-08  0.453034    0.078725    3.138000e-01  0.063002   \n",
       "8540     9.982650e-21 -0.887723    0.078738    1.892290e-09 -0.457733   \n",
       "9191     5.248710e-12 -0.682629    0.089412    2.805060e-03 -0.289091   \n",
       "14611    4.747150e-04 -0.491287    0.134732    9.165740e-01  0.013043   \n",
       "15082    1.165430e-01  0.280837    0.176978    4.431060e-06 -0.558002   \n",
       "17452    7.011230e-06 -0.481758    0.100091    8.857670e-15 -1.068840   \n",
       "\n",
       "       slope_se_m  pval_nominal     slope  slope_se  \n",
       "1096     0.091682  2.747400e-05  0.171830  0.040604  \n",
       "5262     0.079545  8.797530e-08 -0.216374  0.039857  \n",
       "5644     0.069091  5.637370e-52 -0.568916  0.033334  \n",
       "6414     0.059205  1.393160e-08 -0.168762  0.029260  \n",
       "7220     0.062292  2.433600e-08  0.211910  0.037395  \n",
       "8540     0.070571  3.077310e-53 -0.640604  0.036976  \n",
       "9191     0.094806  3.458420e-21 -0.445408  0.045073  \n",
       "14611    0.124116  2.197750e-05 -0.246023  0.057281  \n",
       "15082    0.113158  7.896400e-06 -0.320288  0.070745  \n",
       "17452    0.112111  1.630390e-49 -0.838766  0.049090  \n",
       "\n",
       "[10 rows x 22 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_sig.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mashr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.116258Z",
     "iopub.status.busy": "2023-06-08T13:10:36.115749Z",
     "iopub.status.idle": "2023-06-08T13:10:36.129845Z",
     "shell.execute_reply": "2023-06-08T13:10:36.129172Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>lfsr</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>external_gene_name</th>\n",
       "      <th>entrezgene</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4943</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>chr22:25059120:A:C</td>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>ENSG00000272977</td>\n",
       "      <td>chr22</td>\n",
       "      <td>25476218</td>\n",
       "      <td>25479971</td>\n",
       "      <td>0.011928</td>\n",
       "      <td>0.323847</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000272977</td>\n",
       "      <td>AL008721.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8285</th>\n",
       "      <td>Caudate</td>\n",
       "      <td>ENSG00000270605.1</td>\n",
       "      <td>chr1:28102893:G:C</td>\n",
       "      <td>ENSG00000270605.1</td>\n",
       "      <td>ENSG00000270605</td>\n",
       "      <td>chr1</td>\n",
       "      <td>28239509</td>\n",
       "      <td>28241453</td>\n",
       "      <td>0.049873</td>\n",
       "      <td>-0.261352</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ENSG00000270605</td>\n",
       "      <td>AL353622.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       region            gene_id          variant_id         gencode_id  \\\n",
       "4943  Caudate  ENSG00000272977.1  chr22:25059120:A:C  ENSG00000272977.1   \n",
       "8285  Caudate  ENSG00000270605.1   chr1:28102893:G:C  ENSG00000270605.1   \n",
       "\n",
       "            gene_name seqnames     start       end      lfsr  posterior_mean  \\\n",
       "4943  ENSG00000272977    chr22  25476218  25479971  0.011928        0.323847   \n",
       "8285  ENSG00000270605     chr1  28239509  28241453  0.049873       -0.261352   \n",
       "\n",
       "     feature_type  ensembl_gene_id external_gene_name  entrezgene description  \n",
       "4943         Gene  ENSG00000272977         AL008721.2         NaN         NaN  \n",
       "8285         Gene  ENSG00000270605         AL353622.1         NaN         NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap = bs[(bs['gene_id'].isin(gtex_sig.ensembl_gene_id))].drop_duplicates()\n",
    "print(gtex_overlap.shape)\n",
    "gtex_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.133023Z",
     "iopub.status.busy": "2023-06-08T13:10:36.132768Z",
     "iopub.status.idle": "2023-06-08T13:10:36.137172Z",
     "shell.execute_reply": "2023-06-08T13:10:36.136494Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.2890173410404624"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_overlap.shape[0]/bs.shape[0] * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.140607Z",
     "iopub.status.busy": "2023-06-08T13:10:36.140419Z",
     "iopub.status.idle": "2023-06-08T13:10:36.156404Z",
     "shell.execute_reply": "2023-06-08T13:10:36.155629Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hugo_gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal_sb</th>\n",
       "      <th>slope_sb</th>\n",
       "      <th>slope_se_sb</th>\n",
       "      <th>...</th>\n",
       "      <th>qval</th>\n",
       "      <th>pval_nominal_f</th>\n",
       "      <th>slope_f</th>\n",
       "      <th>slope_se_f</th>\n",
       "      <th>pval_nominal_m</th>\n",
       "      <th>slope_m</th>\n",
       "      <th>slope_se_m</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>297207</th>\n",
       "      <td>ENSG00000270605.1</td>\n",
       "      <td>RP5-1092A3.4</td>\n",
       "      <td>antisense</td>\n",
       "      <td>chr1_28223937_C_T_b38</td>\n",
       "      <td>rs481640</td>\n",
       "      <td>Skin_Not_Sun_Exposed_Suprapubic</td>\n",
       "      <td>0.323985</td>\n",
       "      <td>0.000237</td>\n",
       "      <td>-0.241524</td>\n",
       "      <td>0.065161</td>\n",
       "      <td>...</td>\n",
       "      <td>0.208195</td>\n",
       "      <td>4.046590e-21</td>\n",
       "      <td>-0.760755</td>\n",
       "      <td>0.063142</td>\n",
       "      <td>6.395880e-13</td>\n",
       "      <td>-0.599143</td>\n",
       "      <td>0.07247</td>\n",
       "      <td>9.917080e-60</td>\n",
       "      <td>-0.591873</td>\n",
       "      <td>0.031047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338770</th>\n",
       "      <td>ENSG00000272977.1</td>\n",
       "      <td>CTA-390C10.10</td>\n",
       "      <td>sense_intronic</td>\n",
       "      <td>chr22_25459662_G_A_b38</td>\n",
       "      <td>rs6004655</td>\n",
       "      <td>Spleen</td>\n",
       "      <td>0.167401</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.413128</td>\n",
       "      <td>0.097856</td>\n",
       "      <td>...</td>\n",
       "      <td>0.212883</td>\n",
       "      <td>7.931210e-14</td>\n",
       "      <td>-1.129030</td>\n",
       "      <td>0.107327</td>\n",
       "      <td>5.155260e-16</td>\n",
       "      <td>-1.276590</td>\n",
       "      <td>0.10469</td>\n",
       "      <td>4.896640e-51</td>\n",
       "      <td>-1.193670</td>\n",
       "      <td>0.056659</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_gene_id   hugo_gene_id       gene_type  \\\n",
       "297207  ENSG00000270605.1   RP5-1092A3.4       antisense   \n",
       "338770  ENSG00000272977.1  CTA-390C10.10  sense_intronic   \n",
       "\n",
       "                    variant_id      rs_id                           Tissue  \\\n",
       "297207   chr1_28223937_C_T_b38   rs481640  Skin_Not_Sun_Exposed_Suprapubic   \n",
       "338770  chr22_25459662_G_A_b38  rs6004655                           Spleen   \n",
       "\n",
       "             maf  pval_nominal_sb  slope_sb  slope_se_sb  ...      qval  \\\n",
       "297207  0.323985         0.000237 -0.241524     0.065161  ...  0.208195   \n",
       "338770  0.167401         0.000038  0.413128     0.097856  ...  0.212883   \n",
       "\n",
       "        pval_nominal_f   slope_f  slope_se_f  pval_nominal_m   slope_m  \\\n",
       "297207    4.046590e-21 -0.760755    0.063142    6.395880e-13 -0.599143   \n",
       "338770    7.931210e-14 -1.129030    0.107327    5.155260e-16 -1.276590   \n",
       "\n",
       "        slope_se_m  pval_nominal     slope  slope_se  \n",
       "297207     0.07247  9.917080e-60 -0.591873  0.031047  \n",
       "338770     0.10469  4.896640e-51 -1.193670  0.056659  \n",
       "\n",
       "[2 rows x 22 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtex_sig[(gtex_sig['ensembl_gene_id'].isin(bs.gene_id))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.161178Z",
     "iopub.status.busy": "2023-06-08T13:10:36.160921Z",
     "iopub.status.idle": "2023-06-08T13:10:36.167556Z",
     "shell.execute_reply": "2023-06-08T13:10:36.166895Z"
    }
   },
   "outputs": [],
   "source": [
    "gtex_sig[(gtex_sig['ensembl_gene_id'].isin(bs.gene_id))]\\\n",
    "    .to_csv(\"siEQTL_gtex_comparison.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Session information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-08T13:10:36.171067Z",
     "iopub.status.busy": "2023-06-08T13:10:36.170816Z",
     "iopub.status.idle": "2023-06-08T13:10:45.216918Z",
     "shell.execute_reply": "2023-06-08T13:10:45.214865Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "numpy               1.20.2\n",
       "pandas              1.4.3\n",
       "pyhere              1.0.0\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "anyio               NA\n",
       "arrow               1.2.3\n",
       "asttokens           NA\n",
       "attr                21.4.0\n",
       "babel               2.10.3\n",
       "backcall            0.2.0\n",
       "certifi             2021.10.08\n",
       "chardet             4.0.0\n",
       "charset_normalizer  2.0.12\n",
       "cloudpickle         2.1.0\n",
       "cython_runtime      NA\n",
       "dateutil            2.8.2\n",
       "debugpy             1.6.3\n",
       "decorator           5.1.1\n",
       "executing           0.10.0\n",
       "fastjsonschema      NA\n",
       "fqdn                NA\n",
       "google              NA\n",
       "idna                2.10\n",
       "importlib_metadata  NA\n",
       "ipykernel           6.15.1\n",
       "ipython_genutils    0.2.0\n",
       "isoduration         NA\n",
       "jedi                0.18.1\n",
       "jinja2              3.1.1\n",
       "json5               NA\n",
       "jsonpointer         2.3\n",
       "jsonschema          4.17.3\n",
       "jupyter_server      1.23.5\n",
       "jupyterlab_server   2.19.0\n",
       "markupsafe          2.1.1\n",
       "mpl_toolkits        NA\n",
       "nbformat            5.4.0\n",
       "packaging           21.3\n",
       "parso               0.8.3\n",
       "pexpect             4.8.0\n",
       "pickleshare         0.7.5\n",
       "pkg_resources       NA\n",
       "platformdirs        3.0.0\n",
       "prometheus_client   NA\n",
       "prompt_toolkit      3.0.30\n",
       "psutil              5.9.1\n",
       "ptyprocess          0.7.0\n",
       "pure_eval           0.2.2\n",
       "pvectorc            NA\n",
       "pyarrow             7.0.0\n",
       "pydev_ipython       NA\n",
       "pydevconsole        NA\n",
       "pydevd              2.8.0\n",
       "pydevd_file_utils   NA\n",
       "pydevd_plugins      NA\n",
       "pydevd_tracing      NA\n",
       "pygments            2.13.0\n",
       "pyrsistent          NA\n",
       "pytz                2022.1\n",
       "requests            2.28.2\n",
       "rfc3339_validator   0.1.4\n",
       "rfc3986_validator   0.1.1\n",
       "send2trash          NA\n",
       "six                 1.16.0\n",
       "sniffio             1.2.0\n",
       "stack_data          0.4.0\n",
       "terminado           0.15.0\n",
       "tornado             6.2\n",
       "traitlets           5.9.0\n",
       "typing_extensions   NA\n",
       "uri_template        NA\n",
       "urllib3             1.26.8\n",
       "wcwidth             0.2.5\n",
       "webcolors           1.11.1\n",
       "websocket           1.4.0\n",
       "zipp                NA\n",
       "zmq                 25.0.0\n",
       "zstandard           0.18.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.4.0\n",
       "jupyter_client      8.0.2\n",
       "jupyter_core        5.2.0\n",
       "jupyterlab          3.6.1\n",
       "notebook            6.4.12\n",
       "-----\n",
       "Python 3.9.10 (main, Feb 22 2022, 16:34:24) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]\n",
       "Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.17\n",
       "-----\n",
       "Session information updated at 2023-06-08 09:10\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
